{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PGB1 logistic regression\n",
    "\n",
    "The aim is to build a simple logistic regression model that predicts the probaility of presence or absence of the PG_binding_1 domain in a genome given the distribution of presence or absence of all other Pfam & TIGR domains. Phylogeny is also added (taxonomic rank: order) to manage confounding variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as plt\n",
    "import seaborn as sns\n",
    "import statsmodels.api as sm\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import accuracy_score, confusion_matrix, classification_report\n",
    "from sklearn.utils.class_weight import compute_class_weight\n",
    "\n",
    "cwd = os.getcwd()\n",
    "if cwd.endswith('notebook'):\n",
    "    os.chdir('..')\n",
    "    cwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme(palette='colorblind', font_scale=1.3)\n",
    "palette = sns.color_palette().as_hex()\n",
    "\n",
    "data_folder = Path('./data/')\n",
    "assert data_folder.is_dir()\n",
    "\n",
    "db_proka = Path('../db_proka/')\n",
    "assert db_proka.is_dir()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_domain = 'PG_binding_1'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pfam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genomes: 32,507\n",
      "Number of domains: 3,981\n"
     ]
    }
   ],
   "source": [
    "pfam_df = pd.read_csv(data_folder / 'pfam_bacteria.csv', index_col='assembly_accession')\n",
    "assert target_domain in set(pfam_df.columns)\n",
    "\n",
    "print(f'Number of genomes: {len(pfam_df):,}')\n",
    "print(f'Number of domains: {len(pfam_df.columns):,}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PG_binding_1 is present in 64% of all genomes.\n"
     ]
    }
   ],
   "source": [
    "p = 100 * len(pfam_df[pfam_df[target_domain] == 1]) / len(pfam_df)\n",
    "\n",
    "print(f'{target_domain} is present in {p:.0f}% of all genomes.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TIGR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "TIGR_df = pd.read_csv(data_folder / 'TIGR_bacteria.csv', index_col='assembly_accession')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phylogeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylogeny_df = pd.read_csv(data_folder / 'taxonomy_order_bacteria.csv', index_col='assembly_accession')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df = pd.merge(\n",
    "    pfam_df,\n",
    "    TIGR_df,\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    how='inner',\n",
    ")\n",
    "data_df = pd.merge(\n",
    "    data_df,\n",
    "    phylogeny_df,\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    how='inner',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genomes: 32,507\n",
      "Number of columns: 6,704\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of genomes: {len(data_df):,}')\n",
    "print(f'Number of columns: {len(data_df.columns):,}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PGB1 exploration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdb_metadata = pd.read_csv(db_proka / 'gtdb_metadata.csv', index_col='ncbi_accession')\n",
    "gtdb_metadata['domain'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[0].replace('d__', ''))\n",
    "gtdb_metadata['gtdb_phylum'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[1].replace('p__', ''))\n",
    "gtdb_metadata['gtdb_class'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[2].replace('c__', ''))\n",
    "gtdb_metadata['gtdb_order'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[3].replace('o__', ''))\n",
    "gtdb_metadata['gtdb_family'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[4].replace('f__', ''))\n",
    "gtdb_metadata['gtdb_genus'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[5].replace('g__', ''))\n",
    "gtdb_metadata['gtdb_species'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[6].replace('s__', ''))\n",
    "\n",
    "bacterial_genomes = gtdb_metadata[gtdb_metadata['domain'] == 'Bacteria']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PG_binding_1</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000007325.1</th>\n",
       "      <td>1</td>\n",
       "      <td>Fusobacteriota</td>\n",
       "      <td>Fusobacteriia</td>\n",
       "      <td>Fusobacteriales</td>\n",
       "      <td>Fusobacteriaceae</td>\n",
       "      <td>Fusobacterium</td>\n",
       "      <td>Fusobacterium nucleatum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000008885.1</th>\n",
       "      <td>0</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Wigglesworthia</td>\n",
       "      <td>Wigglesworthia glossinidia_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000009845.1</th>\n",
       "      <td>0</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Acholeplasmatales</td>\n",
       "      <td>Acholeplasmataceae</td>\n",
       "      <td>Phytoplasma</td>\n",
       "      <td>Phytoplasma sp000009845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000010565.1</th>\n",
       "      <td>1</td>\n",
       "      <td>Bacillota_B</td>\n",
       "      <td>Desulfotomaculia</td>\n",
       "      <td>Desulfotomaculales</td>\n",
       "      <td>Pelotomaculaceae</td>\n",
       "      <td>Pelotomaculum</td>\n",
       "      <td>Pelotomaculum thermopropionicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000011445.1</th>\n",
       "      <td>0</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Mycoplasmatales</td>\n",
       "      <td>Mycoplasmataceae</td>\n",
       "      <td>Mycoplasma</td>\n",
       "      <td>Mycoplasma mycoides</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 PG_binding_1     gtdb_phylum           gtdb_class  \\\n",
       "GCA_000007325.1             1  Fusobacteriota        Fusobacteriia   \n",
       "GCA_000008885.1             0  Pseudomonadota  Gammaproteobacteria   \n",
       "GCA_000009845.1             0       Bacillota              Bacilli   \n",
       "GCA_000010565.1             1     Bacillota_B     Desulfotomaculia   \n",
       "GCA_000011445.1             0       Bacillota              Bacilli   \n",
       "\n",
       "                         gtdb_order         gtdb_family      gtdb_genus  \\\n",
       "GCA_000007325.1     Fusobacteriales    Fusobacteriaceae   Fusobacterium   \n",
       "GCA_000008885.1    Enterobacterales  Enterobacteriaceae  Wigglesworthia   \n",
       "GCA_000009845.1   Acholeplasmatales  Acholeplasmataceae     Phytoplasma   \n",
       "GCA_000010565.1  Desulfotomaculales    Pelotomaculaceae   Pelotomaculum   \n",
       "GCA_000011445.1     Mycoplasmatales    Mycoplasmataceae      Mycoplasma   \n",
       "\n",
       "                                    gtdb_species  \n",
       "GCA_000007325.1          Fusobacterium nucleatum  \n",
       "GCA_000008885.1     Wigglesworthia glossinidia_A  \n",
       "GCA_000009845.1          Phytoplasma sp000009845  \n",
       "GCA_000010565.1  Pelotomaculum thermopropionicum  \n",
       "GCA_000011445.1              Mycoplasma mycoides  "
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pgb1_metadata = pd.merge(\n",
    "    pfam_df[[target_domain]],\n",
    "    bacterial_genomes[['gtdb_phylum', 'gtdb_class', 'gtdb_order', 'gtdb_family', 'gtdb_genus', 'gtdb_species']],\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    ")\n",
    "pgb1_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gtdb_phylum: 126 out of 155 (81 %)\n",
      "gtdb_class: 300 out of 385 (78 %)\n",
      "gtdb_order: 840 out of 1,172 (72 %)\n",
      "gtdb_family: 1,850 out of 2,882 (64 %)\n",
      "gtdb_genus: 5,860 out of 12,553 (47 %)\n"
     ]
    }
   ],
   "source": [
    "taxonomy_to_subset = {}\n",
    "for taxonomy in ['gtdb_phylum', 'gtdb_class', 'gtdb_order', 'gtdb_family', 'gtdb_genus']:\n",
    "    g = pgb1_metadata[[taxonomy, target_domain]].groupby(taxonomy).agg(['sum', 'count'])\n",
    "\n",
    "    subset_to_consider = sorted(set(g[\n",
    "        #(g[(target_domain, 'sum')] > 0) &\n",
    "        (g[(target_domain, 'count')] > 1)\n",
    "    ].index))\n",
    "\n",
    "    taxonomy_to_subset[taxonomy] = subset_to_consider\n",
    "\n",
    "    p = 100 * len(subset_to_consider) / len(g.index)\n",
    "    print(f'{taxonomy}: {len(subset_to_consider):,} out of {len(g.index):,} ({p:.0f} %)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Relevant genomes: 25,814 out of 32,507 (79 %)\n"
     ]
    }
   ],
   "source": [
    "pgb1_subset = pgb1_metadata[\n",
    "    pgb1_metadata['gtdb_phylum'].isin(taxonomy_to_subset['gtdb_phylum']) &\n",
    "    pgb1_metadata['gtdb_class'].isin(taxonomy_to_subset['gtdb_class']) &\n",
    "    pgb1_metadata['gtdb_order'].isin(taxonomy_to_subset['gtdb_order']) &\n",
    "    pgb1_metadata['gtdb_family'].isin(taxonomy_to_subset['gtdb_family']) &\n",
    "    pgb1_metadata['gtdb_genus'].isin(taxonomy_to_subset['gtdb_genus'])\n",
    "]\n",
    "\n",
    "p = 100 * len(pgb1_subset) / len(pgb1_metadata)\n",
    "print(f'Relevant genomes: {len(pgb1_subset):,} out of {len(pgb1_metadata):,} ({p:.0f} %)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_df = data_df.loc[pgb1_subset.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genomes: 32,507\n",
      "Number of columns: 6,704\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of genomes: {len(data_df):,}')\n",
    "print(f'Number of columns: {len(data_df.columns):,}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genomes with PG_binding_1: 20,955 (64 %)\n"
     ]
    }
   ],
   "source": [
    "n_genomes_with_domain = data_df[target_domain].sum()\n",
    "p = 100 * n_genomes_with_domain / len(data_df)\n",
    "print(f'Number of genomes with {target_domain}: {n_genomes_with_domain:,} ({p:.0f} %)')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1.4014179  0.77734084]\n"
     ]
    }
   ],
   "source": [
    "response = data_df[target_domain]\n",
    "predictors = data_df.loc[:, data_df.columns != target_domain]\n",
    "\n",
    "# Add constant term for intercept\n",
    "predictors = sm.add_constant(predictors)\n",
    "\n",
    "# Split the data into training and validation sets\n",
    "X_train, X_val, y_train, y_val = train_test_split(predictors, response, test_size=0.1, random_state=123)\n",
    "\n",
    "# Calculate class weights\n",
    "class_weights = compute_class_weight(class_weight='balanced', classes=np.array([0, 1]), y=y_train)\n",
    "weights = np.array([class_weights[0] if x == 0 else class_weights[1] for x in y_train])\n",
    "print(class_weights)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Train logistic regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/Caskroom/miniforge/base/envs/cwb/lib/python3.11/site-packages/statsmodels/genmod/generalized_linear_model.py:1464: UserWarning: Elastic net fitting did not converge\n",
      "  warnings.warn(\"Elastic net fitting did not converge\")\n"
     ]
    }
   ],
   "source": [
    "reg_param = 1e-3\n",
    "\n",
    "# Fitting the logistic regression model\n",
    "model = sm.GLM(\n",
    "    y_train, \n",
    "    X_train, \n",
    "    family=sm.families.Binomial(), \n",
    "    freq_weights=weights,\n",
    ").fit_regularized(\n",
    "    maxiter=100, \n",
    "    alpha=reg_param,\n",
    "    L1_wt=0.5,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluate model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Validation set performance:\n",
      "Accuracy: 0.8741925561365733\n",
      "Confusion Matrix:\n",
      "[[ 993  121]\n",
      " [ 288 1849]]\n",
      "Classification Report:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.78      0.89      0.83      1114\n",
      "           1       0.94      0.87      0.90      2137\n",
      "\n",
      "    accuracy                           0.87      3251\n",
      "   macro avg       0.86      0.88      0.86      3251\n",
      "weighted avg       0.88      0.87      0.88      3251\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Predict on the validation set\n",
    "y_pred = model.predict(X_val)\n",
    "y_pred_binary = [1 if x >= 0.5 else 0 for x in y_pred]\n",
    "\n",
    "# Calculate validation metrics\n",
    "accuracy = accuracy_score(y_val, y_pred_binary)\n",
    "conf_matrix = confusion_matrix(y_val, y_pred_binary)\n",
    "class_report = classification_report(y_val, y_pred_binary)\n",
    "\n",
    "print(\"\\nValidation set performance:\")\n",
    "print(f\"Accuracy: {accuracy}\")\n",
    "print(\"Confusion Matrix:\")\n",
    "print(conf_matrix)\n",
    "print(\"Classification Report:\")\n",
    "print(class_report)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify top domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'RegularizedResults' object has no attribute 'pvalues'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[175], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Extract p-values and coefficients of the coefficients from the training model\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m p_values \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpvalues\u001b[49m\n\u001b[1;32m      3\u001b[0m coefficients \u001b[38;5;241m=\u001b[39m model\u001b[38;5;241m.\u001b[39mparams\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# Filter significant p-values\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/cwb/lib/python3.11/site-packages/statsmodels/base/wrapper.py:34\u001b[0m, in \u001b[0;36mResultsWrapper.__getattribute__\u001b[0;34m(self, attr)\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m:\n\u001b[1;32m     32\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[0;32m---> 34\u001b[0m obj \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mresults\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mattr\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     35\u001b[0m data \u001b[38;5;241m=\u001b[39m results\u001b[38;5;241m.\u001b[39mmodel\u001b[38;5;241m.\u001b[39mdata\n\u001b[1;32m     36\u001b[0m how \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_wrap_attrs\u001b[38;5;241m.\u001b[39mget(attr)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'RegularizedResults' object has no attribute 'pvalues'"
     ]
    }
   ],
   "source": [
    "# Extract p-values and coefficients of the coefficients from the training model\n",
    "p_values = model.pvalues\n",
    "coefficients = model.params\n",
    "\n",
    "# Filter significant p-values\n",
    "significant_vars = p_values[p_values < 1e-2].sort_values()\n",
    "\n",
    "# Removing the intercept from the significant variables\n",
    "try:\n",
    "    significant_vars = significant_vars.drop('const')\n",
    "except:\n",
    "    pass\n",
    "\n",
    "# Extracting the values of the significant parameters\n",
    "significant_coefficients = coefficients[significant_vars.index]\n",
    "sorted_significant_coeff = sorted(significant_coefficients.items(), reverse=True, key=lambda t: t[1])\n",
    "\n",
    "# Display the ranked significant regressors and their coefficients\n",
    "print(\"\\nRanked significant regressors from the training set and their coefficients:\")\n",
    "for i, (var, coef) in enumerate(sorted_significant_coeff):\n",
    "    print(f\"{var}: {coef:.3f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scaffold: 1.906\n",
      "PG_binding_2: 1.567\n",
      "TIGR02283: 1.151\n",
      "SpoIIIAC: 0.557\n",
      "Spore_III_AB: 0.541\n",
      "TIGR01925: 0.424\n",
      "TIGR01424: 0.424\n",
      "Com_YlbF: 0.406\n",
      "TIGR02848: 0.343\n",
      "Peptidase_M15_2: 0.335\n",
      "TIGR00247: 0.335\n",
      "Peptidase_S11: 0.327\n",
      "TIGR01134: 0.289\n",
      "SpoIIE_N: 0.276\n",
      "YkuD: 0.273\n",
      "TIGR01646: 0.269\n",
      "TIGR00690: 0.267\n",
      "DUF1540: 0.266\n",
      "Peptidase_S41: 0.265\n",
      "RimP_N: 0.264\n",
      "Peptidase_M14: 0.249\n",
      "HTH_40: 0.242\n",
      "TIGR01740: 0.239\n",
      "TIGR00225: 0.235\n",
      "TIGR01449: 0.229\n",
      "ExoD: 0.228\n",
      "TIGR00858: 0.223\n",
      "DUF378: 0.220\n",
      "FxsA: 0.218\n",
      "ATP-gua_Ptrans: 0.216\n",
      "Peptidase_U4: 0.214\n",
      "TIGR00117: 0.211\n",
      "TIGR03998: 0.210\n",
      "5-FTHF_cyc-lig: 0.209\n",
      "DUF4430: 0.206\n",
      "HTH_ParB: 0.198\n",
      "PGDH_inter: 0.197\n",
      "TIGR02483: 0.197\n",
      "TIGR00706: 0.196\n",
      "LpxI_C: 0.196\n",
      "Cupin_7: 0.192\n",
      "HATPase_c_2: 0.191\n",
      "DUF808: 0.191\n",
      "LMWPc: 0.189\n",
      "TIGR01459: 0.189\n",
      "CorA: 0.189\n",
      "SepF: 0.188\n",
      "Thioredox_DsbH: 0.188\n",
      "UPF0047: 0.187\n",
      "GerA: 0.186\n",
      "Dodecin: 0.185\n",
      "TIGR04018: 0.183\n",
      "Toprim_C_rpt: 0.182\n",
      "TIGR01887: 0.179\n",
      "TIGR03599: 0.179\n",
      "o_Erysipelotrichales: 0.178\n",
      "CSD2: 0.175\n",
      "PEPCK_N: 0.175\n",
      "Fn3-like: 0.173\n",
      "TIGR00611: 0.172\n",
      "FBPase_2: 0.170\n",
      "TIGR01710: 0.167\n",
      "WS_DGAT_C: 0.165\n",
      "TIGR01490: 0.163\n",
      "TIGR02079: 0.163\n",
      "APP1_cat: 0.163\n",
      "PhoH: 0.162\n",
      "TIGR00970: 0.160\n",
      "DUF480: 0.157\n",
      "TPR_16: 0.156\n",
      "DUF4131: 0.155\n",
      "PPK2: 0.155\n",
      "RelA_AH_RIS: 0.155\n",
      "bMG6: 0.155\n",
      "Peptidase_M48: 0.154\n",
      "FmdA_AmdA: 0.153\n",
      "Lip_A_acyltrans: 0.151\n",
      "DUF3471: 0.150\n",
      "DUF3971: 0.149\n",
      "CysG_dimeriser: 0.149\n",
      "TIGR02098: 0.149\n",
      "DUF512: 0.148\n",
      "Peptidase_M74: 0.147\n",
      "TIGR03300: 0.146\n",
      "HRDC: 0.146\n",
      "YMF19: 0.145\n",
      "NAPRTase_N: 0.144\n",
      "TIGR00207: 0.144\n",
      "TIGR00135: 0.144\n",
      "MGDG_synth: 0.144\n",
      "DivIVA: 0.143\n",
      "Inhibitor_I9: 0.143\n",
      "TIGR03904: 0.142\n",
      "BPL_LplA_LipB_2: 0.139\n",
      "TIGR02531: 0.138\n",
      "DUF937: 0.138\n",
      "Amidase_2: 0.138\n",
      "TIGR01364: 0.137\n",
      "Sulfotransfer_2: 0.137\n",
      "Aldose_epim: 0.137\n",
      "ATP-synt_DE: 0.136\n",
      "FAA_hydrolase_N: 0.136\n",
      "DUF2934: 0.136\n",
      "MoCoBD_1: 0.136\n",
      "TIGR01933: 0.135\n",
      "CSD: 0.134\n",
      "Phage_base_V: 0.133\n",
      "TIGR00656: 0.133\n",
      "TIGR02644: 0.133\n",
      "MbhD: 0.133\n",
      "Fer4_NifH: 0.133\n",
      "Dak1_2: 0.132\n",
      "TIGR00528: 0.132\n",
      "Chlorophyllase2: 0.132\n",
      "DUF554: 0.131\n",
      "Thia_YuaJ: 0.131\n",
      "TIGR01296: 0.131\n",
      "TIGR00040: 0.130\n",
      "TIGR01001: 0.129\n",
      "Peptidase_C14: 0.129\n",
      "TIGR01613: 0.128\n",
      "WhiA_N: 0.127\n",
      "AsmA_2: 0.127\n",
      "TIGR02776: 0.126\n",
      "AAA_22: 0.126\n",
      "OmpA: 0.125\n",
      "LptD_N: 0.125\n",
      "Glu_cyclase_2: 0.124\n",
      "Methyltr_RsmF_N: 0.124\n",
      "TIGR02838: 0.124\n",
      "Thioredoxin_9: 0.123\n",
      "Acyl-CoA_dh_C: 0.123\n",
      "TIGR01266: 0.123\n",
      "Erythro_esteras: 0.123\n",
      "DUF1292: 0.123\n",
      "DZR: 0.122\n",
      "TPP_enzyme_M_2: 0.122\n",
      "MHYT: 0.121\n",
      "Fe_hyd_lg_C: 0.121\n",
      "DUF4136: 0.120\n",
      "VitK2_biosynth: 0.120\n",
      "ClpS: 0.119\n",
      "DUF2232: 0.119\n",
      "Radical_SAM_2: 0.119\n",
      "TIGR00576: 0.119\n",
      "TIGR01726: 0.118\n",
      "FTP: 0.117\n",
      "TIGR03458: 0.117\n",
      "Fer4_15: 0.117\n",
      "Ligase_CoA: 0.116\n",
      "NYN: 0.116\n",
      "PGPGW: 0.116\n",
      "LpxI_N: 0.115\n",
      "TIGR01982: 0.115\n",
      "Acetyltransf_5: 0.115\n",
      "MraY_sig1: 0.115\n",
      "BPL_C: 0.113\n",
      "SbcD_C: 0.113\n",
      "DUF4870: 0.112\n",
      "SurA_N: 0.112\n",
      "Lipocalin_2: 0.112\n",
      "TIGR02967: 0.112\n",
      "Histone_HNS: 0.111\n",
      "o_Christensenellales: 0.111\n",
      "TIGR01317: 0.111\n",
      "TIGR01389: 0.111\n",
      "TIGR01340: 0.111\n",
      "TIGR00513: 0.110\n",
      "YdjC: 0.110\n",
      "UPF0014: 0.110\n",
      "TIGR01973: 0.110\n",
      "TIGR03605: 0.110\n",
      "DUF924: 0.109\n",
      "TPR_4: 0.109\n",
      "Rce1-like: 0.109\n",
      "CDO_I: 0.108\n",
      "RAMPs: 0.108\n",
      "NYN_YacP: 0.107\n",
      "Actin: 0.107\n",
      "TIGR00525: 0.107\n",
      "TIGR03696: 0.107\n",
      "CpsB_CapC: 0.106\n",
      "CHAP: 0.106\n",
      "PFK: 0.105\n",
      "RuBisCO_large: 0.105\n",
      "HAMP_2: 0.104\n",
      "CbiX: 0.104\n",
      "TilS: 0.104\n",
      "TIGR02429: 0.103\n",
      "5TM-5TMR_LYT: 0.103\n",
      "NUDIX_4: 0.102\n",
      "TIGR00084: 0.102\n",
      "DUF4162: 0.102\n",
      "sCache_like: 0.101\n",
      "TIGR00274: 0.101\n",
      "Asp_Arg_Hydrox: 0.101\n",
      "Peptidase_S8: 0.101\n",
      "TIGR01784: 0.101\n",
      "MMPL: 0.101\n",
      "TIGR01260: 0.101\n",
      "CAP: 0.100\n",
      "Creatinase_N: 0.099\n",
      "PrpF: 0.098\n",
      "DUF2157: 0.098\n",
      "SUI1: 0.098\n",
      "Peptidase_M15_3: 0.097\n",
      "Cytochrom_CIII: 0.097\n",
      "DUF2064: 0.097\n",
      "TIGR00374: 0.096\n",
      "Eco57I: 0.096\n",
      "DBI_PRT: 0.096\n",
      "TIGR00219: 0.096\n",
      "DUF1638: 0.096\n",
      "PrmC_N: 0.095\n",
      "Phage_holin_3_6: 0.095\n",
      "TIGR01172: 0.095\n",
      "WbqC: 0.094\n",
      "Zn_protease: 0.093\n",
      "TIGR00521: 0.093\n",
      "TIGR03828: 0.093\n",
      "TIGR03296: 0.093\n",
      "TIGR03542: 0.093\n",
      "DUF370: 0.093\n",
      "PilN: 0.093\n",
      "TetR_C_7: 0.093\n",
      "DUF418: 0.093\n",
      "Na_sulph_symp: 0.093\n",
      "TIGR03426: 0.092\n",
      "Coat_F: 0.092\n",
      "Peptidase_M4: 0.092\n",
      "Sugar-bind: 0.092\n",
      "AbrB: 0.092\n",
      "DUF3656: 0.091\n",
      "Peptidase_S66C: 0.091\n",
      "TIGR01016: 0.091\n",
      "DAPDH_C: 0.091\n",
      "TIGR00826: 0.089\n",
      "Pro_racemase: 0.089\n",
      "TIGR00705: 0.089\n",
      "TIGR02241: 0.088\n",
      "FG-GAP: 0.088\n",
      "TIGR02799: 0.088\n",
      "GHL10: 0.088\n",
      "SBF_like: 0.088\n",
      "QueT: 0.088\n",
      "Mt_ATP-synt_B: 0.087\n",
      "TIGR01430: 0.087\n",
      "TIGR03558: 0.086\n",
      "PriA_C: 0.086\n",
      "RIO1: 0.086\n",
      "Melibiase_C: 0.086\n",
      "Gly-zipper_Omp: 0.086\n",
      "bPH_1: 0.085\n",
      "TIGR02154: 0.085\n",
      "LNT_N: 0.085\n",
      "DUF520: 0.085\n",
      "SlyX: 0.085\n",
      "NiFe_hyd_SSU_C: 0.084\n",
      "FtsJ: 0.084\n",
      "ANF_receptor: 0.083\n",
      "DUF2087: 0.083\n",
      "NDK: 0.083\n",
      "TIGR01362: 0.082\n",
      "Cyanophycin_syn: 0.082\n",
      "DUF4142: 0.082\n",
      "MaoC_dehydrat_N: 0.082\n",
      "TIGR01221: 0.082\n",
      "TamB: 0.081\n",
      "TIGR01222: 0.081\n",
      "TIGR02823: 0.081\n",
      "Glyco_hydro_57: 0.081\n",
      "PEPCK_ATP: 0.081\n",
      "TIGR00556: 0.081\n",
      "ImpA_N: 0.081\n",
      "TIGR00791: 0.080\n",
      "MauE: 0.080\n",
      "OpuAC: 0.080\n",
      "GYF_2: 0.080\n",
      "AidB_N: 0.080\n",
      "TIGR00091: 0.080\n",
      "GDPD_2: 0.080\n",
      "Pur_DNA_glyco: 0.079\n",
      "TIGR00443: 0.079\n",
      "DUF5011: 0.079\n",
      "TIGR01451: 0.079\n",
      "CwfJ_C_1: 0.078\n",
      "DRTGG: 0.078\n",
      "Phage_int_SAM_3: 0.078\n",
      "SLC35F: 0.078\n",
      "CobA_CobO_BtuR: 0.077\n",
      "UDPGT: 0.077\n",
      "TIGR01139: 0.077\n",
      "CinA_KH: 0.077\n",
      "cobW: 0.076\n",
      "FtsQ_DivIB_C: 0.076\n",
      "DNA_photolyase: 0.076\n",
      "TIGR00795: 0.076\n",
      "zf-trcl: 0.075\n",
      "KAP_NTPase: 0.075\n",
      "Abhydrolase_3: 0.075\n",
      "TIGR00759: 0.075\n",
      "HigB_toxin: 0.075\n",
      "DUF3298: 0.074\n",
      "LCAT: 0.074\n",
      "HTH_AsnC-type: 0.074\n",
      "Ser_hydrolase: 0.073\n",
      "Aconitase_2_N: 0.073\n",
      "Acetyltransf_CG: 0.073\n",
      "FAD_binding_6: 0.073\n",
      "TSP_3: 0.073\n",
      "2-oxogl_dehyd_N: 0.072\n",
      "TIGR03423: 0.072\n",
      "ROK: 0.071\n",
      "Glyco_hydro_25: 0.071\n",
      "TIGR00191: 0.071\n",
      "TIGR01341: 0.071\n",
      "ACT_6: 0.071\n",
      "LTXXQ: 0.071\n",
      "3-HAO: 0.070\n",
      "DUF421: 0.070\n",
      "AIG1: 0.070\n",
      "FIST_C: 0.070\n",
      "DNA_pol3_gamma3: 0.070\n",
      "BE_C: 0.070\n",
      "DUF2200: 0.070\n",
      "TIGR02176: 0.069\n",
      "DAO_C: 0.069\n",
      "Glutaredoxin: 0.069\n",
      "4PPT_N: 0.069\n",
      "DUF423: 0.069\n",
      "MucB_RseB: 0.069\n",
      "FlpD: 0.069\n",
      "Glucosaminidase: 0.068\n",
      "Terminase_6C: 0.068\n",
      "NfeD: 0.067\n",
      "Trm112p: 0.067\n",
      "Pentapeptide_4: 0.067\n",
      "DNApol3-delta_C: 0.067\n",
      "TIGR03367: 0.067\n",
      "PDDEXK_2: 0.067\n",
      "Peptidase_S9_N: 0.067\n",
      "TIGR00178: 0.067\n",
      "YbjN: 0.066\n",
      "Macro: 0.066\n",
      "DUF819: 0.065\n",
      "TIGR02150: 0.065\n",
      "SMC_hinge: 0.064\n",
      "2OG-FeII_Oxy_2: 0.064\n",
      "Peptidase_M15_4: 0.064\n",
      "Collagen: 0.064\n",
      "TIGR00462: 0.064\n",
      "RVT_1: 0.064\n",
      "FHA: 0.063\n",
      "DUF3987: 0.062\n",
      "Phosphorylase: 0.062\n",
      "FrhB_FdhB_C: 0.062\n",
      "TIGR00007: 0.062\n",
      "TIGR00676: 0.061\n",
      "TrmB: 0.061\n",
      "TIGR01968: 0.061\n",
      "Glyco_hydro_36N: 0.061\n",
      "Transp_cyt_pur: 0.061\n",
      "TIGR00245: 0.061\n",
      "DUF2188: 0.060\n",
      "Ank: 0.060\n",
      "DUF1003: 0.060\n",
      "TIGR00095: 0.060\n",
      "DUF3362: 0.060\n",
      "TIGR01342: 0.060\n",
      "LuxS: 0.060\n",
      "AsmA: 0.060\n",
      "Flavodoxin_2: 0.059\n",
      "MmeI_TRD: 0.059\n",
      "Met_gamma_lyase: 0.059\n",
      "FtsL: 0.059\n",
      "zf-FPG_IleRS: 0.059\n",
      "TIGR00654: 0.059\n",
      "FliM: 0.059\n",
      "TIGR03302: 0.058\n",
      "UPF0016: 0.058\n",
      "Ribonuclease: 0.058\n",
      "TIGR00639: 0.058\n",
      "HD_assoc: 0.058\n",
      "Choline_kinase: 0.057\n",
      "TPR_5: 0.057\n",
      "NHL: 0.057\n",
      "TIGR01133: 0.057\n",
      "TIGR00278: 0.056\n",
      "zf-ISL3: 0.056\n",
      "ATP-synt_I: 0.056\n",
      "DUF4157: 0.056\n",
      "TIGR00738: 0.056\n",
      "DUF2231: 0.056\n",
      "Abhydrolase_11: 0.056\n",
      "DUF11: 0.055\n",
      "PAAR_motif: 0.055\n",
      "DUF4065: 0.055\n",
      "TIGR00554: 0.055\n",
      "TIGR00367: 0.055\n",
      "BBP2: 0.055\n",
      "DUF4384: 0.055\n",
      "CxxCxxCC: 0.054\n",
      "Y2_Tnp: 0.054\n",
      "Glyco_tranf_2_4: 0.054\n",
      "Ribosomal_S30AE: 0.054\n",
      "Yip1: 0.054\n",
      "H2O2_YaaD: 0.054\n",
      "DUF4126: 0.054\n",
      "iPGM_N: 0.054\n",
      "DUF2065: 0.053\n",
      "Catalase_C: 0.053\n",
      "PCC_BT: 0.052\n",
      "Ppx-GppA: 0.052\n",
      "Asparaginase_2: 0.052\n",
      "TIGR00409: 0.052\n",
      "TrbI: 0.052\n",
      "TIGR01019: 0.052\n",
      "TetR_C_6: 0.052\n",
      "Terminase_5: 0.052\n",
      "HTH_33: 0.051\n",
      "Cation_ATPase: 0.051\n",
      "ThiF: 0.051\n",
      "Mpo1-like: 0.051\n",
      "DUF6537: 0.050\n",
      "DUF150_C: 0.050\n",
      "TIGR00368: 0.050\n",
      "HTH_Tnp_ISL3: 0.050\n",
      "KH_4: 0.050\n",
      "TIGR01179: 0.049\n",
      "DUF4340: 0.049\n",
      "TIGR00915: 0.049\n",
      "Beta-lactamase2: 0.049\n",
      "Band_7_1: 0.049\n",
      "Acyl_transf_3: 0.049\n",
      "PAS_10: 0.049\n",
      "EutQ: 0.049\n",
      "PepSY_TM: 0.049\n",
      "CofH_C: 0.048\n",
      "Phytase-like: 0.048\n",
      "Fructosamin_kin: 0.048\n",
      "PemK_toxin: 0.048\n",
      "LAB_N: 0.048\n",
      "TRAM: 0.048\n",
      "EutB: 0.048\n",
      "Gly-zipper_YMGG: 0.048\n",
      "TIGR00546: 0.048\n",
      "TIGR02140: 0.047\n",
      "Na_Pi_cotrans: 0.047\n",
      "PMT: 0.047\n",
      "Chromate_transp: 0.047\n",
      "HupE_UreJ_2: 0.047\n",
      "Aldolase: 0.047\n",
      "Methyltransf_24: 0.047\n",
      "DUF6597: 0.046\n",
      "TIGR00479: 0.046\n",
      "TIGR03561: 0.046\n",
      "Spore_permease: 0.045\n",
      "DAGK_prokar: 0.045\n",
      "Xan_ur_permease: 0.045\n",
      "TIGR00619: 0.045\n",
      "Glycoamylase: 0.045\n",
      "YgjP-like: 0.045\n",
      "Rossmann-like: 0.044\n",
      "DUF445: 0.044\n",
      "TIGR01065: 0.044\n",
      "VWA_3: 0.044\n",
      "TIGR00636: 0.044\n",
      "OppC_N: 0.044\n",
      "TIGR00011: 0.043\n",
      "Glyco_hydro_88: 0.043\n",
      "PKD_4: 0.043\n",
      "Polysacc_synt_3: 0.043\n",
      "SUFBD: 0.043\n",
      "GCV_T_C: 0.043\n",
      "TIGR01883: 0.043\n",
      "TIGR00968: 0.043\n",
      "PYNP_C: 0.043\n",
      "Exosortase_EpsH: 0.042\n",
      "BtpA: 0.042\n",
      "MTES_1575: 0.042\n",
      "He_PIG: 0.042\n",
      "B12-binding_2: 0.042\n",
      "Proteasome: 0.042\n",
      "FloA: 0.042\n",
      "TANGO2: 0.042\n",
      "PhdYeFM_antitox: 0.041\n",
      "TIGR00215: 0.041\n",
      "BKACE: 0.041\n",
      "Peptidase_A25: 0.041\n",
      "PTE: 0.040\n",
      "CRS1_YhbY: 0.040\n",
      "VWA_2: 0.040\n",
      "YdjM: 0.040\n",
      "CagE_TrbE_VirB: 0.040\n",
      "VIT: 0.040\n",
      "HMG-CoA_red: 0.040\n",
      "Peptidase_M42: 0.040\n",
      "TIGR03804: 0.040\n",
      "H-kinase_dim: 0.039\n",
      "Fer4_22: 0.039\n",
      "AGL_N: 0.039\n",
      "AAA_16: 0.039\n",
      "TIGR00317: 0.039\n",
      "NMN_transporter: 0.038\n",
      "Arf: 0.038\n",
      "MORN: 0.038\n",
      "HTH_16: 0.038\n",
      "Kinase-PPPase: 0.038\n",
      "DUF294: 0.038\n",
      "Ferric_reduct: 0.038\n",
      "C_GCAxxG_C_C: 0.037\n",
      "Lumazine_bd_2: 0.037\n",
      "CCG: 0.037\n",
      "YlmH_RBD: 0.037\n",
      "Transgly_assoc: 0.037\n",
      "TIGR02454: 0.037\n",
      "DrsE_2: 0.037\n",
      "SASA: 0.037\n",
      "THUMP: 0.036\n",
      "TIGR00453: 0.036\n",
      "HupE_UreJ: 0.036\n",
      "DDE_3: 0.036\n",
      "TIGR00282: 0.036\n",
      "TIGR00515: 0.036\n",
      "Thioredoxin_3: 0.035\n",
      "HicB_lk_antitox: 0.035\n",
      "TIGR01708: 0.035\n",
      "PUA_3: 0.035\n",
      "SATase_N: 0.035\n",
      "PD40: 0.034\n",
      "UPF0227: 0.034\n",
      "DUF6580: 0.034\n",
      "HepT-like: 0.033\n",
      "TPR_12: 0.033\n",
      "TIGR00747: 0.033\n",
      "Methyltransf_14: 0.033\n",
      "TIGR00194: 0.033\n",
      "WES_acyltransf: 0.033\n",
      "ORF6N: 0.032\n",
      "TIGR02229: 0.032\n",
      "DDE_Tnp_1_6: 0.032\n",
      "Lipase_GDSL_3: 0.032\n",
      "SE: 0.031\n",
      "TIGR00436: 0.030\n",
      "ALO: 0.030\n",
      "RNase_H: 0.030\n",
      "TIGR00963: 0.030\n",
      "PRD: 0.030\n",
      "TIGR02156: 0.029\n",
      "LigT_PEase: 0.029\n",
      "DUF1232: 0.029\n",
      "TIGR00773: 0.029\n",
      "PRC: 0.029\n",
      "EF-hand_5: 0.029\n",
      "TIGR00979: 0.029\n",
      "PRA-PH: 0.029\n",
      "TIGR00218: 0.028\n",
      "TIGR03168: 0.028\n",
      "TIGR03139: 0.028\n",
      "Ferritin_2: 0.028\n",
      "DUF1801: 0.027\n",
      "Dockerin_1: 0.027\n",
      "Big_7: 0.027\n",
      "FrsA-like: 0.027\n",
      "SIMPL: 0.027\n",
      "dCache_2: 0.027\n",
      "Sigma70_r4_2: 0.027\n",
      "PAD_porph: 0.026\n",
      "Acetyltransf_9: 0.025\n",
      "TIGR02225: 0.025\n",
      "TIGR00221: 0.025\n",
      "ABC2_membrane_6: 0.025\n",
      "TIGR01515: 0.025\n",
      "TIGR02707: 0.025\n",
      "Cadherin_5: 0.025\n",
      "TIGR02379: 0.025\n",
      "Glyco_tranf_2_5: 0.024\n",
      "GTP-bdg_N: 0.024\n",
      "Abhydrolase_4: 0.024\n",
      "HTH_36: 0.024\n",
      "FUR: 0.024\n",
      "TIGR00516: 0.023\n",
      "HTH_25: 0.023\n",
      "NTP_transf_9: 0.023\n",
      "DUF6531: 0.023\n",
      "TIGR00067: 0.023\n",
      "dTDP_sugar_isom: 0.023\n",
      "TIGR00976: 0.023\n",
      "DUF805: 0.023\n",
      "GPW_gp25: 0.022\n",
      "TIGR00419: 0.022\n",
      "RtcB: 0.022\n",
      "HTH_OrfB_IS605: 0.022\n",
      "Sulfotransfer_3: 0.022\n",
      "HTH_30: 0.022\n",
      "TIGR00020: 0.021\n",
      "DnaB_bind: 0.021\n",
      "HTH_38: 0.021\n",
      "CagX: 0.021\n",
      "TIGR01822: 0.021\n",
      "TIGR01536: 0.020\n",
      "G_glu_transpept: 0.020\n",
      "HisKA_3: 0.019\n",
      "TIGR02081: 0.019\n",
      "CbiQ: 0.019\n",
      "LpxD: 0.019\n",
      "TIGR00638: 0.019\n",
      "DUF2061: 0.018\n",
      "OB_aCoA_assoc: 0.018\n",
      "Calx-beta: 0.018\n",
      "RloB: 0.018\n",
      "Glyco_trans_4_4: 0.018\n",
      "TIGR01550: 0.018\n",
      "DUF2339: 0.018\n",
      "TIGR03652: 0.018\n",
      "ArAE_1: 0.018\n",
      "Ig_mannosidase: 0.017\n",
      "GTP-bdg_M: 0.017\n",
      "Cellulase: 0.017\n",
      "ABC1: 0.017\n",
      "TIGR02276: 0.016\n",
      "FUSC: 0.016\n",
      "YafQ_toxin: 0.016\n",
      "Acetyltransf_8: 0.016\n",
      "zinc_ribbon_5: 0.016\n",
      "TAT_signal: 0.015\n",
      "Host_attach: 0.015\n",
      "LamB_YcsF: 0.015\n",
      "TIGR00080: 0.015\n",
      "DUF1778: 0.015\n",
      "Sterol-sensing: 0.014\n",
      "DUF188: 0.014\n",
      "TIGR00842: 0.014\n",
      "HTH_34: 0.014\n",
      "HTH_17: 0.014\n",
      "T2SSC: 0.014\n",
      "NLPC_P60: 0.014\n",
      "HNH_2: 0.014\n",
      "Thioredoxin_5: 0.014\n",
      "zf-CDGSH: 0.014\n",
      "TIGR00036: 0.014\n",
      "GGACT: 0.014\n",
      "B12-binding: 0.013\n",
      "DivIC: 0.013\n",
      "LemA: 0.013\n",
      "Acetyltransf_17: 0.012\n",
      "FG-GAP_3: 0.012\n",
      "TIGR01704: 0.012\n",
      "TIGR00018: 0.012\n",
      "GNVR: 0.011\n",
      "TRI12: 0.011\n",
      "SNF2_assoc: 0.011\n",
      "TIGR00393: 0.010\n",
      "TIGR00390: 0.010\n",
      "Melibiase: 0.010\n",
      "Fer4_9: 0.010\n",
      "DDE_Tnp_1_2: 0.009\n",
      "Lactonase: 0.009\n",
      "DHquinase_II: 0.008\n",
      "HTH_41: 0.008\n",
      "TIGR03025: 0.008\n",
      "HNH_4: 0.007\n",
      "Bac_luciferase: 0.007\n",
      "TM2: 0.007\n",
      "Glyco_trans_1_2: 0.007\n",
      "TIGR01072: 0.007\n",
      "DUF202: 0.007\n",
      "Glyco_hydro_2_C: 0.007\n",
      "TPR_11: 0.007\n",
      "Pectate_lyase_3: 0.007\n",
      "LAGLIDADG_WhiA: 0.006\n",
      "TIGR00417: 0.006\n",
      "NOV_C: 0.006\n",
      "TIGR02046: 0.005\n",
      "CMAS: 0.005\n",
      "TM_helix: 0.005\n",
      "ABC2_membrane_5: 0.003\n",
      "Pro_CA: 0.003\n",
      "NACHT: 0.003\n",
      "MNHE: 0.002\n",
      "DNA_alkylation: 0.002\n",
      "TIGR01128: 0.002\n",
      "Hist_deacetyl: 0.001\n",
      "Epimerase_2: 0.001\n",
      "Glyoxalase_2: -0.001\n",
      "DUF5915: -0.002\n",
      "TIGR03408: -0.002\n",
      "Tautomerase: -0.002\n",
      "TIGR00176: -0.003\n",
      "TarH: -0.004\n",
      "Fic_N: -0.005\n",
      "TSCPD: -0.005\n",
      "TIGR01371: -0.005\n",
      "SBP_bac_8: -0.006\n",
      "TIGR00643: -0.006\n",
      "DUF900: -0.007\n",
      "B3_4: -0.007\n",
      "DUF4386: -0.007\n",
      "4HBT: -0.009\n",
      "TIGR01197: -0.010\n",
      "AnmK: -0.010\n",
      "ResIII: -0.010\n",
      "PIN_3: -0.011\n",
      "TIGR00631: -0.011\n",
      "G3P_acyltransf: -0.012\n",
      "PUCC: -0.012\n",
      "DDE_Tnp_1_3: -0.012\n",
      "TetR_C_24: -0.013\n",
      "TehB: -0.013\n",
      "HDOD: -0.013\n",
      "Sbt_1: -0.013\n",
      "Peptidase_M55: -0.014\n",
      "Rieske: -0.014\n",
      "PhyH: -0.014\n",
      "peroxidase: -0.014\n",
      "GTP_EFTU: -0.014\n",
      "TIGR03916: -0.014\n",
      "Peptidase_C92: -0.015\n",
      "TIGR01350: -0.016\n",
      "TIGR00685: -0.016\n",
      "TIGR00710: -0.017\n",
      "TIGR00726: -0.017\n",
      "CrtC: -0.017\n",
      "DUF4113: -0.017\n",
      "TIGR01498: -0.017\n",
      "TIGR02246: -0.018\n",
      "LptE: -0.018\n",
      "Complex1_51K: -0.018\n",
      "SepSecS: -0.018\n",
      "CutA1: -0.019\n",
      "FMN_dh: -0.019\n",
      "EpsG: -0.019\n",
      "SBP_bac_6: -0.020\n",
      "TIGR01312: -0.020\n",
      "Glyco_hydro_8: -0.020\n",
      "Pyridoxal_deC: -0.020\n",
      "tRNA_synt_2f: -0.020\n",
      "TIGR00035: -0.020\n",
      "GTP_EFTU_D3: -0.021\n",
      "TIGR00250: -0.021\n",
      "Polyketide_cyc: -0.021\n",
      "Y1_Tnp: -0.021\n",
      "Ion_trans: -0.021\n",
      "TIGR01991: -0.022\n",
      "TIGR00457: -0.022\n",
      "Gly_radical: -0.022\n",
      "TIGR03956: -0.022\n",
      "Phos_pyr_kin: -0.022\n",
      "OTCace: -0.022\n",
      "AAA_19: -0.022\n",
      "MPTase-PolyVal: -0.022\n",
      "DUF58: -0.023\n",
      "DUF3592: -0.023\n",
      "TIGR00400: -0.023\n",
      "MobC: -0.023\n",
      "SWI2_SNF2: -0.023\n",
      "Topoisom_I: -0.024\n",
      "Protoglobin: -0.024\n",
      "TIGR00786: -0.024\n",
      "PDH_N: -0.024\n",
      "TIGR04380: -0.024\n",
      "Pyridox_ox_2: -0.024\n",
      "TIGR01104: -0.024\n",
      "TIGR02032: -0.024\n",
      "Sigma70_r4: -0.024\n",
      "Thi4: -0.024\n",
      "DUF386: -0.024\n",
      "HTH_DeoR: -0.024\n",
      "Transglut_core3: -0.025\n",
      "Cons_hypoth698: -0.025\n",
      "PRiA4_ORF3: -0.025\n",
      "Lactamase_B_3: -0.025\n",
      "TIGR01444: -0.025\n",
      "CsbD: -0.026\n",
      "DHDPS: -0.026\n",
      "TIGR01178: -0.026\n",
      "Glyco_hydro_65C: -0.026\n",
      "TIGR00476: -0.026\n",
      "Lycopene_cycl: -0.026\n",
      "TIGR01843: -0.026\n",
      "Dabb: -0.026\n",
      "Pirin: -0.027\n",
      "DUF3387: -0.027\n",
      "TIGR00182: -0.027\n",
      "TIGR00266: -0.027\n",
      "Alliinase_C: -0.027\n",
      "Ldh_2: -0.027\n",
      "TIGR00160: -0.027\n",
      "RimK: -0.028\n",
      "TIGR00675: -0.028\n",
      "TIGR01021: -0.028\n",
      "DUF3641: -0.028\n",
      "YibE_F: -0.028\n",
      "Nitroreductase: -0.028\n",
      "DUF5655: -0.028\n",
      "Sel1: -0.029\n",
      "TIGR00524: -0.029\n",
      "Endonuclease_NS: -0.029\n",
      "ABG_transport: -0.029\n",
      "PseudoU_synth_2: -0.030\n",
      "HU-HIG: -0.030\n",
      "RraA-like: -0.031\n",
      "TIGR01195: -0.031\n",
      "OsmC: -0.031\n",
      "Seryl_tRNA_N: -0.031\n",
      "Ni_hydr_CYTB: -0.031\n",
      "TIGR01728: -0.031\n",
      "PDH_C: -0.032\n",
      "YfhO: -0.032\n",
      "HPP: -0.033\n",
      "zf-TFIIB: -0.033\n",
      "TadE: -0.033\n",
      "YegS_C: -0.033\n",
      "Peptidase_C1_2: -0.033\n",
      "HHH_4: -0.033\n",
      "CCP_MauG: -0.033\n",
      "DUF1329: -0.033\n",
      "TIGR02384: -0.034\n",
      "ThiI: -0.034\n",
      "Desulfoferrodox: -0.034\n",
      "Cyclophil_like2: -0.034\n",
      "TIGR00129: -0.034\n",
      "Acyltransferase: -0.035\n",
      "TIGR00437: -0.035\n",
      "LRR_8: -0.036\n",
      "SIR2: -0.036\n",
      "malic: -0.036\n",
      "TIGR00456: -0.036\n",
      "HEAT_PBS: -0.037\n",
      "PEPcase: -0.037\n",
      "TOBE_2: -0.037\n",
      "NERD: -0.037\n",
      "DUF2229: -0.038\n",
      "HNH: -0.038\n",
      "PTR2: -0.039\n",
      "zinc_ribbon_2: -0.039\n",
      "DHFR_1: -0.039\n",
      "FecR: -0.039\n",
      "BFN_dom: -0.039\n",
      "DUF1285: -0.040\n",
      "AAA_15: -0.040\n",
      "BrnA_antitoxin: -0.040\n",
      "tRNA_NucTran2_2: -0.041\n",
      "TPR_2: -0.041\n",
      "TIGR01352: -0.041\n",
      "TIGR00427: -0.041\n",
      "OCD_Mu_crystall: -0.041\n",
      "PS_pyruv_trans: -0.042\n",
      "TIGR02281: -0.042\n",
      "PqqD: -0.042\n",
      "PHP_C: -0.042\n",
      "TIGR01383: -0.042\n",
      "SNARE_assoc: -0.043\n",
      "TIGR01575: -0.043\n",
      "TIGR02317: -0.043\n",
      "DUF1611: -0.043\n",
      "Spore_YtfJ: -0.043\n",
      "YjdM: -0.043\n",
      "Tex_N: -0.043\n",
      "VKOR: -0.044\n",
      "RskA: -0.044\n",
      "PEGA: -0.044\n",
      "DNA_mis_repair: -0.044\n",
      "HipA_C: -0.044\n",
      "TIGR00127: -0.044\n",
      "Fe-ADH: -0.045\n",
      "EIID-AGA: -0.045\n",
      "TIGR01901: -0.045\n",
      "TIGR00099: -0.045\n",
      "PaaX_C: -0.045\n",
      "TPR_14: -0.045\n",
      "TIGR01493: -0.045\n",
      "TIGR00056: -0.046\n",
      "RelB: -0.046\n",
      "TIGR00332: -0.046\n",
      "TruD: -0.046\n",
      "TPR_7: -0.046\n",
      "NadA: -0.047\n",
      "PmbA_TldD_C: -0.047\n",
      "Peptidase_S10: -0.047\n",
      "Peptidase_S9: -0.047\n",
      "TIGR01428: -0.047\n",
      "Sua5_C: -0.047\n",
      "CaMKII_AD: -0.047\n",
      "TIGR00514: -0.047\n",
      "DNA_pol_B_exo1: -0.048\n",
      "TIGR02673: -0.048\n",
      "Peptidase_C39_2: -0.048\n",
      "VanW: -0.049\n",
      "FKBP_C: -0.049\n",
      "Glyco_tran_WbsX: -0.049\n",
      "Arc: -0.050\n",
      "Sigma70_r2: -0.050\n",
      "DUF4140: -0.050\n",
      "FtsH_ext: -0.050\n",
      "CitMHS_2: -0.051\n",
      "TIGR01123: -0.051\n",
      "OAD_gamma: -0.051\n",
      "DHquinase_I: -0.052\n",
      "PhoU_div: -0.052\n",
      "TIGR01879: -0.052\n",
      "HrcA_DNA-bdg: -0.052\n",
      "Acetyltransf_6: -0.052\n",
      "EutN_CcmL: -0.053\n",
      "HlyD_2: -0.053\n",
      "FTCD_C: -0.055\n",
      "TIGR01344: -0.055\n",
      "XFP_C: -0.055\n",
      "DUF2721: -0.055\n",
      "Oxidored_molyb: -0.056\n",
      "PRAI: -0.056\n",
      "DUF6088: -0.056\n",
      "Glyco_hydro_43: -0.056\n",
      "DUF465: -0.056\n",
      "TIGR00372: -0.056\n",
      "Rad51: -0.056\n",
      "Na_Ala_symp: -0.056\n",
      "DUF1684: -0.056\n",
      "dCMP_cyt_deam_1: -0.056\n",
      "Smr: -0.057\n",
      "DAC: -0.057\n",
      "FRG: -0.057\n",
      "DrsE: -0.057\n",
      "Hydantoinase_B: -0.057\n",
      "Guanylate_cyc: -0.057\n",
      "N6_N4_Mtase: -0.058\n",
      "Helicase_C: -0.058\n",
      "TIGR02817: -0.058\n",
      "LolA_like: -0.058\n",
      "TgpA_N: -0.059\n",
      "Fasciclin: -0.059\n",
      "DUF2071: -0.059\n",
      "TIGR03824: -0.059\n",
      "DUF456: -0.059\n",
      "OTCace_N: -0.059\n",
      "FCD: -0.059\n",
      "DUF934: -0.059\n",
      "UPF0060: -0.059\n",
      "bact-PGI_C: -0.060\n",
      "Virulence_RhuM: -0.060\n",
      "NiFeSe_Hases: -0.060\n",
      "TIGR00420: -0.060\n",
      "Pmp3: -0.060\n",
      "DUF411: -0.060\n",
      "DDAH_eukar: -0.060\n",
      "Ketoacyl-synt_2: -0.060\n",
      "PriA_CRR: -0.061\n",
      "TIGR00614: -0.061\n",
      "Virul_fac_BrkB: -0.061\n",
      "FR47: -0.062\n",
      "3HCDH_N: -0.062\n",
      "Lactamase_B_6: -0.062\n",
      "DUF815: -0.062\n",
      "TIGR00551: -0.062\n",
      "TIGR02891: -0.062\n",
      "PIN_4: -0.063\n",
      "TIGR00073: -0.063\n",
      "Bac_DnaA: -0.063\n",
      "TAL_FSA: -0.063\n",
      "DAHP_synth_2: -0.063\n",
      "PhoD_N: -0.064\n",
      "TIGR01744: -0.064\n",
      "TPR_6: -0.064\n",
      "TIGR02515: -0.064\n",
      "YaeQ: -0.064\n",
      "PLDc_2: -0.064\n",
      "FliH: -0.064\n",
      "Arginase: -0.064\n",
      "FleQ: -0.064\n",
      "TIGR00444: -0.065\n",
      "TIGR00068: -0.065\n",
      "Lon_C: -0.066\n",
      "TIGR00416: -0.066\n",
      "TIGR00382: -0.066\n",
      "Glyco_hydro_114: -0.066\n",
      "RecQ_Zn_bind: -0.066\n",
      "Ribonuc_red_lgN: -0.066\n",
      "TIGR00564: -0.067\n",
      "TIGR01311: -0.067\n",
      "TIGR03409: -0.067\n",
      "TspO_MBR: -0.067\n",
      "DUF4154: -0.067\n",
      "TIGR02210: -0.067\n",
      "HicB: -0.068\n",
      "TIGR00159: -0.068\n",
      "Cauli_VI: -0.068\n",
      "TIGR00735: -0.068\n",
      "Lipase_GDSL_2: -0.068\n",
      "DUF368: -0.069\n",
      "AAA_14: -0.069\n",
      "EVE: -0.069\n",
      "DUF2062: -0.070\n",
      "CoA_binding_3: -0.070\n",
      "TIGR01737: -0.070\n",
      "Rib_5-P_isom_A: -0.070\n",
      "HTH_29: -0.071\n",
      "GAF: -0.071\n",
      "DUF5981: -0.071\n",
      "POTRA_2: -0.071\n",
      "zf-C4_Topoisom: -0.072\n",
      "TIGR02870: -0.072\n",
      "TIGR00299: -0.073\n",
      "TPR_9: -0.074\n",
      "MlaC: -0.074\n",
      "TIGR02039: -0.075\n",
      "Bac_globin: -0.075\n",
      "TIGR00442: -0.075\n",
      "TIGR01212: -0.075\n",
      "TIGR03356: -0.076\n",
      "Cupin_6: -0.076\n",
      "o_Sphingomonadales: -0.076\n",
      "Hpr_kinase_N: -0.076\n",
      "SnoaL_3: -0.076\n",
      "TIGR01941: -0.076\n",
      "Adenylsucc_synt: -0.076\n",
      "DUF1501: -0.077\n",
      "FSH1: -0.077\n",
      "Fn3_assoc: -0.077\n",
      "TIGR00987: -0.078\n",
      "dNK: -0.078\n",
      "LpoB: -0.078\n",
      "YbbR: -0.078\n",
      "TIGR00234: -0.078\n",
      "CLP_protease: -0.078\n",
      "TIGR00460: -0.079\n",
      "TIGR01070: -0.079\n",
      "OmpW: -0.079\n",
      "MutL_C: -0.079\n",
      "TIGR01225: -0.079\n",
      "HTH_12: -0.080\n",
      "Rubredoxin_2: -0.080\n",
      "FliD_N: -0.080\n",
      "G5: -0.080\n",
      "Acid_phosphat_B: -0.081\n",
      "TIGR01662: -0.081\n",
      "TIGR01462: -0.081\n",
      "DUF1328: -0.081\n",
      "Thioredoxin_4: -0.082\n",
      "AAA_28: -0.082\n",
      "TIGR01682: -0.083\n",
      "TIGR03936: -0.083\n",
      "Glyco_tranf_2_2: -0.083\n",
      "TIGR01722: -0.084\n",
      "ASL_C: -0.084\n",
      "HHH_8: -0.084\n",
      "Acatn: -0.085\n",
      "TIGR01347: -0.085\n",
      "HGSNAT_cat: -0.085\n",
      "Dala_Dala_lig_N: -0.085\n",
      "TrmO_C: -0.085\n",
      "Y_phosphatase: -0.086\n",
      "DNA_pol_A_exo1: -0.086\n",
      "TIGR02022: -0.086\n",
      "TIGR00406: -0.086\n",
      "HATPase_c_5: -0.087\n",
      "TIGR00720: -0.087\n",
      "TruB-C_2: -0.088\n",
      "TIGR00741: -0.089\n",
      "TIGR00658: -0.089\n",
      "CstA: -0.089\n",
      "Glyco_hydro_42C: -0.089\n",
      "TIGR00763: -0.089\n",
      "BATS: -0.090\n",
      "DLH: -0.090\n",
      "TIGR01414: -0.090\n",
      "SnoaL_2: -0.091\n",
      "TIGR00711: -0.091\n",
      "EutA: -0.091\n",
      "Peptidase_M3: -0.091\n",
      "RimM: -0.091\n",
      "PNK3P: -0.091\n",
      "TIGR00875: -0.093\n",
      "Bac_DnaA_C: -0.093\n",
      "ketoacyl-synt: -0.093\n",
      "TIGR03023: -0.094\n",
      "zf-CGNR: -0.094\n",
      "Lum_binding: -0.095\n",
      "LptF_LptG: -0.095\n",
      "HATPase_c_4: -0.095\n",
      "TIGR01534: -0.097\n",
      "TIGR00652: -0.097\n",
      "STAS: -0.097\n",
      "DUF3458_C: -0.097\n",
      "TIGR00054: -0.097\n",
      "Peptidase_M28: -0.097\n",
      "Sulf_transp: -0.097\n",
      "SbmA_BacA: -0.098\n",
      "TIGR01579: -0.098\n",
      "TIGR01354: -0.098\n",
      "tRNA-synt_1c_C: -0.098\n",
      "DUF3800: -0.099\n",
      "AMP-binding: -0.100\n",
      "Dala_Dala_lig_C: -0.100\n",
      "HGD-D: -0.100\n",
      "Glyco_hydro_38N: -0.100\n",
      "TIGR04265: -0.101\n",
      "TIGR00231: -0.101\n",
      "TK: -0.102\n",
      "TIGR01472: -0.102\n",
      "YitT_membrane: -0.102\n",
      "HWE_HK: -0.102\n",
      "Asp_decarbox: -0.102\n",
      "TctC: -0.103\n",
      "TIGR01910: -0.103\n",
      "ACT_7: -0.103\n",
      "SprT-like: -0.103\n",
      "TIGR00486: -0.104\n",
      "TMEM175: -0.104\n",
      "TIGR01617: -0.105\n",
      "MltA: -0.105\n",
      "DUF4015: -0.105\n",
      "TIGR00010: -0.105\n",
      "Aminotran_1_2: -0.105\n",
      "TIGR01882: -0.106\n",
      "TIGR00361: -0.107\n",
      "ATP-grasp_5: -0.107\n",
      "TIGR00965: -0.107\n",
      "PQQ_3: -0.107\n",
      "TIGR02945: -0.108\n",
      "NtrY_N: -0.110\n",
      "Mannitol_dh_C: -0.110\n",
      "TIGR02669: -0.110\n",
      "Rhodanese: -0.111\n",
      "zf-ACC: -0.111\n",
      "TIGR02666: -0.112\n",
      "CopG_antitoxin: -0.112\n",
      "YbjQ_1: -0.113\n",
      "Acyl-ACP_TE: -0.114\n",
      "MCM: -0.114\n",
      "CTP_synth_N: -0.114\n",
      "DeoC: -0.115\n",
      "DUF4115: -0.115\n",
      "HA2: -0.115\n",
      "DbpA: -0.116\n",
      "TIGR03464: -0.116\n",
      "TIGR02822: -0.117\n",
      "HlyD_D23: -0.118\n",
      "PyrI_C: -0.118\n",
      "Anticodon_3: -0.119\n",
      "Thymidylate_kin: -0.120\n",
      "DUF3459: -0.120\n",
      "AA_permease_C: -0.120\n",
      "Gluconate_2-dh3: -0.120\n",
      "DnaA_N: -0.120\n",
      "IPT: -0.121\n",
      "FAD-oxidase_C: -0.123\n",
      "Glyco_hydro_42: -0.123\n",
      "DUF2867: -0.123\n",
      "FlgM: -0.123\n",
      "TIGR01036: -0.124\n",
      "Ribosomal_L17: -0.124\n",
      "GATase_3: -0.124\n",
      "Phe_tRNA-synt_N: -0.125\n",
      "TIGR01069: -0.126\n",
      "TIGR00131: -0.126\n",
      "TIGR00209: -0.126\n",
      "TIGR01051: -0.127\n",
      "TIGR00383: -0.127\n",
      "NAD_kinase: -0.128\n",
      "SpoIIM: -0.128\n",
      "TIGR00797: -0.128\n",
      "AAA_18: -0.128\n",
      "AMP_N: -0.128\n",
      "Pyrophosphatase: -0.128\n",
      "TIGR02170: -0.129\n",
      "TIGR00978: -0.129\n",
      "TIGR03330: -0.129\n",
      "Peptidase_M13: -0.130\n",
      "DUF3417: -0.130\n",
      "FeS: -0.130\n",
      "DUF502: -0.131\n",
      "TIGR03263: -0.133\n",
      "TIGR02683: -0.133\n",
      "TIGR00312: -0.134\n",
      "Lactamase_B_2: -0.134\n",
      "Glycos_transf_2: -0.134\n",
      "TIGR00030: -0.134\n",
      "PRK: -0.134\n",
      "TIGR03167: -0.135\n",
      "DcpS_C: -0.136\n",
      "TIGR00755: -0.136\n",
      "TIGR00928: -0.137\n",
      "MASE1: -0.140\n",
      "TIGR00013: -0.141\n",
      "Sacchrp_dh_C: -0.141\n",
      "Bactofilin: -0.142\n",
      "DUF1178: -0.143\n",
      "Lipoprotein_8: -0.146\n",
      "MutS_II: -0.147\n",
      "TIGR00170: -0.149\n",
      "His_Phos_1: -0.150\n",
      "DNA_pol3_delt_C: -0.150\n",
      "FliS: -0.151\n",
      "QueH: -0.154\n",
      "PadR: -0.155\n",
      "DEAD: -0.159\n",
      "TIGR01892: -0.160\n",
      "PucR: -0.161\n",
      "CYTH: -0.163\n",
      "FemAB: -0.163\n",
      "FolB: -0.167\n",
      "CarbopepD_reg_2: -0.167\n",
      "Glyco_hydro_20b: -0.168\n",
      "GlnE: -0.172\n",
      "ATPgrasp_ST: -0.172\n",
      "NFACT_N: -0.175\n",
      "S4_2: -0.176\n",
      "TrkA_C: -0.177\n",
      "TIGR02222: -0.181\n",
      "TIGR01324: -0.184\n",
      "Exonuc_V_gamma: -0.186\n",
      "TIGR00197: -0.197\n",
      "TIGR00344: -0.197\n",
      "Ppnp: -0.198\n",
      "TIGR00878: -0.201\n",
      "TIGR00997: -0.206\n",
      "TIGR01695: -0.209\n",
      "TIGR01191: -0.209\n",
      "Se-cys_synth_N: -0.213\n",
      "Cu_amine_oxidN1: -0.220\n",
      "Porin_O_P: -0.227\n",
      "TIGR00768: -0.233\n",
      "Thioredoxin_7: -0.233\n",
      "OmpH: -0.249\n",
      "HemX: -0.260\n",
      "LysM: -0.261\n",
      "CNNM: -0.262\n",
      "Phage_integrase: -0.273\n",
      "PrmA: -0.304\n",
      "Peptidase_S49: -0.322\n",
      "TIGR02127: -0.340\n",
      "TIGR02729: -0.382\n"
     ]
    }
   ],
   "source": [
    "coefficients = model.params\n",
    "sorted_coeff = sorted(coefficients.items(), reverse=True, key=lambda t: t[1])\n",
    "for i, (var, coef) in enumerate(sorted_coeff):\n",
    "    if coef > 0 or coef < 0:\n",
    "        print(f\"{var}: {coef:.3f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pseudomonadota</th>\n",
       "      <th>Gammaproteobacteria</th>\n",
       "      <th>Burkholderiales</th>\n",
       "      <td>1695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillota_A</th>\n",
       "      <th>Clostridia</th>\n",
       "      <th>Lachnospirales</th>\n",
       "      <td>1257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidota</th>\n",
       "      <th>Bacteroidia</th>\n",
       "      <th>Bacteroidales</th>\n",
       "      <td>1226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillota_A</th>\n",
       "      <th>Clostridia</th>\n",
       "      <th>Oscillospirales</th>\n",
       "      <td>1217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidota</th>\n",
       "      <th>Bacteroidia</th>\n",
       "      <th>Flavobacteriales</th>\n",
       "      <td>1047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonadota</th>\n",
       "      <th>Alphaproteobacteria</th>\n",
       "      <th>Rhizobiales</th>\n",
       "      <td>1022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Actinomycetota</th>\n",
       "      <th>Actinomycetia</th>\n",
       "      <th>Actinomycetales</th>\n",
       "      <td>986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Pseudomonadota</th>\n",
       "      <th>Alphaproteobacteria</th>\n",
       "      <th>Rhodobacterales</th>\n",
       "      <td>866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gammaproteobacteria</th>\n",
       "      <th>Pseudomonadales</th>\n",
       "      <td>731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Enterobacterales_A</th>\n",
       "      <td>574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidota</th>\n",
       "      <th>Bacteroidia</th>\n",
       "      <th>Chitinophagales</th>\n",
       "      <td>536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillota_A</th>\n",
       "      <th>Clostridia</th>\n",
       "      <th>Christensenellales</th>\n",
       "      <td>507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonadota</th>\n",
       "      <th>Gammaproteobacteria</th>\n",
       "      <th>Enterobacterales</th>\n",
       "      <td>495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Bacillota</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Bacilli</th>\n",
       "      <th>Lactobacillales</th>\n",
       "      <td>479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillales</th>\n",
       "      <td>463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Actinomycetota</th>\n",
       "      <th>Actinomycetia</th>\n",
       "      <th>Mycobacteriales</th>\n",
       "      <td>461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidota</th>\n",
       "      <th>Bacteroidia</th>\n",
       "      <th>Cytophagales</th>\n",
       "      <td>391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cyanobacteriota</th>\n",
       "      <th>Cyanobacteriia</th>\n",
       "      <th>Cyanobacteriales</th>\n",
       "      <td>367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Actinomycetota</th>\n",
       "      <th>Acidimicrobiia</th>\n",
       "      <th>Acidimicrobiales</th>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pseudomonadota</th>\n",
       "      <th>Alphaproteobacteria</th>\n",
       "      <th>Sphingomonadales</th>\n",
       "      <td>351</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        assembly_accession\n",
       "gtdb_phylum     gtdb_class          gtdb_order                            \n",
       "Pseudomonadota  Gammaproteobacteria Burkholderiales                   1695\n",
       "Bacillota_A     Clostridia          Lachnospirales                    1257\n",
       "Bacteroidota    Bacteroidia         Bacteroidales                     1226\n",
       "Bacillota_A     Clostridia          Oscillospirales                   1217\n",
       "Bacteroidota    Bacteroidia         Flavobacteriales                  1047\n",
       "Pseudomonadota  Alphaproteobacteria Rhizobiales                       1022\n",
       "Actinomycetota  Actinomycetia       Actinomycetales                    986\n",
       "Pseudomonadota  Alphaproteobacteria Rhodobacterales                    866\n",
       "                Gammaproteobacteria Pseudomonadales                    731\n",
       "                                    Enterobacterales_A                 574\n",
       "Bacteroidota    Bacteroidia         Chitinophagales                    536\n",
       "Bacillota_A     Clostridia          Christensenellales                 507\n",
       "Pseudomonadota  Gammaproteobacteria Enterobacterales                   495\n",
       "Bacillota       Bacilli             Lactobacillales                    479\n",
       "                                    Bacillales                         463\n",
       "Actinomycetota  Actinomycetia       Mycobacteriales                    461\n",
       "Bacteroidota    Bacteroidia         Cytophagales                       391\n",
       "Cyanobacteriota Cyanobacteriia      Cyanobacteriales                   367\n",
       "Actinomycetota  Acidimicrobiia      Acidimicrobiales                   353\n",
       "Pseudomonadota  Alphaproteobacteria Sphingomonadales                   351"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acc = TIGR_df[TIGR_df['TIGR02729'] == 1].index\n",
    "gtdb_metadata.loc[acc].reset_index()[['gtdb_phylum', 'gtdb_class', 'gtdb_order', 'assembly_accession']].groupby([\n",
    "    'gtdb_phylum', 'gtdb_class', 'gtdb_order'\n",
    "]).count().sort_values('assembly_accession', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cwb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
