{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Streptococcus S protein\n",
    "\n",
    "OG: 4HHT0\n",
    "\n",
    "S proteins are LysM-containing proteins playing a central role in pathogenicity:\n",
    "- immune evasion in Group A Streptococci (camouflaging with red blood cell debris) [1]\n",
    "- resistance to beta-lactam antibiotics through recruitment of protein PGP1a [2]\n",
    "- resistance to lysozyme through recruitment of protein PgdA. [2]\n",
    "  - PgdA is a peptidoglycan N-acetylglucosamine deacetylase, i.e. it removes the N-acetyl group from the middle sugar of the peptidoglycan.\n",
    "  - It helps the bacteria resist lysozyme [3].\n",
    "\n",
    "Mutants of _S. pneumoniae_ without S protein show reduce virulence in mouse models, are more suceptible to beta-lactam antibiotics and lysozyme. [2]\n",
    "\n",
    "Refs: \n",
    "- [1] [Wierzbicki et al., 2019](https://doi.org/10.1016/j.celrep.2019.11.001)\n",
    "- [2] [Burnier et al., 2024](https://doi.org/10.1101/2024.11.08.622053)\n",
    "- [3] [Bui et al., 2011](https://doi.org/10.1016/j.bcp.2011.03.028)\n",
    "\n",
    "We found that S proteins are absent from all complete genomes of _S. equi_.\n",
    "\n",
    "_S. equi_ is primarily a horse pathogen but _S. equi subsp. zooepidemicus_ has zoonotic potential: it also infects a wide range of mammals, including dogs, cats, ruminants, pigs, and humans."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as plt\n",
    "import seaborn as sns\n",
    "from Bio import Phylo\n",
    "\n",
    "cwd = os.getcwd()\n",
    "if cwd.endswith('notebook'):\n",
    "    os.chdir('..')\n",
    "    cwd = os.getcwd()\n",
    "\n",
    "from src.tree.tree_util import prune_leaves_with_unknown_id\n",
    "from src.tree.itol_annotation import itol_labels, itol_colored_ranges, itol_binary_annotations, hex_to_rgba"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme(palette='colorblind', font_scale=1.3)\n",
    "palette_colorblind = sns.color_palette('colorblind').as_hex()\n",
    "palette_pastel = sns.color_palette('pastel').as_hex()\n",
    "\n",
    "data_folder = Path('./data/')\n",
    "assert data_folder.is_dir()\n",
    "\n",
    "db_proka = Path('../db_proka/')\n",
    "assert db_proka.is_dir()\n",
    "\n",
    "gtdb_folder = Path('../data/gtdb_r220/')\n",
    "assert gtdb_folder.is_dir()\n",
    "\n",
    "strep_folder = gtdb_folder / 'Streptococcus'\n",
    "assert strep_folder.is_dir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>ambiguous_bases</th>\n",
       "      <th>checkm2_completeness</th>\n",
       "      <th>checkm2_contamination</th>\n",
       "      <th>checkm2_model</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>checkm_marker_count</th>\n",
       "      <th>checkm_marker_lineage</th>\n",
       "      <th>checkm_marker_set_count</th>\n",
       "      <th>...</th>\n",
       "      <th>trna_aa_count</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>trna_selenocysteine_count</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_900636555.1</th>\n",
       "      <td>RS_GCF_900636555.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.14</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>475</td>\n",
       "      <td>o__Lactobacillales (UID544)</td>\n",
       "      <td>267</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus gordonii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_015908985.1</th>\n",
       "      <td>RS_GCF_015908985.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.23</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>475</td>\n",
       "      <td>o__Lactobacillales (UID544)</td>\n",
       "      <td>267</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus suis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001266635.1</th>\n",
       "      <td>RS_GCF_001266635.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.06</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.82</td>\n",
       "      <td>0.00</td>\n",
       "      <td>524</td>\n",
       "      <td>f__Streptococcaceae (UID545)</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_004154025.1</th>\n",
       "      <td>RS_GCF_004154025.1</td>\n",
       "      <td>0</td>\n",
       "      <td>99.99</td>\n",
       "      <td>0.13</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.85</td>\n",
       "      <td>0.00</td>\n",
       "      <td>676</td>\n",
       "      <td>g__Streptococcus (UID722)</td>\n",
       "      <td>182</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_029011635.1</th>\n",
       "      <td>RS_GCF_029011635.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.20</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.18</td>\n",
       "      <td>524</td>\n",
       "      <td>f__Streptococcaceae (UID545)</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             accession  ambiguous_bases  checkm2_completeness  \\\n",
       "assembly_accession                                                              \n",
       "GCF_900636555.1     RS_GCF_900636555.1                0                100.00   \n",
       "GCF_015908985.1     RS_GCF_015908985.1                0                100.00   \n",
       "GCF_001266635.1     RS_GCF_001266635.1                0                100.00   \n",
       "GCF_004154025.1     RS_GCF_004154025.1                0                 99.99   \n",
       "GCF_029011635.1     RS_GCF_029011635.1                0                100.00   \n",
       "\n",
       "                    checkm2_contamination checkm2_model  checkm_completeness  \\\n",
       "assembly_accession                                                             \n",
       "GCF_900636555.1                      0.14      Specific               100.00   \n",
       "GCF_015908985.1                      0.23      Specific               100.00   \n",
       "GCF_001266635.1                      0.06      Specific                99.82   \n",
       "GCF_004154025.1                      0.13      Specific                99.85   \n",
       "GCF_029011635.1                      0.20      Specific               100.00   \n",
       "\n",
       "                    checkm_contamination  checkm_marker_count  \\\n",
       "assembly_accession                                              \n",
       "GCF_900636555.1                     0.00                  475   \n",
       "GCF_015908985.1                     0.00                  475   \n",
       "GCF_001266635.1                     0.00                  524   \n",
       "GCF_004154025.1                     0.00                  676   \n",
       "GCF_029011635.1                     0.18                  524   \n",
       "\n",
       "                           checkm_marker_lineage  checkm_marker_set_count  \\\n",
       "assembly_accession                                                          \n",
       "GCF_900636555.1      o__Lactobacillales (UID544)                      267   \n",
       "GCF_015908985.1      o__Lactobacillales (UID544)                      267   \n",
       "GCF_001266635.1     f__Streptococcaceae (UID545)                      282   \n",
       "GCF_004154025.1        g__Streptococcus (UID722)                      182   \n",
       "GCF_029011635.1     f__Streptococcaceae (UID545)                      282   \n",
       "\n",
       "                    ...  trna_aa_count  trna_count  trna_selenocysteine_count  \\\n",
       "assembly_accession  ...                                                         \n",
       "GCF_900636555.1     ...             19          59                          0   \n",
       "GCF_015908985.1     ...             19          56                          0   \n",
       "GCF_001266635.1     ...             18          63                          0   \n",
       "GCF_004154025.1     ...             19          67                          0   \n",
       "GCF_029011635.1     ...             19          80                          0   \n",
       "\n",
       "                      domain  gtdb_phylum  gtdb_class       gtdb_order  \\\n",
       "assembly_accession                                                       \n",
       "GCF_900636555.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_015908985.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_001266635.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_004154025.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_029011635.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "\n",
       "                         gtdb_family     gtdb_genus              gtdb_species  \n",
       "assembly_accession                                                             \n",
       "GCF_900636555.1     Streptococcaceae  Streptococcus    Streptococcus gordonii  \n",
       "GCF_015908985.1     Streptococcaceae  Streptococcus        Streptococcus suis  \n",
       "GCF_001266635.1     Streptococcaceae  Streptococcus  Streptococcus agalactiae  \n",
       "GCF_004154025.1     Streptococcaceae  Streptococcus    Streptococcus pyogenes  \n",
       "GCF_029011635.1     Streptococcaceae  Streptococcus  Streptococcus agalactiae  \n",
       "\n",
       "[5 rows x 120 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df = pd.read_csv(strep_folder / 'genomes_metadata.csv', index_col='assembly_accession')\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfam_df = pd.read_csv(strep_folder / 'Streptococcus_all_proteins.pfam.csv', index_col='assembly_accession')\n",
    "pfam_df['gtdb_species'] = [metadata_df.loc[a, 'gtdb_species'] for a in pfam_df.index]\n",
    "\n",
    "tigr_df = pd.read_csv(strep_folder / 'Streptococcus_all_proteins.tigr.csv', index_col='assembly_accession')\n",
    "tigr_df['gtdb_species'] = [metadata_df.loc[a, 'gtdb_species'] for a in tigr_df.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467658.1@GCF_003963555.1</td>\n",
       "      <td>1000570.HMPREF9966_1759</td>\n",
       "      <td>9.704000e-82</td>\n",
       "      <td>271.0</td>\n",
       "      <td>COG0716@1|root,COG0716@2|Bacteria,1V45R@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>C</td>\n",
       "      <td>Flavodoxin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Flavodoxin_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467724.1@GCF_003963555.1</td>\n",
       "      <td>862970.SAIN_1577</td>\n",
       "      <td>1.313000e-186</td>\n",
       "      <td>582.0</td>\n",
       "      <td>COG1396@1|root,COG1396@2|Bacteria,1VIH9@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>K</td>\n",
       "      <td>Helix-turn-helix XRE-family like proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HTH_19,HTH_3,TPR_12,TPR_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467790.1@GCF_003963555.1</td>\n",
       "      <td>862970.SAIN_1636</td>\n",
       "      <td>2.776000e-131</td>\n",
       "      <td>418.0</td>\n",
       "      <td>COG1564@1|root,COG1564@2|Bacteria,1VA0W@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>H</td>\n",
       "      <td>Thiamin pyrophosphokinase, vitamin B1 binding ...</td>\n",
       "      <td>thiN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K00949</td>\n",
       "      <td>ko00730,ko01100,map00730,map01100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R00619</td>\n",
       "      <td>RC00002,RC00017</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TPK_B1_binding,TPK_catalytic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467857.1@GCF_003963555.1</td>\n",
       "      <td>176090.SSIN_0693</td>\n",
       "      <td>1.181000e-201</td>\n",
       "      <td>627.0</td>\n",
       "      <td>COG3677@1|root,COG3677@2|Bacteria,1V4D1@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>L</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDE_Tnp_IS1595,Zn_Tnp_IS1595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467993.1@GCF_003963555.1</td>\n",
       "      <td>862969.SCI_1925</td>\n",
       "      <td>1.812000e-256</td>\n",
       "      <td>791.0</td>\n",
       "      <td>COG0612@1|root,COG0612@2|Bacteria,1TPN6@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>Peptidase M16 inactive</td>\n",
       "      <td>ymfF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peptidase_M16,Peptidase_M16_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        protein_id            seed_ortholog  \\\n",
       "assembly_accession                                                            \n",
       "GCF_003963555.1     WP_126467658.1@GCF_003963555.1  1000570.HMPREF9966_1759   \n",
       "GCF_003963555.1     WP_126467724.1@GCF_003963555.1         862970.SAIN_1577   \n",
       "GCF_003963555.1     WP_126467790.1@GCF_003963555.1         862970.SAIN_1636   \n",
       "GCF_003963555.1     WP_126467857.1@GCF_003963555.1         176090.SSIN_0693   \n",
       "GCF_003963555.1     WP_126467993.1@GCF_003963555.1          862969.SCI_1925   \n",
       "\n",
       "                           evalue  score  \\\n",
       "assembly_accession                         \n",
       "GCF_003963555.1      9.704000e-82  271.0   \n",
       "GCF_003963555.1     1.313000e-186  582.0   \n",
       "GCF_003963555.1     2.776000e-131  418.0   \n",
       "GCF_003963555.1     1.181000e-201  627.0   \n",
       "GCF_003963555.1     1.812000e-256  791.0   \n",
       "\n",
       "                                                           eggNOG_OGs  \\\n",
       "assembly_accession                                                      \n",
       "GCF_003963555.1     COG0716@1|root,COG0716@2|Bacteria,1V45R@1239|F...   \n",
       "GCF_003963555.1     COG1396@1|root,COG1396@2|Bacteria,1VIH9@1239|F...   \n",
       "GCF_003963555.1     COG1564@1|root,COG1564@2|Bacteria,1VA0W@1239|F...   \n",
       "GCF_003963555.1     COG3677@1|root,COG3677@2|Bacteria,1V4D1@1239|F...   \n",
       "GCF_003963555.1     COG0612@1|root,COG0612@2|Bacteria,1TPN6@1239|F...   \n",
       "\n",
       "                   max_annot_lvl COG_category  \\\n",
       "assembly_accession                              \n",
       "GCF_003963555.1       2|Bacteria            C   \n",
       "GCF_003963555.1       2|Bacteria            K   \n",
       "GCF_003963555.1       2|Bacteria            H   \n",
       "GCF_003963555.1       2|Bacteria            L   \n",
       "GCF_003963555.1       2|Bacteria            S   \n",
       "\n",
       "                                                          Description  \\\n",
       "assembly_accession                                                      \n",
       "GCF_003963555.1                                            Flavodoxin   \n",
       "GCF_003963555.1             Helix-turn-helix XRE-family like proteins   \n",
       "GCF_003963555.1     Thiamin pyrophosphokinase, vitamin B1 binding ...   \n",
       "GCF_003963555.1                         ISXO2-like transposase domain   \n",
       "GCF_003963555.1                                Peptidase M16 inactive   \n",
       "\n",
       "                   Preferred_name  GOs  ...    KEGG_ko  \\\n",
       "assembly_accession                      ...              \n",
       "GCF_003963555.1               NaN  NaN  ...        NaN   \n",
       "GCF_003963555.1               NaN  NaN  ...        NaN   \n",
       "GCF_003963555.1              thiN  NaN  ...  ko:K00949   \n",
       "GCF_003963555.1               NaN  NaN  ...        NaN   \n",
       "GCF_003963555.1              ymfF  NaN  ...        NaN   \n",
       "\n",
       "                                         KEGG_Pathway KEGG_Module  \\\n",
       "assembly_accession                                                  \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "GCF_003963555.1     ko00730,ko01100,map00730,map01100         NaN   \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "\n",
       "                   KEGG_Reaction      KEGG_rclass                    BRITE  \\\n",
       "assembly_accession                                                           \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "GCF_003963555.1           R00619  RC00002,RC00017  ko00000,ko00001,ko01000   \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "\n",
       "                   KEGG_TC CAZy BiGG_Reaction                          PFAMs  \n",
       "assembly_accession                                                            \n",
       "GCF_003963555.1        NaN  NaN           NaN                   Flavodoxin_4  \n",
       "GCF_003963555.1        NaN  NaN           NaN      HTH_19,HTH_3,TPR_12,TPR_8  \n",
       "GCF_003963555.1        NaN  NaN           NaN   TPK_B1_binding,TPK_catalytic  \n",
       "GCF_003963555.1        NaN  NaN           NaN   DDE_Tnp_IS1595,Zn_Tnp_IS1595  \n",
       "GCF_003963555.1        NaN  NaN           NaN  Peptidase_M16,Peptidase_M16_C  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggNOG_annotations_path = strep_folder / 'Streptococcus_eggNOG_annotations.csv'\n",
    "eggNOG_df = pd.read_csv(eggNOG_annotations_path, index_col='assembly_accession')\n",
    "eggNOG_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_ogs(eggNOG_df, accessions, og_whitelist=None):\n",
    "    index = eggNOG_df.index\n",
    "\n",
    "    og_union = set()\n",
    "    for acc in accessions:\n",
    "        if acc not in index:\n",
    "            print(f'Not in index: {acc}')\n",
    "            continue\n",
    "\n",
    "        genme_ogs = eggNOG_df.loc[[acc]].set_index('protein_id')\n",
    "\n",
    "        ogs = set()\n",
    "        for protein_id in genme_ogs.index:\n",
    "            protein_ogs = genme_ogs.loc[protein_id, 'eggNOG_OGs']\n",
    "\n",
    "            candidate_ogs = {\n",
    "                og_with_tax.split('@')[0]\n",
    "                for og_with_tax in protein_ogs.split(',')\n",
    "            }\n",
    "\n",
    "            if og_whitelist is not None:\n",
    "                candidate_ogs = candidate_ogs & og_whitelist\n",
    "                \n",
    "            ogs |= candidate_ogs\n",
    "\n",
    "        og_union = og_union.union(ogs)\n",
    "\n",
    "    return og_union"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ogs_present_in_all(eggNOG_df, accessions, og_whitelist=None):\n",
    "    index = eggNOG_df.index\n",
    "\n",
    "    og_intersection = None\n",
    "    for acc in accessions:\n",
    "        if acc not in index:\n",
    "            print(f'Not in index: {acc}')\n",
    "            continue\n",
    "\n",
    "        genme_ogs = eggNOG_df.loc[[acc]].set_index('protein_id')\n",
    "\n",
    "        ogs = set()\n",
    "        for protein_id in genme_ogs.index:\n",
    "            protein_ogs = genme_ogs.loc[protein_id, 'eggNOG_OGs']\n",
    "\n",
    "            candidate_ogs = {\n",
    "                og_with_tax.split('@')[0]\n",
    "                for og_with_tax in protein_ogs.split(',')\n",
    "            }\n",
    "\n",
    "            if og_whitelist is not None:\n",
    "                candidate_ogs = candidate_ogs & og_whitelist\n",
    "                \n",
    "            ogs |= candidate_ogs\n",
    "\n",
    "        if og_intersection is None:\n",
    "            og_intersection = ogs\n",
    "        else:\n",
    "            og_intersection &= ogs\n",
    "\n",
    "    return og_intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2190"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_ogs = {\n",
    "    og\n",
    "    for og_str in eggNOG_df['eggNOG_OGs']\n",
    "    for og_raw in og_str.split(',')\n",
    "    if (og := og_raw.split('@')[0]).startswith('COG')\n",
    "}\n",
    "\n",
    "len(cog_ogs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "og_metadata = pd.read_csv(\n",
    "    strep_folder / 'eggNOG' / 'e5.og_annotations.tsv',\n",
    "    sep='\\t', \n",
    "    header=None,\n",
    "    names=['og', 'og_level', 'description'],\n",
    ").drop_duplicates('og').set_index('og')\n",
    "og_metadata.head()\n",
    "\n",
    "def display_ogs(ogs):\n",
    "    for og in ogs:\n",
    "        desc = og_metadata.loc[og, 'description']\n",
    "        print(f'{og}: {desc}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = [\n",
    "    'Streptococcus pyogenes',\n",
    "    'Streptococcus dysgalactiae',\n",
    "    'Streptococcus equi',\n",
    "    'Streptococcus canis',\n",
    "]\n",
    "metadata_subset = metadata_df[metadata_df['gtdb_species'].isin(species)].copy()\n",
    "\n",
    "accessions = metadata_subset.index\n",
    "\n",
    "eggNOG_subset = eggNOG_df.loc[accessions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "accessions_S_protein = sorted(set(eggNOG_subset[eggNOG_subset['eggNOG_OGs'].str.contains('4HHT0')].index))\n",
    "accessions_no_S_protein = sorted(set(accessions) - set(accessions_S_protein))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(856, 1237)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_S_protein = get_ogs_present_in_all(eggNOG_subset, accessions_no_S_protein, cog_ogs)\n",
    "no_S_protein_all = get_unique_ogs(eggNOG_subset, accessions_no_S_protein, cog_ogs)\n",
    "\n",
    "len(no_S_protein), len(no_S_protein_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(586, 1435)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yes_S_protein = get_ogs_present_in_all(eggNOG_subset, accessions_S_protein, cog_ogs)\n",
    "yes_S_protein_all = get_unique_ogs(eggNOG_subset, accessions_S_protein, cog_ogs)\n",
    "\n",
    "len(yes_S_protein), len(yes_S_protein_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gain in S. equi\n",
    "\n",
    "COGs present in all S. equi but in no related Strep."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gain: 13 OGs\n",
      "COG1345: cell adhesion\n",
      "COG1376: peptidoglycan L,D-transpeptidase activity\n",
      "COG1434: Gram-negative-bacterium-type cell wall biogenesis\n",
      "COG2848: Uncharacterised ACR (DUF711)\n",
      "COG3458: cephalosporin-C deacetylase activity\n",
      "COG3830: Belongs to the UPF0237 family\n",
      "COG3958: Transketolase\n",
      "COG3959: Transketolase, thiamine diphosphate binding domain\n",
      "COG4495: Domain of unknown function (DUF4176)\n",
      "COG4935: serine-type endopeptidase activity\n",
      "COG5353: nan\n",
      "COG5444: nuclease activity\n",
      "COG5585: self proteolysis\n"
     ]
    }
   ],
   "source": [
    "missing_in_yes = no_S_protein - yes_S_protein_all\n",
    "print('Gain:', len(missing_in_yes), 'OGs')\n",
    "display_ogs(sorted(missing_in_yes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loss in S. equi\n",
    "\n",
    "COGs present in all related Strep with S protein but in no S. equi."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loss: 9 OGs\n",
      "COG0010: Belongs to the arginase family\n",
      "COG0625: glutathione transferase activity\n",
      "COG0730: response to heat\n",
      "COG0813: purine-nucleoside phosphorylase activity\n",
      "COG1054: Rhodanese Homology Domain\n",
      "COG1902: FMN binding\n",
      "COG2110: O-acetyl-ADP-ribose deacetylase activity\n",
      "COG2141: oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen\n",
      "COG3643: Formiminotransferase domain\n"
     ]
    }
   ],
   "source": [
    "missing_in_no = yes_S_protein - no_S_protein_all\n",
    "print('Loss:', len(missing_in_no), 'OGs')\n",
    "display_ogs(sorted(missing_in_no))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export list of gains & losses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# OGs: 13,782 (COGs: 1,489)\n"
     ]
    }
   ],
   "source": [
    "all_ogs_in_subset = {og_raw.split('@')[0] for l in eggNOG_subset['eggNOG_OGs'].values for og_raw in l.split(',')}\n",
    "all_COGs_in_subset = {og for og in all_ogs_in_subset if og.startswith('COG')}\n",
    "n_ogs = len(all_ogs_in_subset)\n",
    "n_cogs = len(all_COGs_in_subset)\n",
    "\n",
    "print(f'# OGs: {n_ogs:,} (COGs: {n_cogs:,})')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# gains : 133 (1.0 %)\n",
      "\tCOGs: 13 (0.9 %)\n",
      "# losses: 44 (0.3 %)\n",
      "\tCOGs: 9 (0.6 %)\n"
     ]
    }
   ],
   "source": [
    "no_S_protein = get_ogs_present_in_all(eggNOG_subset, accessions_no_S_protein)\n",
    "no_S_protein_all = get_unique_ogs(eggNOG_subset, accessions_no_S_protein)\n",
    "\n",
    "yes_S_protein = get_ogs_present_in_all(eggNOG_subset, accessions_S_protein)\n",
    "yes_S_protein_all = get_unique_ogs(eggNOG_subset, accessions_S_protein)\n",
    "\n",
    "missing_in_yes = no_S_protein - yes_S_protein_all\n",
    "missing_in_no = yes_S_protein - no_S_protein_all\n",
    "\n",
    "missing_in_yes_cogs = {og for og in missing_in_yes if og.startswith('COG')}\n",
    "missing_in_no_cogs = {og for og in missing_in_no if og.startswith('COG')}\n",
    "\n",
    "print(f'# gains : {len(missing_in_yes):,} ({100 * len(missing_in_yes) / n_ogs:.1f} %)') \n",
    "print(f'\\tCOGs: {len(missing_in_yes_cogs):,} ({100 * len(missing_in_yes_cogs) / n_cogs:.1f} %)')\n",
    "print(f'# losses: {len(missing_in_no):,} ({100 * len(missing_in_no) / n_ogs:.1f} %)')\n",
    "print(f'\\tCOGs: {len(missing_in_no_cogs):,} ({100 * len(missing_in_no_cogs) / n_cogs:.1f} %)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>qlen</th>\n",
       "      <th>tlen</th>\n",
       "      <th>fident</th>\n",
       "      <th>alnlen</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>query</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WP_000031175.1@GCF_009730515.1</th>\n",
       "      <td>A0A380KSV8</td>\n",
       "      <td>417</td>\n",
       "      <td>417</td>\n",
       "      <td>1.0</td>\n",
       "      <td>417</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>417</td>\n",
       "      <td>1</td>\n",
       "      <td>417</td>\n",
       "      <td>1.929000e-311</td>\n",
       "      <td>963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_000730403.1@GCF_009730515.1</th>\n",
       "      <td>A0A1T0C0H9</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "      <td>1.0</td>\n",
       "      <td>305</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>305</td>\n",
       "      <td>1</td>\n",
       "      <td>305</td>\n",
       "      <td>2.411000e-230</td>\n",
       "      <td>719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_011999767.1@GCF_009730535.1</th>\n",
       "      <td>A0A1V0H0P2</td>\n",
       "      <td>426</td>\n",
       "      <td>426</td>\n",
       "      <td>1.0</td>\n",
       "      <td>426</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>426</td>\n",
       "      <td>1</td>\n",
       "      <td>426</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_012130945.1@GCF_009730535.1</th>\n",
       "      <td>A8AZI4</td>\n",
       "      <td>849</td>\n",
       "      <td>849</td>\n",
       "      <td>1.0</td>\n",
       "      <td>849</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>849</td>\n",
       "      <td>1</td>\n",
       "      <td>849</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_156011810.1@GCF_009731465.1</th>\n",
       "      <td>A0A6H3S4G9</td>\n",
       "      <td>250</td>\n",
       "      <td>250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>250</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "      <td>2.376000e-184</td>\n",
       "      <td>582</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    target  qlen  tlen  fident  alnlen  \\\n",
       "query                                                                    \n",
       "WP_000031175.1@GCF_009730515.1  A0A380KSV8   417   417     1.0     417   \n",
       "WP_000730403.1@GCF_009730515.1  A0A1T0C0H9   305   305     1.0     305   \n",
       "WP_011999767.1@GCF_009730535.1  A0A1V0H0P2   426   426     1.0     426   \n",
       "WP_012130945.1@GCF_009730535.1      A8AZI4   849   849     1.0     849   \n",
       "WP_156011810.1@GCF_009731465.1  A0A6H3S4G9   250   250     1.0     250   \n",
       "\n",
       "                                mismatch  qstart  qend  tstart  tend  \\\n",
       "query                                                                  \n",
       "WP_000031175.1@GCF_009730515.1         0       1   417       1   417   \n",
       "WP_000730403.1@GCF_009730515.1         0       1   305       1   305   \n",
       "WP_011999767.1@GCF_009730535.1         0       1   426       1   426   \n",
       "WP_012130945.1@GCF_009730535.1         0       1   849       1   849   \n",
       "WP_156011810.1@GCF_009731465.1         0       1   250       1   250   \n",
       "\n",
       "                                       evalue  bits  \n",
       "query                                                \n",
       "WP_000031175.1@GCF_009730515.1  1.929000e-311   963  \n",
       "WP_000730403.1@GCF_009730515.1  2.411000e-230   719  \n",
       "WP_011999767.1@GCF_009730535.1   0.000000e+00  1008  \n",
       "WP_012130945.1@GCF_009730535.1   0.000000e+00  2006  \n",
       "WP_156011810.1@GCF_009731465.1  2.376000e-184   582  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_mapping = pd.read_csv(strep_folder / 'Streptococcus_all_proteins_UniProtKB_map.tsv', sep='\\t', index_col='query')\n",
    "uniprot_set = set(uniprot_mapping.index)\n",
    "uniprot_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "      <th>is_gain</th>\n",
       "      <th>is_cog</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>example_uniprot_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eggNOG_OG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>COG1345</th>\n",
       "      <td>cell adhesion</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Flagellin_IN,FliD_C,FliD_N</td>\n",
       "      <td>C0M7S8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG1376</th>\n",
       "      <td>peptidoglycan L,D-transpeptidase activity</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>YkuD</td>\n",
       "      <td>A0A380JSM9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG1434</th>\n",
       "      <td>Gram-negative-bacterium-type cell wall biogenesis</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>DUF218</td>\n",
       "      <td>B4U0X8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG2848</th>\n",
       "      <td>Uncharacterised ACR (DUF711)</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>DUF711</td>\n",
       "      <td>C0M7F4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG3458</th>\n",
       "      <td>cephalosporin-C deacetylase activity</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AXE1</td>\n",
       "      <td>B4U0U4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 description  is_gain  is_cog  \\\n",
       "eggNOG_OG                                                                       \n",
       "COG1345                                        cell adhesion     True    True   \n",
       "COG1376            peptidoglycan L,D-transpeptidase activity     True    True   \n",
       "COG1434    Gram-negative-bacterium-type cell wall biogenesis     True    True   \n",
       "COG2848                         Uncharacterised ACR (DUF711)     True    True   \n",
       "COG3458                 cephalosporin-C deacetylase activity     True    True   \n",
       "\n",
       "                                PFAMs example_uniprot_id  \n",
       "eggNOG_OG                                                 \n",
       "COG1345    Flagellin_IN,FliD_C,FliD_N             C0M7S8  \n",
       "COG1376                          YkuD         A0A380JSM9  \n",
       "COG1434                        DUF218             B4U0X8  \n",
       "COG2848                        DUF711             C0M7F4  \n",
       "COG3458                          AXE1             B4U0U4  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_og = {\n",
    "    'eggNOG_OG': [],\n",
    "    'description': [],\n",
    "    'is_gain': [],\n",
    "    'is_cog': [],\n",
    "    'PFAMs': [],\n",
    "    'example_uniprot_id': [],\n",
    "}\n",
    "for is_gain, ogs in [(True, missing_in_yes), [False, missing_in_no]]:\n",
    "    for og in sorted(ogs):\n",
    "        description = og_metadata.loc[og, 'description']\n",
    "        if pd.isnull(description) or description == '' or description == 'nan':\n",
    "            description = None\n",
    "\n",
    "        df = eggNOG_subset[eggNOG_subset['eggNOG_OGs'].str.contains(og)]\n",
    "        protein_ids = df['protein_id'].unique()\n",
    "        proteins_with_uniprot_match = sorted(uniprot_set & set(protein_ids))\n",
    "        if len(proteins_with_uniprot_match) > 0:\n",
    "            example_uniprot_id = uniprot_mapping.loc[proteins_with_uniprot_match[0], 'target']\n",
    "        else:\n",
    "            example_uniprot_id = None\n",
    "\n",
    "        pfams = None\n",
    "        for pfam_list in df['PFAMs'].unique():\n",
    "            s = set()\n",
    "            for p in (pfam_list if isinstance(pfam_list, str) else '').split(','):\n",
    "                s.add(p)\n",
    "\n",
    "            if pfams is None:\n",
    "                pfams = s\n",
    "            else:\n",
    "                pfams &= s\n",
    "\n",
    "        pfams = ','.join(sorted(pfams))\n",
    "        \n",
    "        data_og['eggNOG_OG'].append(og)\n",
    "        data_og['description'].append(description)\n",
    "        data_og['is_gain'].append(is_gain)\n",
    "        data_og['is_cog'].append(og.startswith('COG'))\n",
    "        data_og['PFAMs'].append(pfams)\n",
    "        data_og['example_uniprot_id'].append(example_uniprot_id)\n",
    "\n",
    "comparative_genomics_df = pd.DataFrame.from_dict(data_og).set_index('eggNOG_OG', drop=True).sort_values(\n",
    "    ['is_gain', 'is_cog'],\n",
    "    ascending=False,    \n",
    ")\n",
    "comparative_genomics_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "comparative_genomics_df.to_csv(strep_folder / 'S_equi_gain_loss.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "strep_tree_path = strep_folder / 'tree' / 'Streptococcus.tree'\n",
    "strep_tree = Phylo.read(strep_tree_path, 'phyloxml')\n",
    "\n",
    "strep_tree_subset = prune_leaves_with_unknown_id(strep_tree, set(metadata_subset.index))\n",
    "\n",
    "assert len(strep_tree_subset.get_terminals()) == len(metadata_subset)\n",
    "\n",
    "strep_tree_subset_path = strep_folder / 'tree' / 'Streptococcus_zoom.tree'\n",
    "with strep_tree_subset_path.open('w') as f_out:\n",
    "     Phylo.write([strep_tree_subset], f_out, 'phyloxml')\n",
    "\n",
    "strep_tree_subset = Phylo.read(strep_tree_subset_path, 'phyloxml')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = []\n",
    "for accession in metadata_subset.index:\n",
    "    ncbi_organism_name = metadata_df.loc[accession, 'ncbi_organism_name']\n",
    "    label = f'{ncbi_organism_name} [{accession}]'\n",
    "    labels.append([accession, label])\n",
    "\n",
    "itol_labels(\n",
    "    labels, \n",
    "    strep_folder / 'tree' / 'labels_zoom.txt'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Binary annotations: presence or absence of S-protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "cwb_bins = pd.read_csv(strep_folder / 'Streptococcus_cell_wall_binding.csv', index_col='assembly_accession')\n",
    "lysM_accessions = set(cwb_bins[cwb_bins['LysM'] > 0].index)\n",
    "\n",
    "accessions_pgh = sorted(set(eggNOG_subset[\n",
    "    eggNOG_subset['eggNOG_OGs'].str.contains('COG0860') &\n",
    "    eggNOG_subset['eggNOG_OGs'].str.contains('COG1388')\n",
    "].index) & lysM_accessions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "s_protein_binary_data = []\n",
    "for accession in sorted(metadata_subset.index):\n",
    "    s_protein = '1' if accession in accessions_S_protein else '-1'\n",
    "    pgh = '1' if accession in accessions_pgh else '-1'\n",
    "    s_protein_binary_data.append([accession, s_protein, pgh])\n",
    "\n",
    "itol_binary_annotations(\n",
    "    data=s_protein_binary_data,\n",
    "    output_path=strep_folder / 'tree' / 'zoom_S_protein_binary_presence.txt',\n",
    "    field_shapes=[1, 1],\n",
    "    field_labels=['S-protein', 'Amidase_3 + LysM'],\n",
    "    dataset_label='S-protein',\n",
    "    field_colors=['#008000', '#FF6347'],\n",
    "    legend_title='S-protein',\n",
    "    height_factor=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Colored ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "range_colors = {\n",
    "    'Streptococcus pyogenes': '#a1c9f4',\n",
    "    'Streptococcus dysgalactiae': '#debb9b',\n",
    "    'Streptococcus equi': '#fab0e4',\n",
    "}\n",
    "\n",
    "colored_ranges = []\n",
    "for species, color in range_colors.items():\n",
    "    if species == 'Streptococcus canis':\n",
    "        continue\n",
    "    node_id = f's__{species}'\n",
    "    colored_ranges.append(\n",
    "        [node_id, node_id, hex_to_rgba(color, 0.25), '', '', '', '', species, '', '10', 'bold-italic']\n",
    "    )\n",
    "\n",
    "itol_colored_ranges(\n",
    "    colored_ranges,\n",
    "    output_path=strep_folder / 'tree' / 'zoom_species_colored_range.txt',\n",
    "    range_type='box',\n",
    "    range_cover='tree',\n",
    "    dataset_label='Species',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cwb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
