{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Streptococcus comparative genomics\n",
    "\n",
    "Comparing protein domain presence when a specific doamin such as LysM is present vs when it's not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as plt\n",
    "import seaborn as sns\n",
    "from Bio import Phylo\n",
    "\n",
    "cwd = os.getcwd()\n",
    "if cwd.endswith('notebook'):\n",
    "    os.chdir('..')\n",
    "    cwd = os.getcwd()\n",
    "\n",
    "from src.cell_wall_binding_domains import cwb_domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme(palette='colorblind', font_scale=1.3)\n",
    "palette_colorblind = sns.color_palette('colorblind').as_hex()\n",
    "palette_pastel = sns.color_palette('pastel').as_hex()\n",
    "\n",
    "data_folder = Path('./data/')\n",
    "assert data_folder.is_dir()\n",
    "\n",
    "db_proka = Path('../db_proka/')\n",
    "assert db_proka.is_dir()\n",
    "\n",
    "gtdb_folder = Path('../data/gtdb_r220/')\n",
    "assert gtdb_folder.is_dir()\n",
    "\n",
    "strep_folder = gtdb_folder / 'Streptococcus'\n",
    "assert strep_folder.is_dir()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>ambiguous_bases</th>\n",
       "      <th>checkm2_completeness</th>\n",
       "      <th>checkm2_contamination</th>\n",
       "      <th>checkm2_model</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>checkm_marker_count</th>\n",
       "      <th>checkm_marker_lineage</th>\n",
       "      <th>checkm_marker_set_count</th>\n",
       "      <th>...</th>\n",
       "      <th>trna_aa_count</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>trna_selenocysteine_count</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_900636555.1</th>\n",
       "      <td>RS_GCF_900636555.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.14</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>475</td>\n",
       "      <td>o__Lactobacillales (UID544)</td>\n",
       "      <td>267</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus gordonii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_015908985.1</th>\n",
       "      <td>RS_GCF_015908985.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.23</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>475</td>\n",
       "      <td>o__Lactobacillales (UID544)</td>\n",
       "      <td>267</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus suis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001266635.1</th>\n",
       "      <td>RS_GCF_001266635.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.06</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.82</td>\n",
       "      <td>0.00</td>\n",
       "      <td>524</td>\n",
       "      <td>f__Streptococcaceae (UID545)</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_004154025.1</th>\n",
       "      <td>RS_GCF_004154025.1</td>\n",
       "      <td>0</td>\n",
       "      <td>99.99</td>\n",
       "      <td>0.13</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.85</td>\n",
       "      <td>0.00</td>\n",
       "      <td>676</td>\n",
       "      <td>g__Streptococcus (UID722)</td>\n",
       "      <td>182</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_029011635.1</th>\n",
       "      <td>RS_GCF_029011635.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.20</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.18</td>\n",
       "      <td>524</td>\n",
       "      <td>f__Streptococcaceae (UID545)</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             accession  ambiguous_bases  checkm2_completeness  \\\n",
       "assembly_accession                                                              \n",
       "GCF_900636555.1     RS_GCF_900636555.1                0                100.00   \n",
       "GCF_015908985.1     RS_GCF_015908985.1                0                100.00   \n",
       "GCF_001266635.1     RS_GCF_001266635.1                0                100.00   \n",
       "GCF_004154025.1     RS_GCF_004154025.1                0                 99.99   \n",
       "GCF_029011635.1     RS_GCF_029011635.1                0                100.00   \n",
       "\n",
       "                    checkm2_contamination checkm2_model  checkm_completeness  \\\n",
       "assembly_accession                                                             \n",
       "GCF_900636555.1                      0.14      Specific               100.00   \n",
       "GCF_015908985.1                      0.23      Specific               100.00   \n",
       "GCF_001266635.1                      0.06      Specific                99.82   \n",
       "GCF_004154025.1                      0.13      Specific                99.85   \n",
       "GCF_029011635.1                      0.20      Specific               100.00   \n",
       "\n",
       "                    checkm_contamination  checkm_marker_count  \\\n",
       "assembly_accession                                              \n",
       "GCF_900636555.1                     0.00                  475   \n",
       "GCF_015908985.1                     0.00                  475   \n",
       "GCF_001266635.1                     0.00                  524   \n",
       "GCF_004154025.1                     0.00                  676   \n",
       "GCF_029011635.1                     0.18                  524   \n",
       "\n",
       "                           checkm_marker_lineage  checkm_marker_set_count  \\\n",
       "assembly_accession                                                          \n",
       "GCF_900636555.1      o__Lactobacillales (UID544)                      267   \n",
       "GCF_015908985.1      o__Lactobacillales (UID544)                      267   \n",
       "GCF_001266635.1     f__Streptococcaceae (UID545)                      282   \n",
       "GCF_004154025.1        g__Streptococcus (UID722)                      182   \n",
       "GCF_029011635.1     f__Streptococcaceae (UID545)                      282   \n",
       "\n",
       "                    ...  trna_aa_count  trna_count  trna_selenocysteine_count  \\\n",
       "assembly_accession  ...                                                         \n",
       "GCF_900636555.1     ...             19          59                          0   \n",
       "GCF_015908985.1     ...             19          56                          0   \n",
       "GCF_001266635.1     ...             18          63                          0   \n",
       "GCF_004154025.1     ...             19          67                          0   \n",
       "GCF_029011635.1     ...             19          80                          0   \n",
       "\n",
       "                      domain  gtdb_phylum  gtdb_class       gtdb_order  \\\n",
       "assembly_accession                                                       \n",
       "GCF_900636555.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_015908985.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_001266635.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_004154025.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_029011635.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "\n",
       "                         gtdb_family     gtdb_genus              gtdb_species  \n",
       "assembly_accession                                                             \n",
       "GCF_900636555.1     Streptococcaceae  Streptococcus    Streptococcus gordonii  \n",
       "GCF_015908985.1     Streptococcaceae  Streptococcus        Streptococcus suis  \n",
       "GCF_001266635.1     Streptococcaceae  Streptococcus  Streptococcus agalactiae  \n",
       "GCF_004154025.1     Streptococcaceae  Streptococcus    Streptococcus pyogenes  \n",
       "GCF_029011635.1     Streptococcaceae  Streptococcus  Streptococcus agalactiae  \n",
       "\n",
       "[5 rows x 120 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df = pd.read_csv(strep_folder / 'genomes_metadata.csv', index_col='assembly_accession')\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfam_df = pd.read_csv(strep_folder / 'Streptococcus_all_proteins.pfam.csv', index_col='assembly_accession')\n",
    "pfam_df['gtdb_species'] = [metadata_df.loc[a, 'gtdb_species'] for a in pfam_df.index]\n",
    "\n",
    "tigr_df = pd.read_csv(strep_folder / 'Streptococcus_all_proteins.tigr.csv', index_col='assembly_accession')\n",
    "tigr_df['gtdb_species'] = [metadata_df.loc[a, 'gtdb_species'] for a in tigr_df.index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How many LysM per genomes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gtdb_species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Streptococcus pyogenes</th>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus agalactiae</th>\n",
       "      <td>611</td>\n",
       "      <td>197</td>\n",
       "      <td>3.101523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pneumoniae</th>\n",
       "      <td>313</td>\n",
       "      <td>158</td>\n",
       "      <td>1.981013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus suis</th>\n",
       "      <td>268</td>\n",
       "      <td>109</td>\n",
       "      <td>2.458716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus thermophilus</th>\n",
       "      <td>310</td>\n",
       "      <td>88</td>\n",
       "      <td>3.522727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus dysgalactiae</th>\n",
       "      <td>57</td>\n",
       "      <td>43</td>\n",
       "      <td>1.325581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus mutans</th>\n",
       "      <td>68</td>\n",
       "      <td>26</td>\n",
       "      <td>2.615385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus iniae</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus suis_W</th>\n",
       "      <td>39</td>\n",
       "      <td>13</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus gordonii</th>\n",
       "      <td>52</td>\n",
       "      <td>13</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             id  assembly_accession     ratio\n",
       "gtdb_species                                                 \n",
       "Streptococcus pyogenes      271                 271  1.000000\n",
       "Streptococcus agalactiae    611                 197  3.101523\n",
       "Streptococcus pneumoniae    313                 158  1.981013\n",
       "Streptococcus suis          268                 109  2.458716\n",
       "Streptococcus thermophilus  310                  88  3.522727\n",
       "Streptococcus dysgalactiae   57                  43  1.325581\n",
       "Streptococcus mutans         68                  26  2.615385\n",
       "Streptococcus iniae          14                  14  1.000000\n",
       "Streptococcus suis_W         39                  13  3.000000\n",
       "Streptococcus gordonii       52                  13  4.000000"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lysM_df = pfam_df[pfam_df['hmm_query'] == 'LysM'].copy()\n",
    "\n",
    "n_copies_per_species = lysM_df.reset_index()[\n",
    "    ['gtdb_species', 'id', 'assembly_accession']\n",
    "].groupby('gtdb_species').nunique()\n",
    "\n",
    "n_copies_per_species['ratio'] = n_copies_per_species['id'] / n_copies_per_species['assembly_accession']\n",
    "n_copies_per_species.sort_values('assembly_accession', ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>hmm_accession</th>\n",
       "      <th>hmm_query</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gtdb_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_010921965.1@GCF_002055535.1</td>\n",
       "      <td>WP_010921965.1</td>\n",
       "      <td>PF01476.25</td>\n",
       "      <td>LysM</td>\n",
       "      <td>1.600000e-10</td>\n",
       "      <td>37.6</td>\n",
       "      <td>0.97</td>\n",
       "      <td>48</td>\n",
       "      <td>91</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                id      protein_id  \\\n",
       "assembly_accession                                                   \n",
       "GCF_002055535.1     WP_010921965.1@GCF_002055535.1  WP_010921965.1   \n",
       "\n",
       "                   hmm_accession hmm_query        evalue  bitscore  accuracy  \\\n",
       "assembly_accession                                                             \n",
       "GCF_002055535.1       PF01476.25      LysM  1.600000e-10      37.6      0.97   \n",
       "\n",
       "                    start  end            gtdb_species  \n",
       "assembly_accession                                      \n",
       "GCF_002055535.1        48   91  Streptococcus pyogenes  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_pyogenes_repr = 'GCF_002055535.1'\n",
    "df = pfam_df.loc[s_pyogenes_repr]\n",
    "df[df['hmm_query'] == 'LysM']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LysM in S. equi\n",
    "\n",
    "We spotted a pattern of differential presence of LysM in certain strains of _Steptococcus equi_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of S. equi genomes:              38\n",
      "Number of S. equi genones with LysM:    4\n",
      "Number of S. equi genones without LysM: 34\n"
     ]
    }
   ],
   "source": [
    "accessions_s_equi = metadata_df[\n",
    "    metadata_df['gtdb_species'] == 'Streptococcus equi'\n",
    "].index\n",
    "\n",
    "pfam_s_equi = pfam_df.loc[accessions_s_equi]\n",
    "tigr_s_equi = tigr_df.loc[accessions_s_equi]\n",
    "\n",
    "accessions_yes_LysM = set(pfam_s_equi[\n",
    "    pfam_s_equi['hmm_query'] == 'LysM'\n",
    "].index)\n",
    "\n",
    "accessions_no_LysM = set(accessions_s_equi) - accessions_yes_LysM\n",
    "\n",
    "print(f'Number of S. equi genomes:              {len(accessions_s_equi):,}')\n",
    "print(f'Number of S. equi genones with LysM:    {len(accessions_yes_LysM):,}')\n",
    "print(f'Number of S. equi genones without LysM: {len(accessions_no_LysM):,}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "263"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extra_in_no = (\n",
    "    (\n",
    "        set(pfam_s_equi.loc[sorted(accessions_no_LysM)]['hmm_query'].values) |\n",
    "        set(tigr_s_equi.loc[sorted(accessions_no_LysM)]['hmm_query'].values)\n",
    "    ) - \n",
    "    (\n",
    "        set(pfam_s_equi.loc[sorted(accessions_yes_LysM)]['hmm_query'].values) |\n",
    "        set(tigr_s_equi.loc[sorted(accessions_yes_LysM)]['hmm_query'].values)\n",
    "    )\n",
    ")\n",
    "len(extra_in_no)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('TIGR03426', 26, 76.5),\n",
       " ('Malt_amylase_C', 24, 70.6),\n",
       " ('FokI_D1', 21, 61.8),\n",
       " ('FokI_dom_2', 21, 61.8),\n",
       " ('AAA_10', 21, 61.8),\n",
       " ('FokI_D3', 21, 61.8),\n",
       " ('FokI_cleav_dom', 21, 61.8),\n",
       " ('MethyltransfD12', 20, 58.8),\n",
       " ('Polysacc_synt_3', 20, 58.8),\n",
       " ('TIGR00571', 19, 55.9),\n",
       " ('Phage_int_SAM_5', 19, 55.9),\n",
       " ('Gp58', 18, 52.9),\n",
       " ('Cellulase', 17, 50.0),\n",
       " ('RloB', 16, 47.1),\n",
       " ('HTH_36', 16, 47.1),\n",
       " ('T6_Ig_like', 16, 47.1),\n",
       " ('TIGR01637', 15, 44.1),\n",
       " ('TIGR01451', 15, 44.1),\n",
       " ('Bro-N', 14, 41.2),\n",
       " ('AAA_13', 14, 41.2)]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfam_no_lysM = pfam_s_equi.loc[sorted(accessions_no_LysM)]\n",
    "tigr_no_lysM = tigr_s_equi.loc[sorted(accessions_no_LysM)]\n",
    "\n",
    "domain_with_count = []\n",
    "for domain in extra_in_no:\n",
    "    if 'TIGR' in domain:\n",
    "        df = tigr_no_lysM\n",
    "    else:\n",
    "        df = pfam_no_lysM\n",
    "\n",
    "    count = len(set(df[df['hmm_query'] == domain].index))\n",
    "    percent = float(np.round(100 * count / len(accessions_no_LysM), 1))\n",
    "\n",
    "    domain_with_count.append((domain, count, percent))\n",
    "\n",
    "domain_with_count = sorted(domain_with_count, key=lambda t: t[1], reverse=True)\n",
    "domain_with_count[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_in_no = (\n",
    "    (\n",
    "        set(pfam_s_equi.loc[sorted(accessions_yes_LysM)]['hmm_query'].values) |\n",
    "        set(tigr_s_equi.loc[sorted(accessions_yes_LysM)]['hmm_query'].values)\n",
    "    ) - \n",
    "    (\n",
    "        set(pfam_s_equi.loc[sorted(accessions_no_LysM)]['hmm_query'].values) |\n",
    "        set(tigr_s_equi.loc[sorted(accessions_no_LysM)]['hmm_query'].values)\n",
    "    )\n",
    ")\n",
    "len(missing_in_no)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Amidase_3', 4, 100.0),\n",
       " ('LysM', 4, 100.0),\n",
       " ('DUF6900', 4, 100.0),\n",
       " ('SLFN-g3_helicase', 3, 75.0),\n",
       " ('CTD11', 3, 75.0),\n",
       " ('PhdYeFM_antitox', 3, 75.0),\n",
       " ('ToxN_toxin', 3, 75.0),\n",
       " ('MeaB', 3, 75.0),\n",
       " ('MC7', 3, 75.0),\n",
       " ('DUF2326', 3, 75.0),\n",
       " ('DEDD_Tnp_IS110', 2, 50.0),\n",
       " ('DUF3781', 1, 25.0),\n",
       " ('DUF4145', 1, 25.0),\n",
       " ('HNH_3', 1, 25.0),\n",
       " ('Imm63', 1, 25.0),\n",
       " ('TrbL', 1, 25.0)]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfam_yes_lysM = pfam_s_equi.loc[sorted(accessions_yes_LysM)]\n",
    "tigr_yes_lysM = tigr_s_equi.loc[sorted(accessions_yes_LysM)]\n",
    "\n",
    "missing_domain_with_count = []\n",
    "for domain in missing_in_no:\n",
    "    if 'TIGR' in domain:\n",
    "        df = tigr_yes_lysM\n",
    "    else:\n",
    "        df = pfam_yes_lysM\n",
    "\n",
    "    count = len(set(df[df['hmm_query'] == domain].index))\n",
    "    percent = float(np.round(100 * count / len(accessions_yes_LysM), 1))\n",
    "\n",
    "    missing_domain_with_count.append((domain, count, percent))\n",
    "\n",
    "missing_domain_with_count = sorted(missing_domain_with_count, key=lambda t: t[1], reverse=True)\n",
    "missing_domain_with_count[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not particularly conclusive at this level of analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell wall HMM search\n",
    "\n",
    "HMM models from [Megrian et al., 2022](https://doi.org/10.1038/s41564-022-01257-y)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>hmm_accession</th>\n",
       "      <th>hmm_query</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35082.1@GCA_000013525.1</td>\n",
       "      <td>ABF35082.1</td>\n",
       "      <td>newDdlB</td>\n",
       "      <td>newDdlB</td>\n",
       "      <td>6.500000e-08</td>\n",
       "      <td>28.6</td>\n",
       "      <td>0.77</td>\n",
       "      <td>126</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35084.1@GCA_000013525.1</td>\n",
       "      <td>ABF35084.1</td>\n",
       "      <td>newDdlB</td>\n",
       "      <td>newDdlB</td>\n",
       "      <td>3.500000e-11</td>\n",
       "      <td>39.3</td>\n",
       "      <td>0.79</td>\n",
       "      <td>106</td>\n",
       "      <td>289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35137.1@GCA_000013525.1</td>\n",
       "      <td>ABF35137.1</td>\n",
       "      <td>newFtsI</td>\n",
       "      <td>newFtsI</td>\n",
       "      <td>1.900000e-21</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.79</td>\n",
       "      <td>329</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35311.1@GCA_000013525.1</td>\n",
       "      <td>ABF35311.1</td>\n",
       "      <td>newMraY</td>\n",
       "      <td>newMraY</td>\n",
       "      <td>1.400000e-35</td>\n",
       "      <td>119.7</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0</td>\n",
       "      <td>330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35345.1@GCA_000013525.1</td>\n",
       "      <td>ABF35345.1</td>\n",
       "      <td>newMurD</td>\n",
       "      <td>newMurD</td>\n",
       "      <td>3.800000e-05</td>\n",
       "      <td>19.8</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            id  protein_id hmm_accession  \\\n",
       "assembly_accession                                                         \n",
       "GCA_000013525.1     ABF35082.1@GCA_000013525.1  ABF35082.1       newDdlB   \n",
       "GCA_000013525.1     ABF35084.1@GCA_000013525.1  ABF35084.1       newDdlB   \n",
       "GCA_000013525.1     ABF35137.1@GCA_000013525.1  ABF35137.1       newFtsI   \n",
       "GCA_000013525.1     ABF35311.1@GCA_000013525.1  ABF35311.1       newMraY   \n",
       "GCA_000013525.1     ABF35345.1@GCA_000013525.1  ABF35345.1       newMurD   \n",
       "\n",
       "                   hmm_query        evalue  bitscore  accuracy  start  end  \n",
       "assembly_accession                                                          \n",
       "GCA_000013525.1      newDdlB  6.500000e-08      28.6      0.77    126  303  \n",
       "GCA_000013525.1      newDdlB  3.500000e-11      39.3      0.79    106  289  \n",
       "GCA_000013525.1      newFtsI  1.900000e-21      73.0      0.79    329  642  \n",
       "GCA_000013525.1      newMraY  1.400000e-35     119.7      0.79      0  330  \n",
       "GCA_000013525.1      newMurD  3.800000e-05      19.8      0.77      0  117  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cw_df = pd.read_csv(strep_folder / 'Streptococcus_all_proteins.cell_wall.csv', index_col='assembly_accession')\n",
    "\n",
    "# Keep only one model per protein â€“ best scoring one.\n",
    "cw_df = cw_df.sort_values(['id', 'bitscore'], ascending=[True, False]).drop_duplicates(['id'])\n",
    "\n",
    "cw_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hmm_query</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>newDdlB</th>\n",
       "      <td>20</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newFtsI</th>\n",
       "      <td>20</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMraW</th>\n",
       "      <td>12</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurA</th>\n",
       "      <td>12</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurD</th>\n",
       "      <td>12</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newFtsA</th>\n",
       "      <td>8</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newFtsW</th>\n",
       "      <td>8</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMraY</th>\n",
       "      <td>8</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurE</th>\n",
       "      <td>8</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newFtsZ</th>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurB</th>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurC</th>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurF</th>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurG</th>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           count  ratio\n",
       "hmm_query              \n",
       "newDdlB       20    5.0\n",
       "newFtsI       20    5.0\n",
       "newMraW       12    3.0\n",
       "newMurA       12    3.0\n",
       "newMurD       12    3.0\n",
       "newFtsA        8    2.0\n",
       "newFtsW        8    2.0\n",
       "newMraY        8    2.0\n",
       "newMurE        8    2.0\n",
       "newFtsZ        4    1.0\n",
       "newMurB        4    1.0\n",
       "newMurC        4    1.0\n",
       "newMurF        4    1.0\n",
       "newMurG        4    1.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cw_df_yes_lysM = cw_df.loc[sorted(accessions_yes_LysM)]\n",
    "\n",
    "cw_yes_lysM_grouped = cw_df_yes_lysM[\n",
    "    ['id', 'hmm_query']\n",
    "].groupby(\n",
    "    'hmm_query'\n",
    ").nunique().sort_values('id', ascending=False).rename(\n",
    "    columns={\n",
    "        'id': 'count',\n",
    "    }\n",
    ")\n",
    "cw_yes_lysM_grouped['ratio'] = cw_yes_lysM_grouped['count'] / len(accessions_yes_LysM)\n",
    "cw_yes_lysM_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hmm_query</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>newFtsI</th>\n",
       "      <td>170</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newDdlB</th>\n",
       "      <td>169</td>\n",
       "      <td>4.970588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurA</th>\n",
       "      <td>102</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurD</th>\n",
       "      <td>101</td>\n",
       "      <td>2.970588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMraW</th>\n",
       "      <td>94</td>\n",
       "      <td>2.764706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newFtsW</th>\n",
       "      <td>68</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMraY</th>\n",
       "      <td>68</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newFtsA</th>\n",
       "      <td>67</td>\n",
       "      <td>1.970588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurC</th>\n",
       "      <td>52</td>\n",
       "      <td>1.529412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurE</th>\n",
       "      <td>48</td>\n",
       "      <td>1.411765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurF</th>\n",
       "      <td>35</td>\n",
       "      <td>1.029412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurB</th>\n",
       "      <td>34</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMurG</th>\n",
       "      <td>34</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newFtsZ</th>\n",
       "      <td>33</td>\n",
       "      <td>0.970588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newMraZ</th>\n",
       "      <td>1</td>\n",
       "      <td>0.029412</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           count     ratio\n",
       "hmm_query                 \n",
       "newFtsI      170  5.000000\n",
       "newDdlB      169  4.970588\n",
       "newMurA      102  3.000000\n",
       "newMurD      101  2.970588\n",
       "newMraW       94  2.764706\n",
       "newFtsW       68  2.000000\n",
       "newMraY       68  2.000000\n",
       "newFtsA       67  1.970588\n",
       "newMurC       52  1.529412\n",
       "newMurE       48  1.411765\n",
       "newMurF       35  1.029412\n",
       "newMurB       34  1.000000\n",
       "newMurG       34  1.000000\n",
       "newFtsZ       33  0.970588\n",
       "newMraZ        1  0.029412"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cw_df_no_lysM = cw_df.loc[sorted(accessions_no_LysM)]\n",
    "\n",
    "cw_no_lysM_grouped = cw_df_no_lysM[\n",
    "    ['id', 'hmm_query']\n",
    "].groupby(\n",
    "    'hmm_query'\n",
    ").nunique().sort_values('id', ascending=False).rename(\n",
    "    columns={\n",
    "        'id': 'count',\n",
    "    }\n",
    ")\n",
    "cw_no_lysM_grouped['ratio'] = cw_no_lysM_grouped['count'] / len(accessions_no_LysM)\n",
    "cw_no_lysM_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'newMraZ'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(cw_no_lysM_grouped.index) - set(cw_yes_lysM_grouped.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(cw_yes_lysM_grouped.index) - set(cw_no_lysM_grouped.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## eggNOG orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467658.1@GCF_003963555.1</td>\n",
       "      <td>1000570.HMPREF9966_1759</td>\n",
       "      <td>9.704000e-82</td>\n",
       "      <td>271.0</td>\n",
       "      <td>COG0716@1|root,COG0716@2|Bacteria,1V45R@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>C</td>\n",
       "      <td>Flavodoxin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Flavodoxin_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467724.1@GCF_003963555.1</td>\n",
       "      <td>862970.SAIN_1577</td>\n",
       "      <td>1.313000e-186</td>\n",
       "      <td>582.0</td>\n",
       "      <td>COG1396@1|root,COG1396@2|Bacteria,1VIH9@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>K</td>\n",
       "      <td>Helix-turn-helix XRE-family like proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HTH_19,HTH_3,TPR_12,TPR_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467790.1@GCF_003963555.1</td>\n",
       "      <td>862970.SAIN_1636</td>\n",
       "      <td>2.776000e-131</td>\n",
       "      <td>418.0</td>\n",
       "      <td>COG1564@1|root,COG1564@2|Bacteria,1VA0W@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>H</td>\n",
       "      <td>Thiamin pyrophosphokinase, vitamin B1 binding ...</td>\n",
       "      <td>thiN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K00949</td>\n",
       "      <td>ko00730,ko01100,map00730,map01100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R00619</td>\n",
       "      <td>RC00002,RC00017</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TPK_B1_binding,TPK_catalytic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467857.1@GCF_003963555.1</td>\n",
       "      <td>176090.SSIN_0693</td>\n",
       "      <td>1.181000e-201</td>\n",
       "      <td>627.0</td>\n",
       "      <td>COG3677@1|root,COG3677@2|Bacteria,1V4D1@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>L</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDE_Tnp_IS1595,Zn_Tnp_IS1595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467993.1@GCF_003963555.1</td>\n",
       "      <td>862969.SCI_1925</td>\n",
       "      <td>1.812000e-256</td>\n",
       "      <td>791.0</td>\n",
       "      <td>COG0612@1|root,COG0612@2|Bacteria,1TPN6@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>Peptidase M16 inactive</td>\n",
       "      <td>ymfF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peptidase_M16,Peptidase_M16_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        protein_id            seed_ortholog  \\\n",
       "assembly_accession                                                            \n",
       "GCF_003963555.1     WP_126467658.1@GCF_003963555.1  1000570.HMPREF9966_1759   \n",
       "GCF_003963555.1     WP_126467724.1@GCF_003963555.1         862970.SAIN_1577   \n",
       "GCF_003963555.1     WP_126467790.1@GCF_003963555.1         862970.SAIN_1636   \n",
       "GCF_003963555.1     WP_126467857.1@GCF_003963555.1         176090.SSIN_0693   \n",
       "GCF_003963555.1     WP_126467993.1@GCF_003963555.1          862969.SCI_1925   \n",
       "\n",
       "                           evalue  score  \\\n",
       "assembly_accession                         \n",
       "GCF_003963555.1      9.704000e-82  271.0   \n",
       "GCF_003963555.1     1.313000e-186  582.0   \n",
       "GCF_003963555.1     2.776000e-131  418.0   \n",
       "GCF_003963555.1     1.181000e-201  627.0   \n",
       "GCF_003963555.1     1.812000e-256  791.0   \n",
       "\n",
       "                                                           eggNOG_OGs  \\\n",
       "assembly_accession                                                      \n",
       "GCF_003963555.1     COG0716@1|root,COG0716@2|Bacteria,1V45R@1239|F...   \n",
       "GCF_003963555.1     COG1396@1|root,COG1396@2|Bacteria,1VIH9@1239|F...   \n",
       "GCF_003963555.1     COG1564@1|root,COG1564@2|Bacteria,1VA0W@1239|F...   \n",
       "GCF_003963555.1     COG3677@1|root,COG3677@2|Bacteria,1V4D1@1239|F...   \n",
       "GCF_003963555.1     COG0612@1|root,COG0612@2|Bacteria,1TPN6@1239|F...   \n",
       "\n",
       "                   max_annot_lvl COG_category  \\\n",
       "assembly_accession                              \n",
       "GCF_003963555.1       2|Bacteria            C   \n",
       "GCF_003963555.1       2|Bacteria            K   \n",
       "GCF_003963555.1       2|Bacteria            H   \n",
       "GCF_003963555.1       2|Bacteria            L   \n",
       "GCF_003963555.1       2|Bacteria            S   \n",
       "\n",
       "                                                          Description  \\\n",
       "assembly_accession                                                      \n",
       "GCF_003963555.1                                            Flavodoxin   \n",
       "GCF_003963555.1             Helix-turn-helix XRE-family like proteins   \n",
       "GCF_003963555.1     Thiamin pyrophosphokinase, vitamin B1 binding ...   \n",
       "GCF_003963555.1                         ISXO2-like transposase domain   \n",
       "GCF_003963555.1                                Peptidase M16 inactive   \n",
       "\n",
       "                   Preferred_name  GOs  ...    KEGG_ko  \\\n",
       "assembly_accession                      ...              \n",
       "GCF_003963555.1               NaN  NaN  ...        NaN   \n",
       "GCF_003963555.1               NaN  NaN  ...        NaN   \n",
       "GCF_003963555.1              thiN  NaN  ...  ko:K00949   \n",
       "GCF_003963555.1               NaN  NaN  ...        NaN   \n",
       "GCF_003963555.1              ymfF  NaN  ...        NaN   \n",
       "\n",
       "                                         KEGG_Pathway KEGG_Module  \\\n",
       "assembly_accession                                                  \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "GCF_003963555.1     ko00730,ko01100,map00730,map01100         NaN   \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "\n",
       "                   KEGG_Reaction      KEGG_rclass                    BRITE  \\\n",
       "assembly_accession                                                           \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "GCF_003963555.1           R00619  RC00002,RC00017  ko00000,ko00001,ko01000   \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "\n",
       "                   KEGG_TC CAZy BiGG_Reaction                          PFAMs  \n",
       "assembly_accession                                                            \n",
       "GCF_003963555.1        NaN  NaN           NaN                   Flavodoxin_4  \n",
       "GCF_003963555.1        NaN  NaN           NaN      HTH_19,HTH_3,TPR_12,TPR_8  \n",
       "GCF_003963555.1        NaN  NaN           NaN   TPK_B1_binding,TPK_catalytic  \n",
       "GCF_003963555.1        NaN  NaN           NaN   DDE_Tnp_IS1595,Zn_Tnp_IS1595  \n",
       "GCF_003963555.1        NaN  NaN           NaN  Peptidase_M16,Peptidase_M16_C  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggNOG_annotations_path = strep_folder / 'Streptococcus_eggNOG_annotations.csv'\n",
    "\n",
    "if not eggNOG_annotations_path.is_file():\n",
    "    eggNOG_df = pd.read_csv(\n",
    "        strep_folder / 'eggNOG' / 'Streptococcus.emapper.annotations',\n",
    "        sep='\\t',\n",
    "        comment='#',\n",
    "        header=None,\n",
    "        names=[\n",
    "            'protein_id', 'seed_ortholog', 'evalue', 'score', 'eggNOG_OGs', 'max_annot_lvl', \n",
    "            'COG_category', 'Description', 'Preferred_name', 'GOs', 'EC', \n",
    "            'KEGG_ko', 'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction', 'KEGG_rclass', \n",
    "            'BRITE', 'KEGG_TC', 'CAZy', 'BiGG_Reaction', 'PFAMs',\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    eggNOG_df['assembly_accession'] = eggNOG_df['protein_id'].apply(lambda v: v.split('@')[1])\n",
    "    eggNOG_df = eggNOG_df.set_index('assembly_accession', drop=True).replace('-', np.nan)\n",
    "\n",
    "    eggNOG_df.to_csv(eggNOG_annotations_path)\n",
    "\n",
    "eggNOG_df = pd.read_csv(eggNOG_annotations_path, index_col='assembly_accession')\n",
    "eggNOG_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Streptococcus equi subsp. zooepidemicus ATCC 35246'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_equi_yes_lysM_instance = sorted(accessions_yes_LysM)[0]\n",
    "metadata_df.loc[s_equi_yes_lysM_instance, 'ncbi_organism_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'WP_014622634.1@GCF_000219765.1'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pfam_df.loc[s_equi_yes_lysM_instance]\n",
    "s_equi_yes_lysM_protein_id = df[df['hmm_query'] == 'LysM']['id'].values[0]\n",
    "s_equi_yes_lysM_protein_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COG0860@1|root,COG1388@1|root,COG0860@2|Bacteria,COG1388@2|Bacteria,1VINV@1239|Firmicutes,4HPIN@91061|Bacilli\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "protein_id                           WP_014622634.1@GCF_000219765.1\n",
       "seed_ortholog                          1123299.AUKZ01000026_gene846\n",
       "evalue                                                          0.0\n",
       "score                                                         786.0\n",
       "eggNOG_OGs        COG0860@1|root,COG1388@1|root,COG0860@2|Bacter...\n",
       "max_annot_lvl                                            2|Bacteria\n",
       "COG_category                                                      M\n",
       "Description                      N-acetylmuramoyl-L-alanine amidase\n",
       "Preferred_name                                                 yocH\n",
       "GOs                                                             NaN\n",
       "EC                                         3.2.1.1,3.5.1.28,3.5.2.6\n",
       "KEGG_ko           ko:K01176,ko:K01448,ko:K02067,ko:K06385,ko:K17...\n",
       "KEGG_Pathway      ko00311,ko00500,ko01100,ko01130,ko01501,ko0150...\n",
       "KEGG_Module               M00210,M00627,M00628,M00669,M00670,M00727\n",
       "KEGG_Reaction                    R02108,R02112,R04112,R06363,R11262\n",
       "KEGG_rclass                                 RC00064,RC00141,RC01499\n",
       "BRITE             ko00000,ko00001,ko00002,ko01000,ko01002,ko0101...\n",
       "KEGG_TC                                                    3.A.1.27\n",
       "CAZy                                                           GH13\n",
       "BiGG_Reaction                                                   NaN\n",
       "PFAMs             3D,Amidase_3,CHAP,Cu_amine_oxidN1,Lipase_GDSL,...\n",
       "Name: GCF_000219765.1, dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r = eggNOG_df[eggNOG_df['protein_id'] == s_equi_yes_lysM_protein_id].iloc[0]\n",
    "\n",
    "print(r['eggNOG_OGs'])\n",
    "\n",
    "r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COG1388@1|root,COG1388@2|Bacteria,1V61Q@1239|Firmicutes,4HHT0@91061|Bacilli\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "protein_id                           WP_016355818.1@GCF_030732225.1\n",
       "seed_ortholog                                       1346.DQ08_01735\n",
       "evalue                                                          0.0\n",
       "score                                                         705.0\n",
       "eggNOG_OGs        COG1388@1|root,COG1388@2|Bacteria,1V61Q@1239|F...\n",
       "max_annot_lvl                                            2|Bacteria\n",
       "COG_category                                                      M\n",
       "Description                                     LysM domain protein\n",
       "Preferred_name                                                  sip\n",
       "GOs                                                             NaN\n",
       "EC                                                              NaN\n",
       "KEGG_ko                                                         NaN\n",
       "KEGG_Pathway                                                    NaN\n",
       "KEGG_Module                                                     NaN\n",
       "KEGG_Reaction                                                   NaN\n",
       "KEGG_rclass                                                     NaN\n",
       "BRITE                                                           NaN\n",
       "KEGG_TC                                                         NaN\n",
       "CAZy                                                            NaN\n",
       "BiGG_Reaction                                                   NaN\n",
       "PFAMs                                                          LysM\n",
       "Name: GCF_030732225.1, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_iniae_acc = 'GCF_030732225.1'\n",
    "df = pfam_df.loc[s_iniae_acc]\n",
    "s_iniae_yes_lysM_protein_id = df[df['hmm_query'] == 'LysM']['id'].values[0]\n",
    "\n",
    "eggNOG_df[eggNOG_df['protein_id'] == s_iniae_yes_lysM_protein_id].iloc[0]\n",
    "\n",
    "r2 = eggNOG_df[eggNOG_df['protein_id'] == s_iniae_yes_lysM_protein_id].iloc[0]\n",
    "\n",
    "print(r2['eggNOG_OGs'])\n",
    "\n",
    "r2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "accessions_with_S_protein = sorted(set(eggNOG_df[eggNOG_df['eggNOG_OGs'].str.contains('4HHT0')].index))\n",
    "strep_with_S_protein = metadata_df.loc[accessions_with_S_protein]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>ambiguous_bases</th>\n",
       "      <th>checkm2_completeness</th>\n",
       "      <th>checkm2_contamination</th>\n",
       "      <th>checkm2_model</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>checkm_marker_count</th>\n",
       "      <th>checkm_marker_lineage</th>\n",
       "      <th>checkm_marker_set_count</th>\n",
       "      <th>...</th>\n",
       "      <th>trna_aa_count</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>trna_selenocysteine_count</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, ambiguous_bases, checkm2_completeness, checkm2_contamination, checkm2_model, checkm_completeness, checkm_contamination, checkm_marker_count, checkm_marker_lineage, checkm_marker_set_count, checkm_strain_heterogeneity, coding_bases, coding_density, contig_count, gc_count, gc_percentage, genome_size, gtdb_genome_representative, gtdb_representative, gtdb_taxonomy, gtdb_type_designation_ncbi_taxa, gtdb_type_designation_ncbi_taxa_sources, gtdb_type_species_of_genus, l50_contigs, l50_scaffolds, longest_contig, longest_scaffold, lsu_23s_contig_len, lsu_23s_count, lsu_23s_length, lsu_23s_query_id, lsu_5s_contig_len, lsu_5s_count, lsu_5s_length, lsu_5s_query_id, lsu_silva_23s_blast_align_len, lsu_silva_23s_blast_bitscore, lsu_silva_23s_blast_evalue, lsu_silva_23s_blast_perc_identity, lsu_silva_23s_blast_subject_id, lsu_silva_23s_taxonomy, mean_contig_length, mean_scaffold_length, mimag_high_quality, mimag_low_quality, mimag_medium_quality, n50_contigs, n50_scaffolds, ncbi_assembly_level, ncbi_assembly_name, ncbi_assembly_type, ncbi_bioproject, ncbi_biosample, ncbi_contig_count, ncbi_contig_n50, ncbi_country, ncbi_date, ncbi_genbank_assembly_accession, ncbi_genome_category, ncbi_genome_representation, ncbi_isolate, ncbi_isolation_source, ncbi_lat_lon, ncbi_molecule_count, ncbi_ncrna_count, ncbi_organism_name, ncbi_protein_count, ncbi_refseq_category, ncbi_rrna_count, ncbi_scaffold_count, ncbi_scaffold_l50, ncbi_scaffold_n50, ncbi_scaffold_n75, ncbi_scaffold_n90, ncbi_seq_rel_date, ncbi_spanned_gaps, ncbi_species_taxid, ncbi_ssu_count, ncbi_strain_identifiers, ncbi_submitter, ncbi_taxid, ncbi_taxonomy, ncbi_taxonomy_unfiltered, ncbi_total_gap_length, ncbi_total_length, ncbi_translation_table, ncbi_trna_count, ncbi_type_material_designation, ncbi_ungapped_length, ncbi_unspanned_gaps, ncbi_wgs_master, protein_count, scaffold_count, ssu_contig_len, ssu_count, ssu_gg_blast_align_len, ssu_gg_blast_bitscore, ssu_gg_blast_evalue, ssu_gg_blast_perc_identity, ssu_gg_blast_subject_id, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 120 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strep_with_S_protein[strep_with_S_protein['gtdb_species'] == 'Streptococcus equi']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>og_level</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>og</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4HBBX</th>\n",
       "      <td>L</td>\n",
       "      <td>Recombinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1XVC3</th>\n",
       "      <td>K</td>\n",
       "      <td>transcriptional regulator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2TDXG</th>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2QWWR</th>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37HD9</th>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      og_level                description\n",
       "og                                       \n",
       "4HBBX        L                Recombinase\n",
       "1XVC3        K  transcriptional regulator\n",
       "2TDXG        S                        NaN\n",
       "2QWWR        S                        NaN\n",
       "37HD9        S                        NaN"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_metadata = pd.read_csv(\n",
    "    strep_folder / 'eggNOG' / 'e5.og_annotations.tsv',\n",
    "    sep='\\t', \n",
    "    header=None,\n",
    "    names=['og', 'og_level', 'description'],\n",
    ").drop_duplicates('og').set_index('og')\n",
    "og_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_ogs(ogs):\n",
    "    for og in ogs:\n",
    "        desc = og_metadata.loc[og, 'description']\n",
    "        print(f'{og}: {desc}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_ogs(eggNOG_df, accessions, og_whitelist=None):\n",
    "    index = eggNOG_df.index\n",
    "\n",
    "    og_union = set()\n",
    "    for acc in accessions:\n",
    "        if acc not in index:\n",
    "            print(f'Not in index: {acc}')\n",
    "            continue\n",
    "\n",
    "        genme_ogs = eggNOG_df.loc[[acc]].set_index('protein_id')\n",
    "\n",
    "        ogs = set()\n",
    "        for protein_id in genme_ogs.index:\n",
    "            protein_ogs = genme_ogs.loc[protein_id, 'eggNOG_OGs']\n",
    "\n",
    "            candidate_ogs = {\n",
    "                og_with_tax.split('@')[0]\n",
    "                for og_with_tax in protein_ogs.split(',')\n",
    "            }\n",
    "\n",
    "            if og_whitelist is not None:\n",
    "                candidate_ogs = candidate_ogs & og_whitelist\n",
    "                \n",
    "            ogs |= candidate_ogs\n",
    "\n",
    "        og_union = og_union.union(ogs)\n",
    "\n",
    "    return og_union"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ogs_present_in_all(eggNOG_df, accessions, og_whitelist=None):\n",
    "    index = eggNOG_df.index\n",
    "\n",
    "    og_intersection = None\n",
    "    for acc in accessions:\n",
    "        if acc not in index:\n",
    "            print(f'Not in index: {acc}')\n",
    "            continue\n",
    "\n",
    "        genme_ogs = eggNOG_df.loc[[acc]].set_index('protein_id')\n",
    "\n",
    "        ogs = set()\n",
    "        for protein_id in genme_ogs.index:\n",
    "            protein_ogs = genme_ogs.loc[protein_id, 'eggNOG_OGs']\n",
    "\n",
    "            candidate_ogs = {\n",
    "                og_with_tax.split('@')[0]\n",
    "                for og_with_tax in protein_ogs.split(',')\n",
    "            }\n",
    "\n",
    "            if og_whitelist is not None:\n",
    "                candidate_ogs = candidate_ogs & og_whitelist\n",
    "                \n",
    "            ogs |= candidate_ogs\n",
    "\n",
    "        if og_intersection is None:\n",
    "            og_intersection = ogs\n",
    "        else:\n",
    "            og_intersection &= ogs\n",
    "\n",
    "    return og_intersection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### S-protein\n",
    "\n",
    " (OG: `4HHT0`)\n",
    "\n",
    "LysM-containing protein that is essential for immune evasion in Group A Streptococci. \n",
    "\n",
    "Refs: \n",
    "- [Wierzbicki et al., 2019](https://doi.org/10.1016/j.celrep.2019.11.001)\n",
    "- [Burnier et al., 2024](https://doi.org/10.1101/2024.11.08.622053)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This S-protein isn't present in _S. equi_ (group C Strep). Some _S. equi_ strains do however contain OG `4HPIN` (N-acetylmuramoyl-L-alanine amidase) along with a LysM domain.\n",
    "\n",
    "Intriguing... I wonder if the strains of _S. equi_ with a PGH have re-acquired a LysM-containing protein for the purpose of targeting other Streptococci?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is a list of the unique OGs of proteins with LysM in Streptococcus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "88"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes_with_lysM = sorted(set(pfam_df[pfam_df['hmm_query'] == 'LysM'].index))\n",
    "protein_ids_with_lysM = sorted(pfam_df[pfam_df['hmm_query'] == 'LysM']['id'].unique())\n",
    "\n",
    "ogs_list_str = eggNOG_df[eggNOG_df['protein_id'].isin(protein_ids_with_lysM)]['eggNOG_OGs'].unique()\n",
    "unique_ogs = sorted({\n",
    "    og_with_taxa.split('@')[0]\n",
    "    for og_list in ogs_list_str\n",
    "    for og_with_taxa in og_list.split(',')\n",
    "})\n",
    "len(unique_ogs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COG0741: lytic transglycosylase activity\n",
      "COG0860: N-acetylmuramoyl-L-alanine amidase activity\n",
      "COG1388: cell wall organization\n",
      "COG1652: LysM domain\n",
      "COG1705: Flagellar rod assembly protein muramidase FlgJ\n",
      "COG2304: hyaluronan metabolic process\n",
      "COG2340: Cysteine-rich secretory protein family\n",
      "COG3170: Tfp pilus assembly protein FimV\n",
      "COG3583: domain protein\n",
      "COG3757: hydrolase, family 25\n",
      "COG3942: CHAP domain\n",
      "COG5632: N-acetylmuramoyl-L-alanine amidase activity\n"
     ]
    }
   ],
   "source": [
    "unique_cogs = [og for og in unique_ogs if og.startswith('COG')]\n",
    "display_ogs(unique_cogs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WP_010921965.1@GCF_002055535.1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'COG1388@1|root,COG1388@2|Bacteria,1V61Q@1239|Firmicutes,4HHT0@91061|Bacilli'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acc = metadata_df[\n",
    "    (metadata_df['gtdb_species'] == 'Streptococcus pyogenes') &\n",
    "    (metadata_df['gtdb_representative'] == 't')\n",
    "].index[0]\n",
    "df = pfam_df.loc[acc]\n",
    "protein_ids = df[df['hmm_query'] == 'LysM']['id'].unique()\n",
    "assert len(protein_ids) == 1\n",
    "protein_id = protein_ids[0]\n",
    "\n",
    "print(protein_id)\n",
    "eggNOG_df[eggNOG_df['protein_id'] == protein_id]['eggNOG_OGs'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1V61Q: LysM domain protein\n",
      "COG1388: cell wall organization\n",
      "4HHT0: LysM domain protein\n"
     ]
    }
   ],
   "source": [
    "accs = sorted(metadata_df[\n",
    "    metadata_df['gtdb_species'].isin([\n",
    "        'Streptococcus pyogenes',\n",
    "        'Streptococcus dysgalactiae',\n",
    "    ])\n",
    "].index)\n",
    "df = pfam_df.loc[accs]\n",
    "protein_ids_with_lysM = sorted(df[df['hmm_query'] == 'LysM']['id'].unique())\n",
    "\n",
    "eggNOG_df_subset = eggNOG_df[eggNOG_df['protein_id'].isin(protein_ids_with_lysM)].copy()\n",
    "\n",
    "unique_ogs = get_ogs_present_in_all(eggNOG_df_subset, accs)\n",
    "\n",
    "display_ogs(unique_ogs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>total</th>\n",
       "      <th>percent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gtdb_species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Streptococcus pyogenes</th>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus agalactiae</th>\n",
       "      <td>196</td>\n",
       "      <td>197</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pneumoniae</th>\n",
       "      <td>156</td>\n",
       "      <td>158</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus suis</th>\n",
       "      <td>108</td>\n",
       "      <td>109</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus dysgalactiae</th>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus iniae</th>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>93.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus suis_W</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus gordonii</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus salivarius</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus thermophilus</th>\n",
       "      <td>9</td>\n",
       "      <td>88</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus suis_AA</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus oralis</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus suis_P</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus salivarius_D</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus canis</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus sobrinus</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus ruminantium</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus sp001556435</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus mitis</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus ilei</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            count  total  percent\n",
       "gtdb_species                                     \n",
       "Streptococcus pyogenes        271    271    100.0\n",
       "Streptococcus agalactiae      196    197     99.0\n",
       "Streptococcus pneumoniae      156    158     99.0\n",
       "Streptococcus suis            108    109     99.0\n",
       "Streptococcus dysgalactiae     43     43    100.0\n",
       "Streptococcus iniae            13     14     93.0\n",
       "Streptococcus suis_W           13     13    100.0\n",
       "Streptococcus gordonii         13     13    100.0\n",
       "Streptococcus salivarius       11     11    100.0\n",
       "Streptococcus thermophilus      9     88     10.0\n",
       "Streptococcus suis_AA           8      8    100.0\n",
       "Streptococcus oralis            7      7    100.0\n",
       "Streptococcus suis_P            7      7    100.0\n",
       "Streptococcus salivarius_D      7      7    100.0\n",
       "Streptococcus canis             6      6    100.0\n",
       "Streptococcus sobrinus          6      6    100.0\n",
       "Streptococcus ruminantium       5      5    100.0\n",
       "Streptococcus sp001556435       4      4    100.0\n",
       "Streptococcus mitis             4      4    100.0\n",
       "Streptococcus ilei              3      3    100.0"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accessions_with_4HHT0 = sorted(set(eggNOG_df[eggNOG_df['eggNOG_OGs'].str.contains('4HHT0@')].index))\n",
    "\n",
    "species_with_4HHT0 = pd.DataFrame(metadata_df.loc[accessions_with_4HHT0]['gtdb_species'].value_counts(), columns=['count'])\n",
    "species_with_4HHT0['total'] = [len(metadata_df[metadata_df['gtdb_species'] == s]) for s in species_with_4HHT0.index]\n",
    "species_with_4HHT0['percent'] = (100 * species_with_4HHT0['count'] / species_with_4HHT0['total']).round(0)\n",
    "species_with_4HHT0.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see what's different between _S. equi_ strains without LysM and other strains with this S-protein. We'll exclude the S. equi strains with LysM for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4092"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ogs_no_lysM = get_ogs_present_in_all(eggNOG_df, sorted(accessions_no_LysM))\n",
    "len(ogs_no_lysM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5660"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accessions_s_iniae_s_dysgalactiae = sorted(set(metadata_df[\n",
    "    (metadata_df['gtdb_species'] == 'Streptococcus iniae') |\n",
    "    (metadata_df['gtdb_species'] == 'Streptococcus dysgalactiae')\n",
    "].index))\n",
    "\n",
    "# og_yes_lysM = get_ogs_present_in_all(eggNOG_df, accessions_s_iniae_s_dysgalactiae)\n",
    "og_yes_lysM = get_ogs_present_in_all(eggNOG_df, accessions_yes_LysM)\n",
    "len(og_yes_lysM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gain: 28 OGs\n",
      "1M8ZB: ABC-type amino acid transport signal transduction systems, periplasmic component domain\n",
      "1M8ZC: Membrane protein involved in the export of O-antigen and teichoic acid\n",
      "1M91B: Flavin transferase that catalyzes the transfer of the FMN moiety of FAD and its covalent binding to the hydroxyl group of a threonine residue in a target flavoprotein\n",
      "1M96V: Periplasmic binding protein\n",
      "1M9GP: leucine-rich protein\n",
      "1M9KM: Specifically methylates the pseudouridine at position 1915 (m3Psi1915) in 23S rRNA\n",
      "1MA2F: Nicotinamide mononucleotide transporter\n",
      "1TQUG: ABC transporter\n",
      "1TR9C: Flavin transferase that catalyzes the transfer of the FMN moiety of FAD and its covalent binding to the hydroxyl group of a threonine residue in a target flavoprotein\n",
      "1UIJ9: ABC transporter\n",
      "1V21G: nicotinamide mononucleotide transporter\n",
      "1V3JM: Specifically methylates the pseudouridine at position 1915 (m3Psi1915) in 23S rRNA\n",
      "1V649: COG2508, regulator of polyketide synthase expression\n",
      "1W2RT: nan\n",
      "2DEGR: nan\n",
      "2ZMY1: nan\n",
      "4HACG: Membrane protein involved in the export of O-antigen and teichoic acid\n",
      "4HAJ5: ABC-type amino acid transport signal transduction systems periplasmic component domain\n",
      "4HDDF: ABC transporter, substratebinding protein\n",
      "4HDFE: Flavin transferase that catalyzes the transfer of the FMN moiety of FAD and its covalent binding to the hydroxyl group of a threonine residue in a target flavoprotein\n",
      "4HFP8: Specifically methylates the pseudouridine at position 1915 (m3Psi1915) in 23S rRNA\n",
      "4HHCD: COG2508, regulator of polyketide synthase expression\n",
      "4HZZS: nan\n",
      "4I6C8: Nicotinamide mononucleotide transporter\n",
      "COG1477: protein flavinylation\n",
      "COG1576: rRNA processing\n",
      "COG2508: Transcriptional regulator\n",
      "COG3201: nicotinamide riboside transmembrane transporter activity\n"
     ]
    }
   ],
   "source": [
    "missing_in_yes_lysM = ogs_no_lysM - og_yes_lysM\n",
    "print('Gain:', len(ogs_no_lysM - og_yes_lysM), 'OGs')\n",
    "display_ogs(sorted(missing_in_yes_lysM))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1596\n"
     ]
    }
   ],
   "source": [
    "missing_in_no_lysM = og_yes_lysM - ogs_no_lysM\n",
    "print(len(missing_in_no_lysM))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH28842.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_00080</td>\n",
       "      <td>6.156000e-301</td>\n",
       "      <td>925.0</td>\n",
       "      <td>COG2244@1|root,COG2244@2|Bacteria,1TNYX@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>Membrane protein involved in the export of O-a...</td>\n",
       "      <td>yabM</td>\n",
       "      <td>GO:0005575,GO:0005623,GO:0005886,GO:0016020,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K03328,ko:K06409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko02000</td>\n",
       "      <td>2.A.66.2,2.A.66.2.14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Polysacc_synt,Polysacc_synt_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    protein_id    seed_ortholog  \\\n",
       "assembly_accession                                                \n",
       "GCA_900637675.1     VEH28842.1@GCA_900637675.1  40041.SZO_00080   \n",
       "\n",
       "                           evalue  score  \\\n",
       "assembly_accession                         \n",
       "GCA_900637675.1     6.156000e-301  925.0   \n",
       "\n",
       "                                                           eggNOG_OGs  \\\n",
       "assembly_accession                                                      \n",
       "GCA_900637675.1     COG2244@1|root,COG2244@2|Bacteria,1TNYX@1239|F...   \n",
       "\n",
       "                   max_annot_lvl COG_category  \\\n",
       "assembly_accession                              \n",
       "GCA_900637675.1       2|Bacteria            S   \n",
       "\n",
       "                                                          Description  \\\n",
       "assembly_accession                                                      \n",
       "GCA_900637675.1     Membrane protein involved in the export of O-a...   \n",
       "\n",
       "                   Preferred_name  \\\n",
       "assembly_accession                  \n",
       "GCA_900637675.1              yabM   \n",
       "\n",
       "                                                                  GOs  ...  \\\n",
       "assembly_accession                                                     ...   \n",
       "GCA_900637675.1     GO:0005575,GO:0005623,GO:0005886,GO:0016020,GO...  ...   \n",
       "\n",
       "                                KEGG_ko KEGG_Pathway KEGG_Module  \\\n",
       "assembly_accession                                                 \n",
       "GCA_900637675.1     ko:K03328,ko:K06409          NaN         NaN   \n",
       "\n",
       "                   KEGG_Reaction KEGG_rclass            BRITE  \\\n",
       "assembly_accession                                              \n",
       "GCA_900637675.1              NaN         NaN  ko00000,ko02000   \n",
       "\n",
       "                                 KEGG_TC CAZy BiGG_Reaction  \\\n",
       "assembly_accession                                            \n",
       "GCA_900637675.1     2.A.66.2,2.A.66.2.14  NaN           NaN   \n",
       "\n",
       "                                            PFAMs  \n",
       "assembly_accession                                 \n",
       "GCA_900637675.1     Polysacc_synt,Polysacc_synt_C  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m = metadata_df.loc[sorted(accessions_no_LysM)]\n",
    "acc = m[m['gtdb_representative'] == 't'].index[0]\n",
    "\n",
    "df = eggNOG_df.loc[acc]\n",
    "df[df['eggNOG_OGs'].str.contains('1M8ZC')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gtdb_species\n",
       "Streptococcus dysgalactiae    43\n",
       "Streptococcus equi            37\n",
       "Streptococcus halichoeri       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = eggNOG_df[eggNOG_df['eggNOG_OGs'].str.contains('1M8ZC')]\n",
    "\n",
    "metadata_df.loc[sorted(set(df.index))]['gtdb_species'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Focus on cell wall related pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "461"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_pathways = [\n",
    "    'map00550',  # Peptidoglycan biosynthesis\n",
    "    'map00540',  # Lipopolysaccharide biosynthesis\n",
    "]\n",
    "\n",
    "COGs_only = False  # Focus on OGs with a COG ID (no internal eggNOG ones)\n",
    "\n",
    "cw_ogs = set()\n",
    "for kegg_pathway in kegg_pathways:\n",
    "    ogs_str = eggNOG_df[\n",
    "        eggNOG_df['KEGG_Pathway'].notnull() &\n",
    "        eggNOG_df['KEGG_Pathway'].str.contains(kegg_pathway)\n",
    "    ]['eggNOG_OGs'].values\n",
    "\n",
    "    if COGs_only:\n",
    "        cw_ogs |= {\n",
    "            og\n",
    "            for og_str in ogs_str\n",
    "            for og_raw in og_str.split(',')\n",
    "            if (og := og_raw.split('@')[0]).startswith('COG')\n",
    "        }\n",
    "    else:\n",
    "        cw_ogs |= {\n",
    "            og_raw.split('@')[0]\n",
    "            for og_str in ogs_str\n",
    "            for og_raw in og_str.split(',')\n",
    "        }\n",
    "\n",
    "len(cw_ogs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2190"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_ogs = {\n",
    "    og\n",
    "    for og_str in eggNOG_df['eggNOG_OGs']\n",
    "    for og_raw in og_str.split(',')\n",
    "    if (og := og_raw.split('@')[0]).startswith('COG')\n",
    "}\n",
    "\n",
    "len(cog_ogs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(860, 1230)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ogs_no_lysM = get_ogs_present_in_all(eggNOG_df, sorted(accessions_no_LysM), cw_ogs)\n",
    "# ogs_no_lysM_all = get_unique_ogs(eggNOG_df, sorted(accessions_no_LysM), cw_ogs)\n",
    "\n",
    "ogs_no_lysM = get_ogs_present_in_all(eggNOG_df, sorted(accessions_no_LysM), cog_ogs)\n",
    "ogs_no_lysM_all = get_unique_ogs(eggNOG_df, sorted(accessions_no_LysM), cog_ogs)\n",
    "\n",
    "len(ogs_no_lysM), len(ogs_no_lysM_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(853, 1440)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accessions = sorted(set(metadata_df[\n",
    "    (metadata_df['gtdb_species'] == 'Streptococcus iniae') |\n",
    "    (metadata_df['gtdb_species'] == 'Streptococcus pyogens') |\n",
    "    (metadata_df['gtdb_species'] == 'Streptococcus dysgalactiae')\n",
    "].index))\n",
    "\n",
    "df = eggNOG_df.loc[accessions]\n",
    "accessions = sorted(set(df[df['eggNOG_OGs'].str.contains('4HHT0')].index))\n",
    "\n",
    "# og_yes_lysM = get_ogs_present_in_all(eggNOG_df, accessions, cw_ogs)\n",
    "# og_yes_lysM_all = get_unique_ogs(eggNOG_df, accessions, cw_ogs)\n",
    "\n",
    "og_yes_lysM = get_ogs_present_in_all(eggNOG_df, accessions, cog_ogs)\n",
    "og_yes_lysM_all = get_unique_ogs(eggNOG_df, accessions, cog_ogs)\n",
    "\n",
    "len(og_yes_lysM), len(og_yes_lysM_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gain: 8 OGs\n",
      "COG1345: cell adhesion\n",
      "COG1376: peptidoglycan L,D-transpeptidase activity\n",
      "COG1434: Gram-negative-bacterium-type cell wall biogenesis\n",
      "COG3458: cephalosporin-C deacetylase activity\n",
      "COG4495: Domain of unknown function (DUF4176)\n",
      "COG4935: serine-type endopeptidase activity\n",
      "COG5353: nan\n",
      "COG5585: self proteolysis\n"
     ]
    }
   ],
   "source": [
    "missing_in_yes_lysM = ogs_no_lysM - og_yes_lysM_all\n",
    "print('Gain:', len(missing_in_yes_lysM), 'OGs')\n",
    "display_ogs(sorted(missing_in_yes_lysM))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH30075.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_17040</td>\n",
       "      <td>9.956000e-58</td>\n",
       "      <td>199.0</td>\n",
       "      <td>COG3458@1|root,COG3458@2|Bacteria,1TS02@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>Q</td>\n",
       "      <td>Acetyl xylan esterase (AXE1)</td>\n",
       "      <td>cah</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K01060</td>\n",
       "      <td>ko00311,ko01130,map00311,map01130</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R03062</td>\n",
       "      <td>RC00020,RC00041</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AXE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH32169.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_11360</td>\n",
       "      <td>3.874000e-315</td>\n",
       "      <td>964.0</td>\n",
       "      <td>COG4935@1|root,COG4935@2|Bacteria,1UI0J@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>O</td>\n",
       "      <td>CotH kinase protein</td>\n",
       "      <td>cotH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CotH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH31027.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_13680</td>\n",
       "      <td>3.983000e-162</td>\n",
       "      <td>510.0</td>\n",
       "      <td>COG5585@1|root,COG5585@2|Bacteria</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K11044</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko01000,ko02042</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADPrib_exo_Tox,PT-TG,Phage_min_cap2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH34322.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_17320</td>\n",
       "      <td>1.009000e-92</td>\n",
       "      <td>307.0</td>\n",
       "      <td>COG1345@1|root,COG1345@2|Bacteria,1URWC@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>N</td>\n",
       "      <td>LXG domain of WXG superfamily</td>\n",
       "      <td>fliD</td>\n",
       "      <td>GO:0001539,GO:0005575,GO:0005576,GO:0005623,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K02407</td>\n",
       "      <td>ko02040,map02040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko00001,ko02035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Flagellin_IN,FliD_C,FliD_N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH31778.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_12260</td>\n",
       "      <td>1.951000e-84</td>\n",
       "      <td>280.0</td>\n",
       "      <td>COG5353@1|root,COG5353@2|Bacteria</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>protein conserved in bacteria</td>\n",
       "      <td>ypmB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PepSY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH29995.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_17320</td>\n",
       "      <td>2.667000e-267</td>\n",
       "      <td>833.0</td>\n",
       "      <td>COG1345@1|root,COG1345@2|Bacteria,1URWC@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>N</td>\n",
       "      <td>LXG domain of WXG superfamily</td>\n",
       "      <td>fliD</td>\n",
       "      <td>GO:0001539,GO:0005575,GO:0005576,GO:0005623,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K02407</td>\n",
       "      <td>ko02040,map02040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko00001,ko02035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Flagellin_IN,FliD_C,FliD_N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH33832.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_08060</td>\n",
       "      <td>8.211000e-65</td>\n",
       "      <td>220.0</td>\n",
       "      <td>COG4495@1|root,COG4495@2|Bacteria,1VIE1@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>Domain of unknown function (DUF4176)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DUF4176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH29160.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_00810</td>\n",
       "      <td>1.507000e-57</td>\n",
       "      <td>199.0</td>\n",
       "      <td>COG4495@1|root,COG4495@2|Bacteria,1VFG4@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>Domain of unknown function (DUF4176)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DUF4176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH34822.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_04710</td>\n",
       "      <td>2.247000e-96</td>\n",
       "      <td>315.0</td>\n",
       "      <td>COG1376@1|root,COG1376@2|Bacteria,1V4KP@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>M</td>\n",
       "      <td>ErfK YbiS YcfS YnhG</td>\n",
       "      <td>yciB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YkuD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH34825.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_04710</td>\n",
       "      <td>1.537000e-04</td>\n",
       "      <td>44.0</td>\n",
       "      <td>COG1376@1|root,COG1376@2|Bacteria,1V4KP@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>M</td>\n",
       "      <td>ErfK YbiS YcfS YnhG</td>\n",
       "      <td>yciB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YkuD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH30168.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_16720</td>\n",
       "      <td>1.243000e-193</td>\n",
       "      <td>606.0</td>\n",
       "      <td>COG1434@1|root,COG1434@2|Bacteria,1UR34@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>DUF218 domain</td>\n",
       "      <td>XK27_05540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DUF218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH34344.1@GCA_900637675.1</td>\n",
       "      <td>638301.HMPREF0444_1676</td>\n",
       "      <td>6.320000e-90</td>\n",
       "      <td>302.0</td>\n",
       "      <td>COG1345@1|root,COG1345@2|Bacteria</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>N</td>\n",
       "      <td>Required for morphogenesis and for the elongat...</td>\n",
       "      <td>fliD</td>\n",
       "      <td>GO:0001539,GO:0005575,GO:0005576,GO:0005623,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K02407</td>\n",
       "      <td>ko02040,map02040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko00001,ko02035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Flagellin_IN,FliD_C,FliD_N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_900637675.1</th>\n",
       "      <td>VEH30825.1@GCA_900637675.1</td>\n",
       "      <td>40041.SZO_14620</td>\n",
       "      <td>1.314000e-63</td>\n",
       "      <td>217.0</td>\n",
       "      <td>COG4495@1|root,COG4495@2|Bacteria,1VFG4@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>Domain of unknown function (DUF4176)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DUF4176</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    protein_id           seed_ortholog  \\\n",
       "assembly_accession                                                       \n",
       "GCA_900637675.1     VEH30075.1@GCA_900637675.1         40041.SZO_17040   \n",
       "GCA_900637675.1     VEH32169.1@GCA_900637675.1         40041.SZO_11360   \n",
       "GCA_900637675.1     VEH31027.1@GCA_900637675.1         40041.SZO_13680   \n",
       "GCA_900637675.1     VEH34322.1@GCA_900637675.1         40041.SZO_17320   \n",
       "GCA_900637675.1     VEH31778.1@GCA_900637675.1         40041.SZO_12260   \n",
       "GCA_900637675.1     VEH29995.1@GCA_900637675.1         40041.SZO_17320   \n",
       "GCA_900637675.1     VEH33832.1@GCA_900637675.1         40041.SZO_08060   \n",
       "GCA_900637675.1     VEH29160.1@GCA_900637675.1         40041.SZO_00810   \n",
       "GCA_900637675.1     VEH34822.1@GCA_900637675.1         40041.SZO_04710   \n",
       "GCA_900637675.1     VEH34825.1@GCA_900637675.1         40041.SZO_04710   \n",
       "GCA_900637675.1     VEH30168.1@GCA_900637675.1         40041.SZO_16720   \n",
       "GCA_900637675.1     VEH34344.1@GCA_900637675.1  638301.HMPREF0444_1676   \n",
       "GCA_900637675.1     VEH30825.1@GCA_900637675.1         40041.SZO_14620   \n",
       "\n",
       "                           evalue  score  \\\n",
       "assembly_accession                         \n",
       "GCA_900637675.1      9.956000e-58  199.0   \n",
       "GCA_900637675.1     3.874000e-315  964.0   \n",
       "GCA_900637675.1     3.983000e-162  510.0   \n",
       "GCA_900637675.1      1.009000e-92  307.0   \n",
       "GCA_900637675.1      1.951000e-84  280.0   \n",
       "GCA_900637675.1     2.667000e-267  833.0   \n",
       "GCA_900637675.1      8.211000e-65  220.0   \n",
       "GCA_900637675.1      1.507000e-57  199.0   \n",
       "GCA_900637675.1      2.247000e-96  315.0   \n",
       "GCA_900637675.1      1.537000e-04   44.0   \n",
       "GCA_900637675.1     1.243000e-193  606.0   \n",
       "GCA_900637675.1      6.320000e-90  302.0   \n",
       "GCA_900637675.1      1.314000e-63  217.0   \n",
       "\n",
       "                                                           eggNOG_OGs  \\\n",
       "assembly_accession                                                      \n",
       "GCA_900637675.1     COG3458@1|root,COG3458@2|Bacteria,1TS02@1239|F...   \n",
       "GCA_900637675.1     COG4935@1|root,COG4935@2|Bacteria,1UI0J@1239|F...   \n",
       "GCA_900637675.1                     COG5585@1|root,COG5585@2|Bacteria   \n",
       "GCA_900637675.1     COG1345@1|root,COG1345@2|Bacteria,1URWC@1239|F...   \n",
       "GCA_900637675.1                     COG5353@1|root,COG5353@2|Bacteria   \n",
       "GCA_900637675.1     COG1345@1|root,COG1345@2|Bacteria,1URWC@1239|F...   \n",
       "GCA_900637675.1     COG4495@1|root,COG4495@2|Bacteria,1VIE1@1239|F...   \n",
       "GCA_900637675.1     COG4495@1|root,COG4495@2|Bacteria,1VFG4@1239|F...   \n",
       "GCA_900637675.1     COG1376@1|root,COG1376@2|Bacteria,1V4KP@1239|F...   \n",
       "GCA_900637675.1     COG1376@1|root,COG1376@2|Bacteria,1V4KP@1239|F...   \n",
       "GCA_900637675.1     COG1434@1|root,COG1434@2|Bacteria,1UR34@1239|F...   \n",
       "GCA_900637675.1                     COG1345@1|root,COG1345@2|Bacteria   \n",
       "GCA_900637675.1     COG4495@1|root,COG4495@2|Bacteria,1VFG4@1239|F...   \n",
       "\n",
       "                   max_annot_lvl COG_category  \\\n",
       "assembly_accession                              \n",
       "GCA_900637675.1       2|Bacteria            Q   \n",
       "GCA_900637675.1       2|Bacteria            O   \n",
       "GCA_900637675.1       2|Bacteria          NaN   \n",
       "GCA_900637675.1       2|Bacteria            N   \n",
       "GCA_900637675.1       2|Bacteria            S   \n",
       "GCA_900637675.1       2|Bacteria            N   \n",
       "GCA_900637675.1       2|Bacteria            S   \n",
       "GCA_900637675.1       2|Bacteria            S   \n",
       "GCA_900637675.1       2|Bacteria            M   \n",
       "GCA_900637675.1       2|Bacteria            M   \n",
       "GCA_900637675.1       2|Bacteria            S   \n",
       "GCA_900637675.1       2|Bacteria            N   \n",
       "GCA_900637675.1       2|Bacteria            S   \n",
       "\n",
       "                                                          Description  \\\n",
       "assembly_accession                                                      \n",
       "GCA_900637675.1                          Acetyl xylan esterase (AXE1)   \n",
       "GCA_900637675.1                                   CotH kinase protein   \n",
       "GCA_900637675.1                                                   NaN   \n",
       "GCA_900637675.1                         LXG domain of WXG superfamily   \n",
       "GCA_900637675.1                         protein conserved in bacteria   \n",
       "GCA_900637675.1                         LXG domain of WXG superfamily   \n",
       "GCA_900637675.1                  Domain of unknown function (DUF4176)   \n",
       "GCA_900637675.1                  Domain of unknown function (DUF4176)   \n",
       "GCA_900637675.1                                   ErfK YbiS YcfS YnhG   \n",
       "GCA_900637675.1                                   ErfK YbiS YcfS YnhG   \n",
       "GCA_900637675.1                                         DUF218 domain   \n",
       "GCA_900637675.1     Required for morphogenesis and for the elongat...   \n",
       "GCA_900637675.1                  Domain of unknown function (DUF4176)   \n",
       "\n",
       "                   Preferred_name  \\\n",
       "assembly_accession                  \n",
       "GCA_900637675.1               cah   \n",
       "GCA_900637675.1              cotH   \n",
       "GCA_900637675.1               NaN   \n",
       "GCA_900637675.1              fliD   \n",
       "GCA_900637675.1              ypmB   \n",
       "GCA_900637675.1              fliD   \n",
       "GCA_900637675.1               NaN   \n",
       "GCA_900637675.1               NaN   \n",
       "GCA_900637675.1              yciB   \n",
       "GCA_900637675.1              yciB   \n",
       "GCA_900637675.1        XK27_05540   \n",
       "GCA_900637675.1              fliD   \n",
       "GCA_900637675.1               NaN   \n",
       "\n",
       "                                                                  GOs  ...  \\\n",
       "assembly_accession                                                     ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "GCA_900637675.1     GO:0001539,GO:0005575,GO:0005576,GO:0005623,GO...  ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "GCA_900637675.1     GO:0001539,GO:0005575,GO:0005576,GO:0005623,GO...  ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "GCA_900637675.1     GO:0001539,GO:0005575,GO:0005576,GO:0005623,GO...  ...   \n",
       "GCA_900637675.1                                                   NaN  ...   \n",
       "\n",
       "                      KEGG_ko                       KEGG_Pathway KEGG_Module  \\\n",
       "assembly_accession                                                             \n",
       "GCA_900637675.1     ko:K01060  ko00311,ko01130,map00311,map01130         NaN   \n",
       "GCA_900637675.1           NaN                                NaN         NaN   \n",
       "GCA_900637675.1     ko:K11044                                NaN         NaN   \n",
       "GCA_900637675.1     ko:K02407                   ko02040,map02040         NaN   \n",
       "GCA_900637675.1           NaN                                NaN         NaN   \n",
       "GCA_900637675.1     ko:K02407                   ko02040,map02040         NaN   \n",
       "GCA_900637675.1           NaN                                NaN         NaN   \n",
       "GCA_900637675.1           NaN                                NaN         NaN   \n",
       "GCA_900637675.1           NaN                                NaN         NaN   \n",
       "GCA_900637675.1           NaN                                NaN         NaN   \n",
       "GCA_900637675.1           NaN                                NaN         NaN   \n",
       "GCA_900637675.1     ko:K02407                   ko02040,map02040         NaN   \n",
       "GCA_900637675.1           NaN                                NaN         NaN   \n",
       "\n",
       "                   KEGG_Reaction      KEGG_rclass                    BRITE  \\\n",
       "assembly_accession                                                           \n",
       "GCA_900637675.1           R03062  RC00020,RC00041  ko00000,ko00001,ko01000   \n",
       "GCA_900637675.1              NaN              NaN                      NaN   \n",
       "GCA_900637675.1              NaN              NaN  ko00000,ko01000,ko02042   \n",
       "GCA_900637675.1              NaN              NaN  ko00000,ko00001,ko02035   \n",
       "GCA_900637675.1              NaN              NaN                      NaN   \n",
       "GCA_900637675.1              NaN              NaN  ko00000,ko00001,ko02035   \n",
       "GCA_900637675.1              NaN              NaN                      NaN   \n",
       "GCA_900637675.1              NaN              NaN                      NaN   \n",
       "GCA_900637675.1              NaN              NaN                      NaN   \n",
       "GCA_900637675.1              NaN              NaN                      NaN   \n",
       "GCA_900637675.1              NaN              NaN                      NaN   \n",
       "GCA_900637675.1              NaN              NaN  ko00000,ko00001,ko02035   \n",
       "GCA_900637675.1              NaN              NaN                      NaN   \n",
       "\n",
       "                   KEGG_TC CAZy BiGG_Reaction  \\\n",
       "assembly_accession                              \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "GCA_900637675.1        NaN  NaN           NaN   \n",
       "\n",
       "                                                  PFAMs  \n",
       "assembly_accession                                       \n",
       "GCA_900637675.1                                    AXE1  \n",
       "GCA_900637675.1                                    CotH  \n",
       "GCA_900637675.1     ADPrib_exo_Tox,PT-TG,Phage_min_cap2  \n",
       "GCA_900637675.1              Flagellin_IN,FliD_C,FliD_N  \n",
       "GCA_900637675.1                                   PepSY  \n",
       "GCA_900637675.1              Flagellin_IN,FliD_C,FliD_N  \n",
       "GCA_900637675.1                                 DUF4176  \n",
       "GCA_900637675.1                                 DUF4176  \n",
       "GCA_900637675.1                                    YkuD  \n",
       "GCA_900637675.1                                    YkuD  \n",
       "GCA_900637675.1                                  DUF218  \n",
       "GCA_900637675.1              Flagellin_IN,FliD_C,FliD_N  \n",
       "GCA_900637675.1                                 DUF4176  \n",
       "\n",
       "[13 rows x 21 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m = metadata_df.loc[sorted(accessions_no_LysM)]\n",
    "acc = m[m['gtdb_representative'] == 't'].index[0]\n",
    "df = eggNOG_df.loc[acc]\n",
    "\n",
    "protein_ids = set()\n",
    "for og in sorted(missing_in_yes_lysM):\n",
    "    protein_ids |= set(df[df['eggNOG_OGs'].str.contains(og)]['protein_id'].unique())\n",
    "\n",
    "eggNOG_df[eggNOG_df['protein_id'].isin(sorted(protein_ids))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loss: 31 OGs\n",
      "COG0010: Belongs to the arginase family\n",
      "COG0390: UDP-glucose transmembrane transporter activity\n",
      "COG0411: L-isoleucine import across plasma membrane\n",
      "COG0415: DNA photolyase activity\n",
      "COG0559: Belongs to the binding-protein-dependent transport system permease family\n",
      "COG0574: phosphotransferase activity, paired acceptors\n",
      "COG0578: glycerol-3-phosphate dehydrogenase (quinone) activity\n",
      "COG0625: glutathione transferase activity\n",
      "COG0727: PFAM Uncharacterised protein family (UPF0153)\n",
      "COG0730: response to heat\n",
      "COG0813: purine-nucleoside phosphorylase activity\n",
      "COG1054: Rhodanese Homology Domain\n",
      "COG1228: amidohydrolase\n",
      "COG1252: NADH oxidation\n",
      "COG1271: oxidase subunit\n",
      "COG1575: vitamin K biosynthetic process\n",
      "COG1600: epoxyqueuosine reductase activity\n",
      "COG1620: lactate transmembrane transporter activity\n",
      "COG1765: OsmC-like protein\n",
      "COG1806: Bifunctional serine threonine kinase and phosphorylase involved in the regulation of the\n",
      "COG1937: Protein conserved in bacteria\n",
      "COG2072: flavoprotein involved in K transport\n",
      "COG2321: neutral zinc metallopeptidase\n",
      "COG2764: nan\n",
      "COG2987: Catalyzes the conversion of urocanate to 4-imidazolone- 5-propionate\n",
      "COG3275: phosphorelay sensor kinase activity\n",
      "COG3404: glutamate formimidoyltransferase activity\n",
      "COG3643: Formiminotransferase domain\n",
      "COG3758: protein conserved in bacteria\n",
      "COG3775: system Galactitol-specific IIC component\n",
      "COG4767: VanZ like family\n"
     ]
    }
   ],
   "source": [
    "missing_in_no_lysM = og_yes_lysM - ogs_no_lysM_all\n",
    "print('Loss:', len(missing_in_no_lysM), 'OGs')\n",
    "display_ogs(sorted(missing_in_no_lysM))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_010922738.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_01888</td>\n",
       "      <td>6.300000e-119</td>\n",
       "      <td>382.0</td>\n",
       "      <td>COG3758@1|root,COG3758@2|Bacteria,1VH52@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>HutD</td>\n",
       "      <td>hutD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K09975</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HutD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_015055933.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_00540</td>\n",
       "      <td>4.453000e-246</td>\n",
       "      <td>759.0</td>\n",
       "      <td>COG1600@1|root,COG1600@2|Bacteria,1TP6Q@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>C</td>\n",
       "      <td>Catalyzes the conversion of epoxyqueuosine (oQ...</td>\n",
       "      <td>queG</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0006139,GO:0006725,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K18979</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko01000,ko03016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DUF1730,Fer4_16,HEAT_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_010922741.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_01891</td>\n",
       "      <td>6.454000e-220</td>\n",
       "      <td>681.0</td>\n",
       "      <td>COG0010@1|root,COG0010@2|Bacteria,1TQR5@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>E</td>\n",
       "      <td>Catalyzes the conversion of N-formimidoyl-L-gl...</td>\n",
       "      <td>hutG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K01479</td>\n",
       "      <td>ko00340,ko01100,map00340,map01100</td>\n",
       "      <td>M00045</td>\n",
       "      <td>R02285</td>\n",
       "      <td>RC00221,RC00681</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Arginase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_029714211.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_01694</td>\n",
       "      <td>5.048000e-183</td>\n",
       "      <td>572.0</td>\n",
       "      <td>COG0730@1|root,COG0730@2|Bacteria,1V5K2@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>response to heat</td>\n",
       "      <td>XK27_00825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K07090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TauE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_010922563.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_01432</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1206.0</td>\n",
       "      <td>COG0578@1|root,COG0578@2|Bacteria,1TQJN@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>C</td>\n",
       "      <td>Belongs to the FAD-dependent glycerol-3-phosph...</td>\n",
       "      <td>glpD</td>\n",
       "      <td>GO:0005575,GO:0005622,GO:0005623,GO:0005737,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K00105,ko:K00111</td>\n",
       "      <td>ko00564,ko01110,map00564,map01110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R00846,R00848</td>\n",
       "      <td>RC00029</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DAO,DAO_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_010922187.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_00730</td>\n",
       "      <td>3.261000e-226</td>\n",
       "      <td>699.0</td>\n",
       "      <td>COG1054@1|root,COG1054@2|Bacteria,1TRG7@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>Belongs to the UPF0176 family</td>\n",
       "      <td>yceA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K07146</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Rhodanese,Rhodanese_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_029713996.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_00604</td>\n",
       "      <td>3.272000e-198</td>\n",
       "      <td>617.0</td>\n",
       "      <td>COG2321@1|root,COG2321@2|Bacteria,1TQKF@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>metalloprotease</td>\n",
       "      <td>ypfJ</td>\n",
       "      <td>GO:0005575,GO:0005576</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K07054</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zn_peptidase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_002990147.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_00708</td>\n",
       "      <td>2.844000e-151</td>\n",
       "      <td>477.0</td>\n",
       "      <td>COG0813@1|root,COG0813@2|Bacteria,1TQPG@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>F</td>\n",
       "      <td>purine nucleoside phosphorylase</td>\n",
       "      <td>deoD</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0004731,GO:0005488,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K00772,ko:K03784</td>\n",
       "      <td>ko00230,ko00240,ko00270,ko00760,ko01100,ko0111...</td>\n",
       "      <td>M00034</td>\n",
       "      <td>R01402,R01561,R01863,R01969,R02147,R02294,R022...</td>\n",
       "      <td>RC00033,RC00063,RC00122,RC02819</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>iAPECO1_1312.APECO1_1997,iB21_1397.B21_04226,i...</td>\n",
       "      <td>PNP_UDP_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_010922487.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_01278</td>\n",
       "      <td>9.881313e-324</td>\n",
       "      <td>988.0</td>\n",
       "      <td>COG0415@1|root,COG0415@2|Bacteria,1TQAK@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>L</td>\n",
       "      <td>Belongs to the DNA photolyase family</td>\n",
       "      <td>phr</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K01669</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko01000,ko03400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNA_photolyase,FAD_binding_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_010922735.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_01885</td>\n",
       "      <td>1.193000e-197</td>\n",
       "      <td>615.0</td>\n",
       "      <td>COG3643@1|root,COG3643@2|Bacteria,1TP5T@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>E</td>\n",
       "      <td>glutamate</td>\n",
       "      <td>ftcD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K00603</td>\n",
       "      <td>ko00340,ko00670,ko01100,map00340,map00670,map0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R02287,R03189</td>\n",
       "      <td>RC00165,RC00221,RC00223,RC00870</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FTCD,FTCD_N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_020750283.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_01453</td>\n",
       "      <td>3.992000e-306</td>\n",
       "      <td>938.0</td>\n",
       "      <td>COG3775@1|root,COG3775@2|Bacteria,1TQ10@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>G</td>\n",
       "      <td>system Galactitol-specific IIC component</td>\n",
       "      <td>gatC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K02775,ko:K20114</td>\n",
       "      <td>ko00052,ko01100,ko02060,map00052,map01100,map0...</td>\n",
       "      <td>M00279,M00807</td>\n",
       "      <td>R05570,R11171</td>\n",
       "      <td>RC00017,RC03206</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko02000</td>\n",
       "      <td>4.A.5,4.A.5.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIIC-GAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_011285756.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_01883</td>\n",
       "      <td>3.475000e-276</td>\n",
       "      <td>848.0</td>\n",
       "      <td>COG1228@1|root,COG1228@2|Bacteria,1TP2J@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>Q</td>\n",
       "      <td>Imidazolone-5-propionate hydrolase</td>\n",
       "      <td>hutI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K01468</td>\n",
       "      <td>ko00340,ko01100,map00340,map01100</td>\n",
       "      <td>M00045</td>\n",
       "      <td>R02288</td>\n",
       "      <td>RC00683</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>iYO844.BSU39370</td>\n",
       "      <td>Amidohydro_1,Amidohydro_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_010922734.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_01884</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1379.0</td>\n",
       "      <td>COG2987@1|root,COG2987@2|Bacteria,1TPZ9@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>E</td>\n",
       "      <td>Catalyzes the conversion of urocanate to 4-imi...</td>\n",
       "      <td>hutU</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0006082,GO:0006520,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K01712</td>\n",
       "      <td>ko00340,ko01100,map00340,map01100</td>\n",
       "      <td>M00045</td>\n",
       "      <td>R02914</td>\n",
       "      <td>RC00804</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>iECIAI39_1322.ECIAI39_0688</td>\n",
       "      <td>Urocanase,Urocanase_C,Urocanase_N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_002993519.1@GCF_002055535.1</td>\n",
       "      <td>482234.SCAZ3_07220</td>\n",
       "      <td>6.140000e-71</td>\n",
       "      <td>240.0</td>\n",
       "      <td>COG0625@1|root,COG0625@2|Bacteria,1V559@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>O</td>\n",
       "      <td>Glutathione S-transferase, N-terminal domain</td>\n",
       "      <td>gst</td>\n",
       "      <td>GO:0005575,GO:0005622,GO:0005623,GO:0005737,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K00799</td>\n",
       "      <td>ko00480,ko00980,ko00982,ko00983,ko01524,ko0520...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R03522,R07002,R07003,R07004,R07023,R07024,R070...</td>\n",
       "      <td>RC00004,RC00069,RC00840,RC00948,RC01704,RC0170...</td>\n",
       "      <td>ko00000,ko00001,ko01000,ko02000</td>\n",
       "      <td>1.A.12.2.2,1.A.12.3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GST_C,GST_C_3,GST_N,GST_N_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_002993028.1@GCF_002055535.1</td>\n",
       "      <td>1314.HKU360_00600</td>\n",
       "      <td>4.964000e-76</td>\n",
       "      <td>254.0</td>\n",
       "      <td>COG1765@1|root,COG1765@2|Bacteria</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>O</td>\n",
       "      <td>OsmC-like protein</td>\n",
       "      <td>ycaO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K06889,ko:K07397,ko:K09136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko03009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hydrolase_4,OsmC,YcaO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_002988560.1@GCF_002055535.1</td>\n",
       "      <td>759913.SDSE_1921</td>\n",
       "      <td>4.668000e-36</td>\n",
       "      <td>136.0</td>\n",
       "      <td>COG0574@1|root,COG1080@1|root,COG0574@2|Bacter...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>G</td>\n",
       "      <td>Belongs to the PEP-utilizing enzyme family</td>\n",
       "      <td>ppdK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K01006</td>\n",
       "      <td>ko00620,ko00710,ko00720,ko01100,ko01120,ko0120...</td>\n",
       "      <td>M00169,M00171,M00172,M00173</td>\n",
       "      <td>R00206</td>\n",
       "      <td>RC00002,RC00015</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PEP-utilizers,PEP-utilizers_C,PPDK_N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_002055535.1</th>\n",
       "      <td>WP_010922736.1@GCF_002055535.1</td>\n",
       "      <td>663952.SDD27957_10750</td>\n",
       "      <td>1.672000e-120</td>\n",
       "      <td>387.0</td>\n",
       "      <td>COG3404@1|root,COG3404@2|Bacteria,1V1X8@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>E</td>\n",
       "      <td>Formiminotransferase-cyclodeaminase</td>\n",
       "      <td>fchA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K01746</td>\n",
       "      <td>ko00670,ko01100,map00670,map01100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R02302</td>\n",
       "      <td>RC00688</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FTCD_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        protein_id          seed_ortholog  \\\n",
       "assembly_accession                                                          \n",
       "GCF_002055535.1     WP_010922738.1@GCF_002055535.1      1314.HKU360_01888   \n",
       "GCF_002055535.1     WP_015055933.1@GCF_002055535.1      1314.HKU360_00540   \n",
       "GCF_002055535.1     WP_010922741.1@GCF_002055535.1      1314.HKU360_01891   \n",
       "GCF_002055535.1     WP_029714211.1@GCF_002055535.1      1314.HKU360_01694   \n",
       "GCF_002055535.1     WP_010922563.1@GCF_002055535.1      1314.HKU360_01432   \n",
       "GCF_002055535.1     WP_010922187.1@GCF_002055535.1      1314.HKU360_00730   \n",
       "GCF_002055535.1     WP_029713996.1@GCF_002055535.1      1314.HKU360_00604   \n",
       "GCF_002055535.1     WP_002990147.1@GCF_002055535.1      1314.HKU360_00708   \n",
       "GCF_002055535.1     WP_010922487.1@GCF_002055535.1      1314.HKU360_01278   \n",
       "GCF_002055535.1     WP_010922735.1@GCF_002055535.1      1314.HKU360_01885   \n",
       "GCF_002055535.1     WP_020750283.1@GCF_002055535.1      1314.HKU360_01453   \n",
       "GCF_002055535.1     WP_011285756.1@GCF_002055535.1      1314.HKU360_01883   \n",
       "GCF_002055535.1     WP_010922734.1@GCF_002055535.1      1314.HKU360_01884   \n",
       "GCF_002055535.1     WP_002993519.1@GCF_002055535.1     482234.SCAZ3_07220   \n",
       "GCF_002055535.1     WP_002993028.1@GCF_002055535.1      1314.HKU360_00600   \n",
       "GCF_002055535.1     WP_002988560.1@GCF_002055535.1       759913.SDSE_1921   \n",
       "GCF_002055535.1     WP_010922736.1@GCF_002055535.1  663952.SDD27957_10750   \n",
       "\n",
       "                           evalue   score  \\\n",
       "assembly_accession                          \n",
       "GCF_002055535.1     6.300000e-119   382.0   \n",
       "GCF_002055535.1     4.453000e-246   759.0   \n",
       "GCF_002055535.1     6.454000e-220   681.0   \n",
       "GCF_002055535.1     5.048000e-183   572.0   \n",
       "GCF_002055535.1      0.000000e+00  1206.0   \n",
       "GCF_002055535.1     3.261000e-226   699.0   \n",
       "GCF_002055535.1     3.272000e-198   617.0   \n",
       "GCF_002055535.1     2.844000e-151   477.0   \n",
       "GCF_002055535.1     9.881313e-324   988.0   \n",
       "GCF_002055535.1     1.193000e-197   615.0   \n",
       "GCF_002055535.1     3.992000e-306   938.0   \n",
       "GCF_002055535.1     3.475000e-276   848.0   \n",
       "GCF_002055535.1      0.000000e+00  1379.0   \n",
       "GCF_002055535.1      6.140000e-71   240.0   \n",
       "GCF_002055535.1      4.964000e-76   254.0   \n",
       "GCF_002055535.1      4.668000e-36   136.0   \n",
       "GCF_002055535.1     1.672000e-120   387.0   \n",
       "\n",
       "                                                           eggNOG_OGs  \\\n",
       "assembly_accession                                                      \n",
       "GCF_002055535.1     COG3758@1|root,COG3758@2|Bacteria,1VH52@1239|F...   \n",
       "GCF_002055535.1     COG1600@1|root,COG1600@2|Bacteria,1TP6Q@1239|F...   \n",
       "GCF_002055535.1     COG0010@1|root,COG0010@2|Bacteria,1TQR5@1239|F...   \n",
       "GCF_002055535.1     COG0730@1|root,COG0730@2|Bacteria,1V5K2@1239|F...   \n",
       "GCF_002055535.1     COG0578@1|root,COG0578@2|Bacteria,1TQJN@1239|F...   \n",
       "GCF_002055535.1     COG1054@1|root,COG1054@2|Bacteria,1TRG7@1239|F...   \n",
       "GCF_002055535.1     COG2321@1|root,COG2321@2|Bacteria,1TQKF@1239|F...   \n",
       "GCF_002055535.1     COG0813@1|root,COG0813@2|Bacteria,1TQPG@1239|F...   \n",
       "GCF_002055535.1     COG0415@1|root,COG0415@2|Bacteria,1TQAK@1239|F...   \n",
       "GCF_002055535.1     COG3643@1|root,COG3643@2|Bacteria,1TP5T@1239|F...   \n",
       "GCF_002055535.1     COG3775@1|root,COG3775@2|Bacteria,1TQ10@1239|F...   \n",
       "GCF_002055535.1     COG1228@1|root,COG1228@2|Bacteria,1TP2J@1239|F...   \n",
       "GCF_002055535.1     COG2987@1|root,COG2987@2|Bacteria,1TPZ9@1239|F...   \n",
       "GCF_002055535.1     COG0625@1|root,COG0625@2|Bacteria,1V559@1239|F...   \n",
       "GCF_002055535.1                     COG1765@1|root,COG1765@2|Bacteria   \n",
       "GCF_002055535.1     COG0574@1|root,COG1080@1|root,COG0574@2|Bacter...   \n",
       "GCF_002055535.1     COG3404@1|root,COG3404@2|Bacteria,1V1X8@1239|F...   \n",
       "\n",
       "                   max_annot_lvl COG_category  \\\n",
       "assembly_accession                              \n",
       "GCF_002055535.1       2|Bacteria            S   \n",
       "GCF_002055535.1       2|Bacteria            C   \n",
       "GCF_002055535.1       2|Bacteria            E   \n",
       "GCF_002055535.1       2|Bacteria            S   \n",
       "GCF_002055535.1       2|Bacteria            C   \n",
       "GCF_002055535.1       2|Bacteria            S   \n",
       "GCF_002055535.1       2|Bacteria            S   \n",
       "GCF_002055535.1       2|Bacteria            F   \n",
       "GCF_002055535.1       2|Bacteria            L   \n",
       "GCF_002055535.1       2|Bacteria            E   \n",
       "GCF_002055535.1       2|Bacteria            G   \n",
       "GCF_002055535.1       2|Bacteria            Q   \n",
       "GCF_002055535.1       2|Bacteria            E   \n",
       "GCF_002055535.1       2|Bacteria            O   \n",
       "GCF_002055535.1       2|Bacteria            O   \n",
       "GCF_002055535.1       2|Bacteria            G   \n",
       "GCF_002055535.1       2|Bacteria            E   \n",
       "\n",
       "                                                          Description  \\\n",
       "assembly_accession                                                      \n",
       "GCF_002055535.1                                                  HutD   \n",
       "GCF_002055535.1     Catalyzes the conversion of epoxyqueuosine (oQ...   \n",
       "GCF_002055535.1     Catalyzes the conversion of N-formimidoyl-L-gl...   \n",
       "GCF_002055535.1                                      response to heat   \n",
       "GCF_002055535.1     Belongs to the FAD-dependent glycerol-3-phosph...   \n",
       "GCF_002055535.1                         Belongs to the UPF0176 family   \n",
       "GCF_002055535.1                                       metalloprotease   \n",
       "GCF_002055535.1                       purine nucleoside phosphorylase   \n",
       "GCF_002055535.1                  Belongs to the DNA photolyase family   \n",
       "GCF_002055535.1                                             glutamate   \n",
       "GCF_002055535.1              system Galactitol-specific IIC component   \n",
       "GCF_002055535.1                    Imidazolone-5-propionate hydrolase   \n",
       "GCF_002055535.1     Catalyzes the conversion of urocanate to 4-imi...   \n",
       "GCF_002055535.1          Glutathione S-transferase, N-terminal domain   \n",
       "GCF_002055535.1                                     OsmC-like protein   \n",
       "GCF_002055535.1            Belongs to the PEP-utilizing enzyme family   \n",
       "GCF_002055535.1                   Formiminotransferase-cyclodeaminase   \n",
       "\n",
       "                   Preferred_name  \\\n",
       "assembly_accession                  \n",
       "GCF_002055535.1              hutD   \n",
       "GCF_002055535.1              queG   \n",
       "GCF_002055535.1              hutG   \n",
       "GCF_002055535.1        XK27_00825   \n",
       "GCF_002055535.1              glpD   \n",
       "GCF_002055535.1              yceA   \n",
       "GCF_002055535.1              ypfJ   \n",
       "GCF_002055535.1              deoD   \n",
       "GCF_002055535.1               phr   \n",
       "GCF_002055535.1              ftcD   \n",
       "GCF_002055535.1              gatC   \n",
       "GCF_002055535.1              hutI   \n",
       "GCF_002055535.1              hutU   \n",
       "GCF_002055535.1               gst   \n",
       "GCF_002055535.1              ycaO   \n",
       "GCF_002055535.1              ppdK   \n",
       "GCF_002055535.1              fchA   \n",
       "\n",
       "                                                                  GOs  ...  \\\n",
       "assembly_accession                                                     ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1     GO:0003674,GO:0003824,GO:0006139,GO:0006725,GO...  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1     GO:0005575,GO:0005622,GO:0005623,GO:0005737,GO...  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1                                 GO:0005575,GO:0005576  ...   \n",
       "GCF_002055535.1     GO:0003674,GO:0003824,GO:0004731,GO:0005488,GO...  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1     GO:0003674,GO:0003824,GO:0006082,GO:0006520,GO...  ...   \n",
       "GCF_002055535.1     GO:0005575,GO:0005622,GO:0005623,GO:0005737,GO...  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "GCF_002055535.1                                                   NaN  ...   \n",
       "\n",
       "                                          KEGG_ko  \\\n",
       "assembly_accession                                  \n",
       "GCF_002055535.1                         ko:K09975   \n",
       "GCF_002055535.1                         ko:K18979   \n",
       "GCF_002055535.1                         ko:K01479   \n",
       "GCF_002055535.1                         ko:K07090   \n",
       "GCF_002055535.1               ko:K00105,ko:K00111   \n",
       "GCF_002055535.1                         ko:K07146   \n",
       "GCF_002055535.1                         ko:K07054   \n",
       "GCF_002055535.1               ko:K00772,ko:K03784   \n",
       "GCF_002055535.1                         ko:K01669   \n",
       "GCF_002055535.1                         ko:K00603   \n",
       "GCF_002055535.1               ko:K02775,ko:K20114   \n",
       "GCF_002055535.1                         ko:K01468   \n",
       "GCF_002055535.1                         ko:K01712   \n",
       "GCF_002055535.1                         ko:K00799   \n",
       "GCF_002055535.1     ko:K06889,ko:K07397,ko:K09136   \n",
       "GCF_002055535.1                         ko:K01006   \n",
       "GCF_002055535.1                         ko:K01746   \n",
       "\n",
       "                                                         KEGG_Pathway  \\\n",
       "assembly_accession                                                      \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                     ko00340,ko01100,map00340,map01100   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                     ko00564,ko01110,map00564,map01110   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1     ko00230,ko00240,ko00270,ko00760,ko01100,ko0111...   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1     ko00340,ko00670,ko01100,map00340,map00670,map0...   \n",
       "GCF_002055535.1     ko00052,ko01100,ko02060,map00052,map01100,map0...   \n",
       "GCF_002055535.1                     ko00340,ko01100,map00340,map01100   \n",
       "GCF_002055535.1                     ko00340,ko01100,map00340,map01100   \n",
       "GCF_002055535.1     ko00480,ko00980,ko00982,ko00983,ko01524,ko0520...   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1     ko00620,ko00710,ko00720,ko01100,ko01120,ko0120...   \n",
       "GCF_002055535.1                     ko00670,ko01100,map00670,map01100   \n",
       "\n",
       "                                    KEGG_Module  \\\n",
       "assembly_accession                                \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1                          M00045   \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1                          M00034   \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1                   M00279,M00807   \n",
       "GCF_002055535.1                          M00045   \n",
       "GCF_002055535.1                          M00045   \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1                             NaN   \n",
       "GCF_002055535.1     M00169,M00171,M00172,M00173   \n",
       "GCF_002055535.1                             NaN   \n",
       "\n",
       "                                                        KEGG_Reaction  \\\n",
       "assembly_accession                                                      \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                                R02285   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                         R00846,R00848   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1     R01402,R01561,R01863,R01969,R02147,R02294,R022...   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                         R02287,R03189   \n",
       "GCF_002055535.1                                         R05570,R11171   \n",
       "GCF_002055535.1                                                R02288   \n",
       "GCF_002055535.1                                                R02914   \n",
       "GCF_002055535.1     R03522,R07002,R07003,R07004,R07023,R07024,R070...   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                                R00206   \n",
       "GCF_002055535.1                                                R02302   \n",
       "\n",
       "                                                          KEGG_rclass  \\\n",
       "assembly_accession                                                      \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                       RC00221,RC00681   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                               RC00029   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                       RC00033,RC00063,RC00122,RC02819   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                       RC00165,RC00221,RC00223,RC00870   \n",
       "GCF_002055535.1                                       RC00017,RC03206   \n",
       "GCF_002055535.1                                               RC00683   \n",
       "GCF_002055535.1                                               RC00804   \n",
       "GCF_002055535.1     RC00004,RC00069,RC00840,RC00948,RC01704,RC0170...   \n",
       "GCF_002055535.1                                                   NaN   \n",
       "GCF_002055535.1                                       RC00002,RC00015   \n",
       "GCF_002055535.1                                               RC00688   \n",
       "\n",
       "                                              BRITE                KEGG_TC  \\\n",
       "assembly_accession                                                           \n",
       "GCF_002055535.1                             ko00000                    NaN   \n",
       "GCF_002055535.1             ko00000,ko01000,ko03016                    NaN   \n",
       "GCF_002055535.1     ko00000,ko00001,ko00002,ko01000                    NaN   \n",
       "GCF_002055535.1                             ko00000                    NaN   \n",
       "GCF_002055535.1             ko00000,ko00001,ko01000                    NaN   \n",
       "GCF_002055535.1                             ko00000                    NaN   \n",
       "GCF_002055535.1                             ko00000                    NaN   \n",
       "GCF_002055535.1     ko00000,ko00001,ko00002,ko01000                    NaN   \n",
       "GCF_002055535.1             ko00000,ko01000,ko03400                    NaN   \n",
       "GCF_002055535.1             ko00000,ko00001,ko01000                    NaN   \n",
       "GCF_002055535.1     ko00000,ko00001,ko00002,ko02000          4.A.5,4.A.5.1   \n",
       "GCF_002055535.1     ko00000,ko00001,ko00002,ko01000                    NaN   \n",
       "GCF_002055535.1     ko00000,ko00001,ko00002,ko01000                    NaN   \n",
       "GCF_002055535.1     ko00000,ko00001,ko01000,ko02000  1.A.12.2.2,1.A.12.3.2   \n",
       "GCF_002055535.1                     ko00000,ko03009                    NaN   \n",
       "GCF_002055535.1     ko00000,ko00001,ko00002,ko01000                    NaN   \n",
       "GCF_002055535.1             ko00000,ko00001,ko01000                    NaN   \n",
       "\n",
       "                   CAZy                                      BiGG_Reaction  \\\n",
       "assembly_accession                                                           \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN  iAPECO1_1312.APECO1_1997,iB21_1397.B21_04226,i...   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                    iYO844.BSU39370   \n",
       "GCF_002055535.1     NaN                         iECIAI39_1322.ECIAI39_0688   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "GCF_002055535.1     NaN                                                NaN   \n",
       "\n",
       "                                                   PFAMs  \n",
       "assembly_accession                                        \n",
       "GCF_002055535.1                                     HutD  \n",
       "GCF_002055535.1                   DUF1730,Fer4_16,HEAT_2  \n",
       "GCF_002055535.1                                 Arginase  \n",
       "GCF_002055535.1                                     TauE  \n",
       "GCF_002055535.1                                DAO,DAO_C  \n",
       "GCF_002055535.1                    Rhodanese,Rhodanese_C  \n",
       "GCF_002055535.1                             Zn_peptidase  \n",
       "GCF_002055535.1                                PNP_UDP_1  \n",
       "GCF_002055535.1             DNA_photolyase,FAD_binding_7  \n",
       "GCF_002055535.1                              FTCD,FTCD_N  \n",
       "GCF_002055535.1                                 EIIC-GAT  \n",
       "GCF_002055535.1                Amidohydro_1,Amidohydro_3  \n",
       "GCF_002055535.1        Urocanase,Urocanase_C,Urocanase_N  \n",
       "GCF_002055535.1              GST_C,GST_C_3,GST_N,GST_N_3  \n",
       "GCF_002055535.1                    Hydrolase_4,OsmC,YcaO  \n",
       "GCF_002055535.1     PEP-utilizers,PEP-utilizers_C,PPDK_N  \n",
       "GCF_002055535.1                                   FTCD_C  \n",
       "\n",
       "[17 rows x 21 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acc = metadata_df[(metadata_df['gtdb_species'] == 'Streptococcus pyogenes') & (metadata_df['gtdb_representative'] == 't')].index[0]\n",
    "df = eggNOG_df.loc[acc]\n",
    "\n",
    "protein_ids = set()\n",
    "for og in sorted(missing_in_no_lysM):\n",
    "    protein_ids |= set(df[df['eggNOG_OGs'].str.contains(og)]['protein_id'].unique())\n",
    "\n",
    "eggNOG_df[eggNOG_df['protein_id'].isin(sorted(protein_ids))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### YukD in S. equi with a LysM protein (but no S-protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'GCF_000219765.1', 'GCF_009676645.1', 'GCF_009676685.2', 'GCF_900636805.1'},\n",
       " {'GCF_000219765.1', 'GCF_009676645.1', 'GCF_009676685.2', 'GCF_900636805.1'})"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = eggNOG_df.loc[sorted(accessions_yes_LysM)]\n",
    "set(df[df['eggNOG_OGs'].str.contains('COG1376')].index), accessions_yes_LysM\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "They all have YukD â€“ suggesting that their cell wall composition is the same as other S. equi without LysM proteins.\n",
    "\n",
    "The LysM + Amidase they encode might be a weapon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>hmm_accession</th>\n",
       "      <th>hmm_query</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gtdb_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_009676685.2</th>\n",
       "      <td>WP_014622634.1@GCF_009676685.2</td>\n",
       "      <td>WP_014622634.1</td>\n",
       "      <td>PF01476.25</td>\n",
       "      <td>LysM</td>\n",
       "      <td>5.400000e-14</td>\n",
       "      <td>48.8</td>\n",
       "      <td>0.99</td>\n",
       "      <td>270</td>\n",
       "      <td>314</td>\n",
       "      <td>Streptococcus equi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_900636805.1</th>\n",
       "      <td>WP_154803714.1@GCF_900636805.1</td>\n",
       "      <td>WP_154803714.1</td>\n",
       "      <td>PF01476.25</td>\n",
       "      <td>LysM</td>\n",
       "      <td>2.400000e-15</td>\n",
       "      <td>53.1</td>\n",
       "      <td>0.98</td>\n",
       "      <td>269</td>\n",
       "      <td>313</td>\n",
       "      <td>Streptococcus equi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000219765.1</th>\n",
       "      <td>WP_014622634.1@GCF_000219765.1</td>\n",
       "      <td>WP_014622634.1</td>\n",
       "      <td>PF01476.25</td>\n",
       "      <td>LysM</td>\n",
       "      <td>5.300000e-14</td>\n",
       "      <td>48.8</td>\n",
       "      <td>0.99</td>\n",
       "      <td>270</td>\n",
       "      <td>314</td>\n",
       "      <td>Streptococcus equi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_009676645.1</th>\n",
       "      <td>WP_014622634.1@GCF_009676645.1</td>\n",
       "      <td>WP_014622634.1</td>\n",
       "      <td>PF01476.25</td>\n",
       "      <td>LysM</td>\n",
       "      <td>5.400000e-14</td>\n",
       "      <td>48.8</td>\n",
       "      <td>0.99</td>\n",
       "      <td>270</td>\n",
       "      <td>314</td>\n",
       "      <td>Streptococcus equi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                id      protein_id  \\\n",
       "assembly_accession                                                   \n",
       "GCF_009676685.2     WP_014622634.1@GCF_009676685.2  WP_014622634.1   \n",
       "GCF_900636805.1     WP_154803714.1@GCF_900636805.1  WP_154803714.1   \n",
       "GCF_000219765.1     WP_014622634.1@GCF_000219765.1  WP_014622634.1   \n",
       "GCF_009676645.1     WP_014622634.1@GCF_009676645.1  WP_014622634.1   \n",
       "\n",
       "                   hmm_accession hmm_query        evalue  bitscore  accuracy  \\\n",
       "assembly_accession                                                             \n",
       "GCF_009676685.2       PF01476.25      LysM  5.400000e-14      48.8      0.99   \n",
       "GCF_900636805.1       PF01476.25      LysM  2.400000e-15      53.1      0.98   \n",
       "GCF_000219765.1       PF01476.25      LysM  5.300000e-14      48.8      0.99   \n",
       "GCF_009676645.1       PF01476.25      LysM  5.400000e-14      48.8      0.99   \n",
       "\n",
       "                    start  end        gtdb_species  \n",
       "assembly_accession                                  \n",
       "GCF_009676685.2       270  314  Streptococcus equi  \n",
       "GCF_900636805.1       269  313  Streptococcus equi  \n",
       "GCF_000219765.1       270  314  Streptococcus equi  \n",
       "GCF_009676645.1       270  314  Streptococcus equi  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfam_s_equi[pfam_s_equi['hmm_query'] == 'LysM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cwb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
