{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Focus on genus Streptococcus\n",
    "\n",
    "We found differential presence of `LysM` in 9 out of 10 representative genomes of Streptococci species, i.e. 1 species, _Streptococcus equi_, doesn't encode LysM in its otherwise complete genome. Does it mean that LysM does not bind the cell wall of _Streptococcus equi_? We don't know, but the omission is certainly intriguing.\n",
    "\n",
    "Our phylogenetically balanced dataset `db_proka` includes a maximum of 10 species per genus. To continue exploring the patterns of presence or absence of LysM and other CWB domains, this notebook is looking to gather a list of all Streptococci with complete genome in GTDB (release 220)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as plt\n",
    "import seaborn as sns\n",
    "from Bio import Phylo\n",
    "\n",
    "cwd = os.getcwd()\n",
    "if cwd.endswith('notebook'):\n",
    "    os.chdir('..')\n",
    "    cwd = os.getcwd()\n",
    "\n",
    "from src.cell_wall_binding_domains import cwb_domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme(palette='colorblind', font_scale=1.3)\n",
    "palette = sns.color_palette().as_hex()\n",
    "\n",
    "data_folder = Path('./data/')\n",
    "assert data_folder.is_dir()\n",
    "\n",
    "db_proka = Path('../db_proka/')\n",
    "assert db_proka.is_dir()\n",
    "\n",
    "gtdb_folder = Path('../data/gtdb_r220/')\n",
    "assert gtdb_folder.is_dir()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load GTDB r220 bacterial metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genomes: 584,382\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>ambiguous_bases</th>\n",
       "      <th>checkm2_completeness</th>\n",
       "      <th>checkm2_contamination</th>\n",
       "      <th>checkm2_model</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>checkm_marker_count</th>\n",
       "      <th>checkm_marker_lineage</th>\n",
       "      <th>checkm_marker_set_count</th>\n",
       "      <th>...</th>\n",
       "      <th>trna_aa_count</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>trna_selenocysteine_count</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000657795.2</th>\n",
       "      <td>RS_GCF_000657795.2</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.14</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.53</td>\n",
       "      <td>0.00</td>\n",
       "      <td>426</td>\n",
       "      <td>o__Burkholderiales (UID4000)</td>\n",
       "      <td>213</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Burkholderiaceae</td>\n",
       "      <td>Bordetella</td>\n",
       "      <td>Bordetella pseudohinzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001072555.1</th>\n",
       "      <td>RS_GCF_001072555.1</td>\n",
       "      <td>7</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.52</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.81</td>\n",
       "      <td>0.09</td>\n",
       "      <td>773</td>\n",
       "      <td>g__Staphylococcus (UID294)</td>\n",
       "      <td>178</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcus epidermidis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003050715.1</th>\n",
       "      <td>RS_GCF_003050715.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.04</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.60</td>\n",
       "      <td>0.22</td>\n",
       "      <td>769</td>\n",
       "      <td>g__Burkholderia (UID4006)</td>\n",
       "      <td>248</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Burkholderiaceae</td>\n",
       "      <td>Paraburkholderia</td>\n",
       "      <td>Paraburkholderia graminis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_016772635.1</th>\n",
       "      <td>RS_GCF_016772635.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.16</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.04</td>\n",
       "      <td>1169</td>\n",
       "      <td>f__Enterobacteriaceae (UID5139)</td>\n",
       "      <td>340</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>86</td>\n",
       "      <td>1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000615405.1</th>\n",
       "      <td>GB_GCA_000615405.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.37</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.62</td>\n",
       "      <td>0.57</td>\n",
       "      <td>471</td>\n",
       "      <td>o__Lactobacillales (UID543)</td>\n",
       "      <td>264</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Lactococcus</td>\n",
       "      <td>Lactococcus lactis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             accession  ambiguous_bases  checkm2_completeness  \\\n",
       "assembly_accession                                                              \n",
       "GCF_000657795.2     RS_GCF_000657795.2                0                 100.0   \n",
       "GCF_001072555.1     RS_GCF_001072555.1                7                 100.0   \n",
       "GCF_003050715.1     RS_GCF_003050715.1                0                 100.0   \n",
       "GCF_016772635.1     RS_GCF_016772635.1                0                 100.0   \n",
       "GCA_000615405.1     GB_GCA_000615405.1                0                 100.0   \n",
       "\n",
       "                    checkm2_contamination checkm2_model  checkm_completeness  \\\n",
       "assembly_accession                                                             \n",
       "GCF_000657795.2                      0.14      Specific                99.53   \n",
       "GCF_001072555.1                      0.52      Specific                99.81   \n",
       "GCF_003050715.1                      0.04      Specific                99.60   \n",
       "GCF_016772635.1                      0.16      Specific               100.00   \n",
       "GCA_000615405.1                      1.37      Specific                99.62   \n",
       "\n",
       "                    checkm_contamination  checkm_marker_count  \\\n",
       "assembly_accession                                              \n",
       "GCF_000657795.2                     0.00                  426   \n",
       "GCF_001072555.1                     0.09                  773   \n",
       "GCF_003050715.1                     0.22                  769   \n",
       "GCF_016772635.1                     0.04                 1169   \n",
       "GCA_000615405.1                     0.57                  471   \n",
       "\n",
       "                              checkm_marker_lineage  checkm_marker_set_count  \\\n",
       "assembly_accession                                                             \n",
       "GCF_000657795.2        o__Burkholderiales (UID4000)                      213   \n",
       "GCF_001072555.1          g__Staphylococcus (UID294)                      178   \n",
       "GCF_003050715.1           g__Burkholderia (UID4006)                      248   \n",
       "GCF_016772635.1     f__Enterobacteriaceae (UID5139)                      340   \n",
       "GCA_000615405.1         o__Lactobacillales (UID543)                      264   \n",
       "\n",
       "                    ...  trna_aa_count  trna_count  trna_selenocysteine_count  \\\n",
       "assembly_accession  ...                                                         \n",
       "GCF_000657795.2     ...             19          55                          1   \n",
       "GCF_001072555.1     ...             14          36                          0   \n",
       "GCF_003050715.1     ...             19          52                          0   \n",
       "GCF_016772635.1     ...             19          86                          1   \n",
       "GCA_000615405.1     ...             17          45                          0   \n",
       "\n",
       "                      domain     gtdb_phylum           gtdb_class  \\\n",
       "assembly_accession                                                  \n",
       "GCF_000657795.2     Bacteria  Pseudomonadota  Gammaproteobacteria   \n",
       "GCF_001072555.1     Bacteria       Bacillota              Bacilli   \n",
       "GCF_003050715.1     Bacteria  Pseudomonadota  Gammaproteobacteria   \n",
       "GCF_016772635.1     Bacteria  Pseudomonadota  Gammaproteobacteria   \n",
       "GCA_000615405.1     Bacteria       Bacillota              Bacilli   \n",
       "\n",
       "                          gtdb_order         gtdb_family        gtdb_genus  \\\n",
       "assembly_accession                                                           \n",
       "GCF_000657795.2      Burkholderiales    Burkholderiaceae        Bordetella   \n",
       "GCF_001072555.1     Staphylococcales   Staphylococcaceae    Staphylococcus   \n",
       "GCF_003050715.1      Burkholderiales    Burkholderiaceae  Paraburkholderia   \n",
       "GCF_016772635.1     Enterobacterales  Enterobacteriaceae        Salmonella   \n",
       "GCA_000615405.1      Lactobacillales    Streptococcaceae       Lactococcus   \n",
       "\n",
       "                                  gtdb_species  \n",
       "assembly_accession                              \n",
       "GCF_000657795.2        Bordetella pseudohinzii  \n",
       "GCF_001072555.1     Staphylococcus epidermidis  \n",
       "GCF_003050715.1      Paraburkholderia graminis  \n",
       "GCF_016772635.1            Salmonella enterica  \n",
       "GCA_000615405.1             Lactococcus lactis  \n",
       "\n",
       "[5 rows x 120 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bac_metadata = pd.read_csv(\n",
    "    gtdb_folder / 'bac120_metadata_r220.tsv.gz', \n",
    "    sep='\\t',\n",
    ")\n",
    "bac_metadata['assembly_accession'] = [a[3:] for a in bac_metadata['accession'].values]\n",
    "\n",
    "bac_metadata['domain'] = bac_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[0].replace('d__', ''))\n",
    "bac_metadata['gtdb_phylum'] = bac_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[1].replace('p__', ''))\n",
    "bac_metadata['gtdb_class'] = bac_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[2].replace('c__', ''))\n",
    "bac_metadata['gtdb_order'] = bac_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[3].replace('o__', ''))\n",
    "bac_metadata['gtdb_family'] = bac_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[4].replace('f__', ''))\n",
    "bac_metadata['gtdb_genus'] = bac_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[5].replace('g__', ''))\n",
    "bac_metadata['gtdb_species'] = bac_metadata['gtdb_taxonomy'].apply(lambda t: t.split(';')[6].replace('s__', ''))\n",
    "\n",
    "bac_metadata = bac_metadata.set_index('assembly_accession', drop=True)\n",
    "\n",
    "print(f'Number of genomes: {len(bac_metadata):,}')\n",
    "\n",
    "bac_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Streptococcus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Streptococci genomes: 20,045\n",
      "Number of Streptococci complete genomes: 1,228\n",
      "Number of Streptococci > 98% complete  : 20,045\n"
     ]
    }
   ],
   "source": [
    "strep_df = bac_metadata[bac_metadata['gtdb_genus'] == 'Streptococcus'].copy()\n",
    "\n",
    "n_streptococci = len(strep_df)\n",
    "print(f'Number of Streptococci genomes: {n_streptococci:,}')\n",
    "\n",
    "n_complete_genomes = len(strep_df[strep_df['ncbi_assembly_level'] == 'Complete Genome'])\n",
    "n_good_genomes = len(strep_df[\n",
    "    (strep_df['ncbi_assembly_level'] == 'Complete Genome') |\n",
    "    (strep_df['checkm2_completeness'] > 0.98)\n",
    "])\n",
    "\n",
    "print(f'Number of Streptococci complete genomes: {n_complete_genomes:,}')\n",
    "print(f'Number of Streptococci > 98% complete  : {n_good_genomes:,}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gtdb_genome_representative',\n",
       " 'gtdb_representative',\n",
       " 'ncbi_genome_representation']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[c for c in strep_df.columns if 'repr' in c]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll focus on complete genomes first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of species: 110\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gtdb_species</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Streptococcus pyogenes</th>\n",
       "      <td>271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus agalactiae</th>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pneumoniae</th>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus suis</th>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus thermophilus</th>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus dysgalactiae</th>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus equi</th>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus mutans</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus iniae</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus gordonii</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            accession\n",
       "gtdb_species                         \n",
       "Streptococcus pyogenes            271\n",
       "Streptococcus agalactiae          197\n",
       "Streptococcus pneumoniae          158\n",
       "Streptococcus suis                109\n",
       "Streptococcus thermophilus         88\n",
       "Streptococcus dysgalactiae         43\n",
       "Streptococcus equi                 38\n",
       "Streptococcus mutans               26\n",
       "Streptococcus iniae                14\n",
       "Streptococcus gordonii             13"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strep_cg = strep_df[strep_df['ncbi_assembly_level'] == 'Complete Genome'].copy()\n",
    "\n",
    "n_genomes_per_species = strep_cg[['gtdb_species', 'accession']].groupby('gtdb_species').nunique().sort_values(\n",
    "    'accession', ascending=False,\n",
    ")\n",
    "\n",
    "print(f'Number of species: {len(n_genomes_per_species):,}')\n",
    "\n",
    "n_genomes_per_species.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 38 genomes of _Streptococcus equi_, great."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manual accession changed\n",
    "\n",
    "Since GTDB 220 was release in April 2024, 1 genome of our shortlist of Streptococcus has been updated:\n",
    "\n",
    "GCA_029203915.1 --> GCA_029203915.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Streptococcus suis_W'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strep_cg.index = strep_cg.index.where(strep_cg.index != 'GCA_029203915.1', 'GCA_029203915.2')\n",
    "\n",
    "assert 'GCA_029203915.1' not in strep_cg.index\n",
    "\n",
    "strep_cg.loc['GCA_029203915.2', 'gtdb_species']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export assembly accession list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "strep_cg.to_csv(gtdb_folder / 'Streptococcus' / 'genomes_metadata.csv')\n",
    "\n",
    "strep_cg.reset_index()[['assembly_accession']].to_csv(\n",
    "    gtdb_folder / 'Streptococcus' / 'assembly_accessions.txt', \n",
    "    index=False, \n",
    "    header=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check outputs\n",
    "\n",
    "Have all genomes been downloaded? Are all protein files present?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genomes without predicted proteins: 0\n"
     ]
    }
   ],
   "source": [
    "accessions = set()\n",
    "missing_protein_files = []\n",
    "for p in (gtdb_folder / 'Streptococcus' / 'genomes').iterdir():\n",
    "    if p.name.startswith('GC'):\n",
    "        accession = '_'.join(p.name.split('_')[:2])\n",
    "        accessions.add(accession)\n",
    "\n",
    "        if not (p / f'{p.name}_protein.faa.gz').is_file():\n",
    "            missing_protein_files.append(p.name)\n",
    "\n",
    "assert len(accessions & set(strep_cg.index)) == len(strep_cg)\n",
    "\n",
    "print(f'Number of genomes without predicted proteins: {len(missing_protein_files):,}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GTDB tree\n",
    "\n",
    "How many Streptococci species are present on the tree? All the species representative I suppose, let's see."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "bac_tree = Phylo.read(gtdb_folder / 'bac120_r220.tree', 'newick')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Streptococci in GTDB tree: 79\n"
     ]
    }
   ],
   "source": [
    "gtdb_ids = {leaf.name for leaf in bac_tree.get_terminals()}\n",
    "\n",
    "strep_ids_in_tree = gtdb_ids & set(strep_cg['accession'].unique())\n",
    "\n",
    "print(f'Number of Streptococci in GTDB tree: {len(strep_ids_in_tree):,}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to make a Streptococci tree. \n",
    "\n",
    "- GTDBTk to identify genes\n",
    "- Concatenate core 120 genes\n",
    "- Align with mafft L-INS-i\n",
    "- Build tree with IQ-Tree\n",
    "\n",
    "TBD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CWB domains in Streptococcus\n",
    "\n",
    "From the database assembled with software [github.com/srom/assembly](https://github.com/srom/assembly)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "strep_folder = gtdb_folder / 'Streptococcus'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>hmm_accession</th>\n",
       "      <th>hmm_query</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35053.1@GCA_000013525.1</td>\n",
       "      <td>ABF35053.1</td>\n",
       "      <td>PF01695.22</td>\n",
       "      <td>IstB_IS21</td>\n",
       "      <td>3.500000e-07</td>\n",
       "      <td>26.8</td>\n",
       "      <td>0.71</td>\n",
       "      <td>61</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35053.1@GCA_000013525.1</td>\n",
       "      <td>ABF35053.1</td>\n",
       "      <td>PF00308.23</td>\n",
       "      <td>Bac_DnaA</td>\n",
       "      <td>1.000000e-74</td>\n",
       "      <td>247.2</td>\n",
       "      <td>0.98</td>\n",
       "      <td>111</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35053.1@GCA_000013525.1</td>\n",
       "      <td>ABF35053.1</td>\n",
       "      <td>PF00004.34</td>\n",
       "      <td>AAA</td>\n",
       "      <td>6.300000e-08</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.74</td>\n",
       "      <td>147</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35053.1@GCA_000013525.1</td>\n",
       "      <td>ABF35053.1</td>\n",
       "      <td>PF08299.16</td>\n",
       "      <td>Bac_DnaA_C</td>\n",
       "      <td>3.100000e-31</td>\n",
       "      <td>104.1</td>\n",
       "      <td>0.99</td>\n",
       "      <td>360</td>\n",
       "      <td>429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>ABF35054.1@GCA_000013525.1</td>\n",
       "      <td>ABF35054.1</td>\n",
       "      <td>PF00712.24</td>\n",
       "      <td>DNA_pol3_beta</td>\n",
       "      <td>7.200000e-27</td>\n",
       "      <td>90.8</td>\n",
       "      <td>0.98</td>\n",
       "      <td>1</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            id  protein_id hmm_accession  \\\n",
       "assembly_accession                                                         \n",
       "GCA_000013525.1     ABF35053.1@GCA_000013525.1  ABF35053.1    PF01695.22   \n",
       "GCA_000013525.1     ABF35053.1@GCA_000013525.1  ABF35053.1    PF00308.23   \n",
       "GCA_000013525.1     ABF35053.1@GCA_000013525.1  ABF35053.1    PF00004.34   \n",
       "GCA_000013525.1     ABF35053.1@GCA_000013525.1  ABF35053.1    PF08299.16   \n",
       "GCA_000013525.1     ABF35054.1@GCA_000013525.1  ABF35054.1    PF00712.24   \n",
       "\n",
       "                        hmm_query        evalue  bitscore  accuracy  start  \\\n",
       "assembly_accession                                                           \n",
       "GCA_000013525.1         IstB_IS21  3.500000e-07      26.8      0.71     61   \n",
       "GCA_000013525.1          Bac_DnaA  1.000000e-74     247.2      0.98    111   \n",
       "GCA_000013525.1               AAA  6.300000e-08      30.0      0.74    147   \n",
       "GCA_000013525.1        Bac_DnaA_C  3.100000e-31     104.1      0.99    360   \n",
       "GCA_000013525.1     DNA_pol3_beta  7.200000e-27      90.8      0.98      1   \n",
       "\n",
       "                    end  \n",
       "assembly_accession       \n",
       "GCA_000013525.1     266  \n",
       "GCA_000013525.1     293  \n",
       "GCA_000013525.1     272  \n",
       "GCA_000013525.1     429  \n",
       "GCA_000013525.1     127  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfam_df = pd.read_csv(strep_folder / 'Streptococcus_all_proteins.pfam.csv', index_col='assembly_accession')\n",
    "pfam_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(set(pfam_df.index)) == len(strep_cg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PG_binding_1</th>\n",
       "      <th>PG_binding_2</th>\n",
       "      <th>PG_binding_3</th>\n",
       "      <th>AMIN</th>\n",
       "      <th>SPOR</th>\n",
       "      <th>SH3_1</th>\n",
       "      <th>SH3_2</th>\n",
       "      <th>SH3_3</th>\n",
       "      <th>SH3_4</th>\n",
       "      <th>SH3_5</th>\n",
       "      <th>...</th>\n",
       "      <th>Choline_bind_1</th>\n",
       "      <th>Choline_bind_2</th>\n",
       "      <th>Choline_bind_3</th>\n",
       "      <th>CW_binding_2</th>\n",
       "      <th>CW_7</th>\n",
       "      <th>PSA_CBD</th>\n",
       "      <th>ZoocinA_TRD</th>\n",
       "      <th>GW</th>\n",
       "      <th>OapA</th>\n",
       "      <th>WxL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000014305.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000014325.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000188715.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000211095.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    PG_binding_1  PG_binding_2  PG_binding_3  AMIN  SPOR  \\\n",
       "assembly_accession                                                         \n",
       "GCA_000013525.1                0             0             0     0     0   \n",
       "GCA_000014305.1                0             0             0     0     0   \n",
       "GCA_000014325.1                0             0             0     0     0   \n",
       "GCA_000188715.1                0             0             0     0     0   \n",
       "GCA_000211095.1                0             0             0     0     0   \n",
       "\n",
       "                    SH3_1  SH3_2  SH3_3  SH3_4  SH3_5  ...  Choline_bind_1  \\\n",
       "assembly_accession                                     ...                   \n",
       "GCA_000013525.1         0      0      0      0      3  ...               0   \n",
       "GCA_000014305.1         0      0      0      0      1  ...               0   \n",
       "GCA_000014325.1         0      0      0      0      1  ...               0   \n",
       "GCA_000188715.1         0      0      0      0      1  ...               0   \n",
       "GCA_000211095.1         0      0      0      0      1  ...              12   \n",
       "\n",
       "                    Choline_bind_2  Choline_bind_3  CW_binding_2  CW_7  \\\n",
       "assembly_accession                                                       \n",
       "GCA_000013525.1                  0               0             0     0   \n",
       "GCA_000014305.1                  0               0             0     0   \n",
       "GCA_000014325.1                  0               0             0     0   \n",
       "GCA_000188715.1                  0               0             0     0   \n",
       "GCA_000211095.1                  7               8             0     0   \n",
       "\n",
       "                    PSA_CBD  ZoocinA_TRD  GW  OapA  WxL  \n",
       "assembly_accession                                       \n",
       "GCA_000013525.1           0            0   0     0    0  \n",
       "GCA_000014305.1           0            0   0     0    0  \n",
       "GCA_000014325.1           0            0   0     0    0  \n",
       "GCA_000188715.1           0            0   0     0    0  \n",
       "GCA_000211095.1           0            0   0     0    0  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'assembly_accession': [],\n",
    "}\n",
    "for cwb in cwb_domains:\n",
    "    data[cwb] = []\n",
    "\n",
    "for accession in sorted(set(pfam_df.index)):\n",
    "    data['assembly_accession'].append(accession)\n",
    "    df = pfam_df.loc[[accession]]\n",
    "    for cwb in cwb_domains:\n",
    "        n = len(df[df['hmm_query'] == cwb]['protein_id'].unique())\n",
    "        data[cwb].append(n)\n",
    "\n",
    "cwb_domain_counts = pd.DataFrame.from_dict(data).set_index('assembly_accession', drop=True)\n",
    "cwb_domain_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PG_binding_1</th>\n",
       "      <th>PG_binding_2</th>\n",
       "      <th>PG_binding_3</th>\n",
       "      <th>AMIN</th>\n",
       "      <th>SPOR</th>\n",
       "      <th>SH3_1</th>\n",
       "      <th>SH3_2</th>\n",
       "      <th>SH3_3</th>\n",
       "      <th>SH3_4</th>\n",
       "      <th>SH3_5</th>\n",
       "      <th>...</th>\n",
       "      <th>Choline_bind_1</th>\n",
       "      <th>Choline_bind_2</th>\n",
       "      <th>Choline_bind_3</th>\n",
       "      <th>CW_binding_2</th>\n",
       "      <th>CW_7</th>\n",
       "      <th>PSA_CBD</th>\n",
       "      <th>ZoocinA_TRD</th>\n",
       "      <th>GW</th>\n",
       "      <th>OapA</th>\n",
       "      <th>WxL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000013525.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000014305.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000014325.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000188715.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000211095.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    PG_binding_1  PG_binding_2  PG_binding_3  AMIN  SPOR  \\\n",
       "assembly_accession                                                         \n",
       "GCA_000013525.1                0             0             0     0     0   \n",
       "GCA_000014305.1                0             0             0     0     0   \n",
       "GCA_000014325.1                0             0             0     0     0   \n",
       "GCA_000188715.1                0             0             0     0     0   \n",
       "GCA_000211095.1                0             0             0     0     0   \n",
       "\n",
       "                    SH3_1  SH3_2  SH3_3  SH3_4  SH3_5  ...  Choline_bind_1  \\\n",
       "assembly_accession                                     ...                   \n",
       "GCA_000013525.1         0      0      0      0      1  ...               0   \n",
       "GCA_000014305.1         0      0      0      0      1  ...               0   \n",
       "GCA_000014325.1         0      0      0      0      1  ...               0   \n",
       "GCA_000188715.1         0      0      0      0      1  ...               0   \n",
       "GCA_000211095.1         0      0      0      0      1  ...               1   \n",
       "\n",
       "                    Choline_bind_2  Choline_bind_3  CW_binding_2  CW_7  \\\n",
       "assembly_accession                                                       \n",
       "GCA_000013525.1                  0               0             0     0   \n",
       "GCA_000014305.1                  0               0             0     0   \n",
       "GCA_000014325.1                  0               0             0     0   \n",
       "GCA_000188715.1                  0               0             0     0   \n",
       "GCA_000211095.1                  1               1             0     0   \n",
       "\n",
       "                    PSA_CBD  ZoocinA_TRD  GW  OapA  WxL  \n",
       "assembly_accession                                       \n",
       "GCA_000013525.1           0            0   0     0    0  \n",
       "GCA_000014305.1           0            0   0     0    0  \n",
       "GCA_000014325.1           0            0   0     0    0  \n",
       "GCA_000188715.1           0            0   0     0    0  \n",
       "GCA_000211095.1           0            0   0     0    0  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cwb_domain_bins = cwb_domain_counts.copy()\n",
    "\n",
    "for cwb in cwb_domains:\n",
    "    cwb_domain_bins[cwb] = (cwb_domain_bins[cwb] > 0).astype(int)\n",
    "\n",
    "cwb_domain_bins.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Streptococci strains without LysM: 52 (4.2%)\n"
     ]
    }
   ],
   "source": [
    "n_no_lysm = len(cwb_domain_bins[cwb_domain_bins['LysM'] == 0])\n",
    "total = len(cwb_domain_bins)\n",
    "p = 100 * n_no_lysm / total\n",
    "\n",
    "print(f'Number of Streptococci strains without LysM: {n_no_lysm:,} ({p:.1f}%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "cwb_domain_bins_with_species = cwb_domain_bins.copy()\n",
    "cwb_domain_bins_with_species['gtdb_species'] = [strep_cg.loc[a, 'gtdb_species'] for a in cwb_domain_bins_with_species.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "cwb_domain_bins_with_species[['gtdb_species'] + cwb_domains].to_csv(strep_folder / 'Streptococcus_cell_wall_binding.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_LysM</th>\n",
       "      <th>total</th>\n",
       "      <th>percentage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gtdb_species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Streptococcus parauberis</th>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus porcinus</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus halichoeri</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus porcinus_A</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus equi</th>\n",
       "      <td>4</td>\n",
       "      <td>38</td>\n",
       "      <td>10.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pseudoporcinus</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>33.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pyogenes</th>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus agalactiae</th>\n",
       "      <td>197</td>\n",
       "      <td>197</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pneumoniae</th>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus suis</th>\n",
       "      <td>109</td>\n",
       "      <td>109</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus thermophilus</th>\n",
       "      <td>88</td>\n",
       "      <td>88</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus dysgalactiae</th>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus mutans</th>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus iniae</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus gordonii</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              n_LysM  total  percentage\n",
       "gtdb_species                                           \n",
       "Streptococcus parauberis           0     11        0.00\n",
       "Streptococcus porcinus             0      3        0.00\n",
       "Streptococcus halichoeri           0      1        0.00\n",
       "Streptococcus porcinus_A           0      1        0.00\n",
       "Streptococcus equi                 4     38       10.53\n",
       "Streptococcus pseudoporcinus       1      3       33.33\n",
       "Streptococcus pyogenes           271    271      100.00\n",
       "Streptococcus agalactiae         197    197      100.00\n",
       "Streptococcus pneumoniae         158    158      100.00\n",
       "Streptococcus suis               109    109      100.00\n",
       "Streptococcus thermophilus        88     88      100.00\n",
       "Streptococcus dysgalactiae        43     43      100.00\n",
       "Streptococcus mutans              26     26      100.00\n",
       "Streptococcus iniae               14     14      100.00\n",
       "Streptococcus gordonii            13     13      100.00"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lysM_per_species = cwb_domain_bins_with_species.reset_index()[\n",
    "    ['assembly_accession', 'gtdb_species', 'LysM']\n",
    "].groupby('gtdb_species').agg(\n",
    "    {'LysM': 'sum', 'assembly_accession': 'nunique'}\n",
    ").rename(columns={\n",
    "    'LysM': 'n_LysM',\n",
    "    'assembly_accession': 'total',\n",
    "})\n",
    "\n",
    "lysM_per_species['percentage'] = (100 * lysM_per_species['n_LysM'] / lysM_per_species['total']).round(2)\n",
    "lysM_per_species.sort_values(['percentage', 'total'], ascending=[True, False]).head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000219765.1</th>\n",
       "      <td>Streptococcus equi subsp. zooepidemicus ATCC 3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_009676645.1</th>\n",
       "      <td>Streptococcus equi subsp. zooepidemicus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_009676685.2</th>\n",
       "      <td>Streptococcus equi subsp. zooepidemicus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_900636805.1</th>\n",
       "      <td>Streptococcus equi subsp. zooepidemicus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   ncbi_organism_name\n",
       "assembly_accession                                                   \n",
       "GCF_000219765.1     Streptococcus equi subsp. zooepidemicus ATCC 3...\n",
       "GCF_009676645.1               Streptococcus equi subsp. zooepidemicus\n",
       "GCF_009676685.2               Streptococcus equi subsp. zooepidemicus\n",
       "GCF_900636805.1               Streptococcus equi subsp. zooepidemicus"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_equi_with_lysM = cwb_domain_bins_with_species[\n",
    "    (cwb_domain_bins_with_species['gtdb_species'] == 'Streptococcus equi') &\n",
    "    (cwb_domain_bins_with_species['LysM'] > 0)\n",
    "].index.tolist()\n",
    "\n",
    "strep_cg.loc[s_equi_with_lysM][['ncbi_organism_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_015689395.1</th>\n",
       "      <td>Streptococcus equi subsp. zooepidemicus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_015767555.1</th>\n",
       "      <td>Streptococcus equi subsp. zooepidemicus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_900636805.1</th>\n",
       "      <td>Streptococcus equi subsp. zooepidemicus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         ncbi_organism_name\n",
       "assembly_accession                                         \n",
       "GCF_015689395.1     Streptococcus equi subsp. zooepidemicus\n",
       "GCF_015767555.1     Streptococcus equi subsp. zooepidemicus\n",
       "GCF_900636805.1     Streptococcus equi subsp. zooepidemicus"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_equi_with_zooA = cwb_domain_bins_with_species[\n",
    "    (cwb_domain_bins_with_species['gtdb_species'] == 'Streptococcus equi') &\n",
    "    (cwb_domain_bins_with_species['ZoocinA_TRD'] > 0)\n",
    "].index.tolist()\n",
    "\n",
    "strep_cg.loc[s_equi_with_zooA][['ncbi_organism_name']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How often LysM-containing proteins associated with peptidoglycan catalytic domains?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "catalytic_domains = [\n",
    "    'Amidase',\n",
    "    'Amidase_2',\n",
    "    'Amidase_3',\n",
    "    'Amidase_5',\n",
    "    'Amidase_6',\n",
    "    'Pepdidase_M14_N',\n",
    "    'Peptidase_C107',\n",
    "    'Peptidase_C21',\n",
    "    'Peptidase_C23',\n",
    "    'Peptidase_C24',\n",
    "    'Peptidase_C27',\n",
    "    'Peptidase_C30',\n",
    "    'Peptidase_C34',\n",
    "    'Peptidase_C36',\n",
    "    'Peptidase_C42',\n",
    "    'Peptidase_C53',\n",
    "    'Peptidase_C92',\n",
    "    'Peptidase_M14',\n",
    "    'Peptidase_M15',\n",
    "    'Peptidase_M15_2',\n",
    "    'Peptidase_M15_3',\n",
    "    'Peptidase_M15_4',\n",
    "    'Peptidase_M2',\n",
    "    'Peptidase_M23',\n",
    "    'Peptidase_M26_C',\n",
    "    'Peptidase_M26_N',\n",
    "    'Peptidase_M32',\n",
    "    'Peptidase_M73',\n",
    "    'Peptidase_M74',\n",
    "    'Peptidase_M99',\n",
    "    'Peptidase_M99_C',\n",
    "    'Peptidase_M99_m',\n",
    "    'Peptidase_S10',\n",
    "    'Peptidase_S11',\n",
    "    'Peptidase_S13',\n",
    "    'Peptidase_S28',\n",
    "    'Peptidase_S32',\n",
    "    'Peptidase_S37',\n",
    "    'Peptidase_S66',\n",
    "    'Peptidase_S66C',\n",
    "    'Propep_M14',\n",
    "    'Prophage_tail',\n",
    "    'Prophage_tailD1',\n",
    "    'VanY',\n",
    "    'Glucosaminidase',\n",
    "    'CHAP',\n",
    "    'Aspzincin_M35',\n",
    "    'Glyco_hydro_25',\n",
    "    'Glyco_hydro_108',\n",
    "    'Lysozyme_like',\n",
    "    'Phage_lysozyme',\n",
    "    'Phage_lysozyme2',\n",
    "    'SLT',\n",
    "    'SLT_2',\n",
    "    'NLPC_P60',\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = cwb_domain_bins_with_species[cwb_domain_bins_with_species['LysM'] > 0].index\n",
    "df = pfam_df.loc[ids]\n",
    "protein_ids = df[df['hmm_query'] == 'LysM']['id'].unique()\n",
    "accessions = set(df[\n",
    "    df['id'].isin(protein_ids) &\n",
    "    df['hmm_query'].isin(catalytic_domains)\n",
    "].index)\n",
    "\n",
    "cwb_domain_bins_with_species['has_LysM_pgh'] = [a in accessions for a in cwb_domain_bins_with_species.index]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "lysM_pgh_per_species = cwb_domain_bins_with_species.reset_index()[\n",
    "    ['assembly_accession', 'gtdb_species', 'has_LysM_pgh']\n",
    "].groupby('gtdb_species').agg(\n",
    "    {'has_LysM_pgh': 'sum', 'assembly_accession': 'nunique'}\n",
    ").rename(columns={\n",
    "    'assembly_accession': 'total',\n",
    "})\n",
    "\n",
    "lysM_pgh_per_species['percentage'] = (100 * lysM_pgh_per_species['has_LysM_pgh'] / lysM_pgh_per_species['total']).round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_LysM</th>\n",
       "      <th>total</th>\n",
       "      <th>percentage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gtdb_species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Streptococcus parauberis</th>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus porcinus</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus halichoeri</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus porcinus_A</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus equi</th>\n",
       "      <td>4</td>\n",
       "      <td>38</td>\n",
       "      <td>10.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pseudoporcinus</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>33.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pyogenes</th>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus agalactiae</th>\n",
       "      <td>197</td>\n",
       "      <td>197</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus pneumoniae</th>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus suis</th>\n",
       "      <td>109</td>\n",
       "      <td>109</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus thermophilus</th>\n",
       "      <td>88</td>\n",
       "      <td>88</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus dysgalactiae</th>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus mutans</th>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus iniae</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptococcus gordonii</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              n_LysM  total  percentage\n",
       "gtdb_species                                           \n",
       "Streptococcus parauberis           0     11        0.00\n",
       "Streptococcus porcinus             0      3        0.00\n",
       "Streptococcus halichoeri           0      1        0.00\n",
       "Streptococcus porcinus_A           0      1        0.00\n",
       "Streptococcus equi                 4     38       10.53\n",
       "Streptococcus pseudoporcinus       1      3       33.33\n",
       "Streptococcus pyogenes           271    271      100.00\n",
       "Streptococcus agalactiae         197    197      100.00\n",
       "Streptococcus pneumoniae         158    158      100.00\n",
       "Streptococcus suis               109    109      100.00\n",
       "Streptococcus thermophilus        88     88      100.00\n",
       "Streptococcus dysgalactiae        43     43      100.00\n",
       "Streptococcus mutans              26     26      100.00\n",
       "Streptococcus iniae               14     14      100.00\n",
       "Streptococcus gordonii            13     13      100.00"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lysM_per_species.sort_values(['percentage', 'total'], ascending=[True, False]).head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyogenes_ids = strep_cg[strep_cg['gtdb_species'] == 'Streptococcus pyogenes'].index\n",
    "df = pfam_df.loc[pyogenes_ids]\n",
    "protein_ids = df[df['hmm_query'] == 'LysM']['id'].unique()\n",
    "pfam_df[pfam_df['id'].isin(protein_ids) & (pfam_df['hmm_query'] != 'LysM')]['hmm_query'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All pyogenes proteins containing LysM only contain LysM (no other Pfam domain predicted)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Has LysM been replaced with another CWB domain?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "no_LysM_species = lysM_per_species.sort_values(['percentage', 'total'], ascending=[True, False]).iloc[:6].index.tolist()\n",
    "yes_LysM_species = lysM_per_species.sort_values(['percentage', 'total'], ascending=[True, False]).iloc[6:].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>yes_LysM</th>\n",
       "      <th>no_LysM</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LysM</th>\n",
       "      <td>11.259615</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>10.426282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GW</th>\n",
       "      <td>0.144231</td>\n",
       "      <td>4.333333</td>\n",
       "      <td>4.189103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Choline_bind_3</th>\n",
       "      <td>3.307692</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.307692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Choline_bind_1</th>\n",
       "      <td>3.278846</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.278846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Choline_bind_2</th>\n",
       "      <td>2.182692</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.182692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZoocinA_TRD</th>\n",
       "      <td>2.096154</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.596154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SH3_5</th>\n",
       "      <td>10.432692</td>\n",
       "      <td>9.500000</td>\n",
       "      <td>0.932692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CW_7</th>\n",
       "      <td>1.432692</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.432692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SH3_3</th>\n",
       "      <td>0.201923</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.131410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 yes_LysM   no_LysM       diff\n",
       "LysM            11.259615  0.833333  10.426282\n",
       "GW               0.144231  4.333333   4.189103\n",
       "Choline_bind_3   3.307692  0.000000   3.307692\n",
       "Choline_bind_1   3.278846  0.000000   3.278846\n",
       "Choline_bind_2   2.182692  0.000000   2.182692\n",
       "ZoocinA_TRD      2.096154  0.500000   1.596154\n",
       "SH3_5           10.432692  9.500000   0.932692\n",
       "CW_7             1.432692  1.000000   0.432692\n",
       "SH3_3            0.201923  0.333333   0.131410"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_series = cwb_domain_bins_with_species[\n",
    "    cwb_domain_bins_with_species['gtdb_species'].isin(no_LysM_species)\n",
    "][cwb_domains].sum() / len(no_LysM_species)\n",
    "\n",
    "yes_series = cwb_domain_bins_with_species[\n",
    "    cwb_domain_bins_with_species['gtdb_species'].isin(yes_LysM_species)\n",
    "][cwb_domains].sum() / len(yes_LysM_species)\n",
    "\n",
    "cwb_stats = pd.DataFrame({\n",
    "    'yes_LysM': yes_series,\n",
    "    'no_LysM': no_series\n",
    "})\n",
    "cwb_stats['diff'] = (cwb_stats['yes_LysM'] - cwb_stats['no_LysM']).abs()\n",
    "\n",
    "cwb_stats = cwb_stats.sort_values('diff', ascending=False)\n",
    "cwb_stats[cwb_stats['diff'] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_companion_domains(strep_cg, hmm_query):\n",
    "    strep_ids = strep_cg.index\n",
    "    df = pfam_df.loc[strep_ids]\n",
    "    protein_ids = df[df['hmm_query'] == hmm_query]['id'].unique()\n",
    "\n",
    "    companion_domains = pfam_df[\n",
    "        pfam_df['id'].isin(protein_ids) & \n",
    "        (pfam_df['hmm_query'] != hmm_query)\n",
    "    ]\n",
    "\n",
    "    n_standalone = len(set(protein_ids) - set(companion_domains['id'].unique()))\n",
    "    percent = 100 * n_standalone / len(protein_ids)\n",
    "\n",
    "    print(f'Number of standalone {hmm_query}: {n_standalone:,} ({percent:.0f}%)')\n",
    "    print('Top 5 companion domains:')\n",
    "    print(companion_domains['hmm_query'].value_counts()[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of standalone LysM: 2,172 (82%)\n",
      "Top 5 companion domains:\n",
      "hmm_query\n",
      "CHAP              365\n",
      "Amidase_3         116\n",
      "SLT                82\n",
      "GA                 21\n",
      "Glyco_hydro_25     15\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Number of standalone GW: 0 (0%)\n",
      "Top 5 companion domains:\n",
      "hmm_query\n",
      "Beta-lactamase2    41\n",
      "DUF5776            13\n",
      "Peptidase_S11       9\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Number of standalone Choline_bind_3: 38 (1%)\n",
      "Top 5 companion domains:\n",
      "hmm_query\n",
      "Choline_bind_1    14864\n",
      "Choline_bind_4      941\n",
      "Choline_bind_2      787\n",
      "KxYKxGKxW_sig       499\n",
      "SH3_5               333\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Number of standalone Choline_bind_1: 120 (4%)\n",
      "Top 5 companion domains:\n",
      "hmm_query\n",
      "Choline_bind_3    8911\n",
      "Choline_bind_2    1651\n",
      "Choline_bind_4     941\n",
      "KxYKxGKxW_sig      439\n",
      "SH3_5              333\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Number of standalone Choline_bind_2: 145 (11%)\n",
      "Top 5 companion domains:\n",
      "hmm_query\n",
      "Choline_bind_1     8205\n",
      "Choline_bind_3     2932\n",
      "Choline_bind_4      933\n",
      "Glucosaminidase     198\n",
      "LytB_SH3            198\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Number of standalone ZoocinA_TRD: 0 (0%)\n",
      "Top 5 companion domains:\n",
      "hmm_query\n",
      "Peptidase_M23    200\n",
      "Amidase_5         22\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Number of standalone SH3_5: 177 (9%)\n",
      "Top 5 companion domains:\n",
      "hmm_query\n",
      "CHAP               1364\n",
      "Choline_bind_1      812\n",
      "GBS_Bsp-like        675\n",
      "Glucosaminidase     414\n",
      "Choline_bind_3      398\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Number of standalone CW_7: 0 (0%)\n",
      "Top 5 companion domains:\n",
      "hmm_query\n",
      "Amidase_5          155\n",
      "NLPC_P60            10\n",
      "Amidase_2            9\n",
      "Glucosaminidase      2\n",
      "CHAP                 1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Number of standalone SH3_3: 0 (0%)\n",
      "Top 5 companion domains:\n",
      "hmm_query\n",
      "DUF6287    20\n",
      "CHAP        3\n",
      "Name: count, dtype: int64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for hmm_query in cwb_stats[cwb_stats['diff'] > 0].index:\n",
    "    get_companion_domains(strep_cg, hmm_query)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cwb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
