{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Streptococcus S protein\n",
    "\n",
    "OG: 4HHT0\n",
    "\n",
    "S proteins are LysM-containing proteins playing a central role in pathogenicity:\n",
    "- immune evasion in Group A Streptococci (camouflaging with red blood cell debris) [1]\n",
    "- resistance to beta-lactam antibiotics through recruitment of protein PGP1a [2]\n",
    "- resistance to lysozyme through recruitment of protein PgdA. [2]\n",
    "  - PgdA is a peptidoglycan N-acetylglucosamine deacetylase, i.e. it removes the N-acetyl group from the middle sugar of the peptidoglycan.\n",
    "  - It helps the bacteria resist lysozyme [3].\n",
    "\n",
    "Mutants of _S. pneumoniae_ without S protein show reduce virulence in mouse models, are more suceptible to beta-lactam antibiotics and lysozyme. [2]\n",
    "\n",
    "Refs: \n",
    "- [1] [Wierzbicki et al., 2019](https://doi.org/10.1016/j.celrep.2019.11.001)\n",
    "- [2] [Burnier et al., 2024](https://doi.org/10.1101/2024.11.08.622053)\n",
    "- [3] [Bui et al., 2011](https://doi.org/10.1016/j.bcp.2011.03.028)\n",
    "\n",
    "We found that S proteins are absent from all complete genomes of _S. equi_.\n",
    "\n",
    "_S. equi_ is primarily a horse pathogen but _S. equi subsp. zooepidemicus_ has zoonotic potential: it also infects a wide range of mammals, including dogs, cats, ruminants, pigs, and humans."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as plt\n",
    "import seaborn as sns\n",
    "from Bio import Phylo\n",
    "\n",
    "cwd = os.getcwd()\n",
    "if cwd.endswith('notebook'):\n",
    "    os.chdir('..')\n",
    "    cwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme(palette='colorblind', font_scale=1.3)\n",
    "palette_colorblind = sns.color_palette('colorblind').as_hex()\n",
    "palette_pastel = sns.color_palette('pastel').as_hex()\n",
    "\n",
    "data_folder = Path('./data/')\n",
    "assert data_folder.is_dir()\n",
    "\n",
    "db_proka = Path('../db_proka/')\n",
    "assert db_proka.is_dir()\n",
    "\n",
    "gtdb_folder = Path('../data/gtdb_r220/')\n",
    "assert gtdb_folder.is_dir()\n",
    "\n",
    "strep_folder = gtdb_folder / 'Streptococcus'\n",
    "assert strep_folder.is_dir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>ambiguous_bases</th>\n",
       "      <th>checkm2_completeness</th>\n",
       "      <th>checkm2_contamination</th>\n",
       "      <th>checkm2_model</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>checkm_marker_count</th>\n",
       "      <th>checkm_marker_lineage</th>\n",
       "      <th>checkm_marker_set_count</th>\n",
       "      <th>...</th>\n",
       "      <th>trna_aa_count</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>trna_selenocysteine_count</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_900636555.1</th>\n",
       "      <td>RS_GCF_900636555.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.14</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>475</td>\n",
       "      <td>o__Lactobacillales (UID544)</td>\n",
       "      <td>267</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus gordonii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_015908985.1</th>\n",
       "      <td>RS_GCF_015908985.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.23</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>475</td>\n",
       "      <td>o__Lactobacillales (UID544)</td>\n",
       "      <td>267</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus suis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_001266635.1</th>\n",
       "      <td>RS_GCF_001266635.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.06</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.82</td>\n",
       "      <td>0.00</td>\n",
       "      <td>524</td>\n",
       "      <td>f__Streptococcaceae (UID545)</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_004154025.1</th>\n",
       "      <td>RS_GCF_004154025.1</td>\n",
       "      <td>0</td>\n",
       "      <td>99.99</td>\n",
       "      <td>0.13</td>\n",
       "      <td>Specific</td>\n",
       "      <td>99.85</td>\n",
       "      <td>0.00</td>\n",
       "      <td>676</td>\n",
       "      <td>g__Streptococcus (UID722)</td>\n",
       "      <td>182</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_029011635.1</th>\n",
       "      <td>RS_GCF_029011635.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.20</td>\n",
       "      <td>Specific</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.18</td>\n",
       "      <td>524</td>\n",
       "      <td>f__Streptococcaceae (UID545)</td>\n",
       "      <td>282</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             accession  ambiguous_bases  checkm2_completeness  \\\n",
       "assembly_accession                                                              \n",
       "GCF_900636555.1     RS_GCF_900636555.1                0                100.00   \n",
       "GCF_015908985.1     RS_GCF_015908985.1                0                100.00   \n",
       "GCF_001266635.1     RS_GCF_001266635.1                0                100.00   \n",
       "GCF_004154025.1     RS_GCF_004154025.1                0                 99.99   \n",
       "GCF_029011635.1     RS_GCF_029011635.1                0                100.00   \n",
       "\n",
       "                    checkm2_contamination checkm2_model  checkm_completeness  \\\n",
       "assembly_accession                                                             \n",
       "GCF_900636555.1                      0.14      Specific               100.00   \n",
       "GCF_015908985.1                      0.23      Specific               100.00   \n",
       "GCF_001266635.1                      0.06      Specific                99.82   \n",
       "GCF_004154025.1                      0.13      Specific                99.85   \n",
       "GCF_029011635.1                      0.20      Specific               100.00   \n",
       "\n",
       "                    checkm_contamination  checkm_marker_count  \\\n",
       "assembly_accession                                              \n",
       "GCF_900636555.1                     0.00                  475   \n",
       "GCF_015908985.1                     0.00                  475   \n",
       "GCF_001266635.1                     0.00                  524   \n",
       "GCF_004154025.1                     0.00                  676   \n",
       "GCF_029011635.1                     0.18                  524   \n",
       "\n",
       "                           checkm_marker_lineage  checkm_marker_set_count  \\\n",
       "assembly_accession                                                          \n",
       "GCF_900636555.1      o__Lactobacillales (UID544)                      267   \n",
       "GCF_015908985.1      o__Lactobacillales (UID544)                      267   \n",
       "GCF_001266635.1     f__Streptococcaceae (UID545)                      282   \n",
       "GCF_004154025.1        g__Streptococcus (UID722)                      182   \n",
       "GCF_029011635.1     f__Streptococcaceae (UID545)                      282   \n",
       "\n",
       "                    ...  trna_aa_count  trna_count  trna_selenocysteine_count  \\\n",
       "assembly_accession  ...                                                         \n",
       "GCF_900636555.1     ...             19          59                          0   \n",
       "GCF_015908985.1     ...             19          56                          0   \n",
       "GCF_001266635.1     ...             18          63                          0   \n",
       "GCF_004154025.1     ...             19          67                          0   \n",
       "GCF_029011635.1     ...             19          80                          0   \n",
       "\n",
       "                      domain  gtdb_phylum  gtdb_class       gtdb_order  \\\n",
       "assembly_accession                                                       \n",
       "GCF_900636555.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_015908985.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_001266635.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_004154025.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "GCF_029011635.1     Bacteria    Bacillota     Bacilli  Lactobacillales   \n",
       "\n",
       "                         gtdb_family     gtdb_genus              gtdb_species  \n",
       "assembly_accession                                                             \n",
       "GCF_900636555.1     Streptococcaceae  Streptococcus    Streptococcus gordonii  \n",
       "GCF_015908985.1     Streptococcaceae  Streptococcus        Streptococcus suis  \n",
       "GCF_001266635.1     Streptococcaceae  Streptococcus  Streptococcus agalactiae  \n",
       "GCF_004154025.1     Streptococcaceae  Streptococcus    Streptococcus pyogenes  \n",
       "GCF_029011635.1     Streptococcaceae  Streptococcus  Streptococcus agalactiae  \n",
       "\n",
       "[5 rows x 120 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df = pd.read_csv(strep_folder / 'genomes_metadata.csv', index_col='assembly_accession')\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfam_df = pd.read_csv(strep_folder / 'Streptococcus_all_proteins.pfam.csv', index_col='assembly_accession')\n",
    "pfam_df['gtdb_species'] = [metadata_df.loc[a, 'gtdb_species'] for a in pfam_df.index]\n",
    "\n",
    "tigr_df = pd.read_csv(strep_folder / 'Streptococcus_all_proteins.tigr.csv', index_col='assembly_accession')\n",
    "tigr_df['gtdb_species'] = [metadata_df.loc[a, 'gtdb_species'] for a in tigr_df.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467658.1@GCF_003963555.1</td>\n",
       "      <td>1000570.HMPREF9966_1759</td>\n",
       "      <td>9.704000e-82</td>\n",
       "      <td>271.0</td>\n",
       "      <td>COG0716@1|root,COG0716@2|Bacteria,1V45R@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>C</td>\n",
       "      <td>Flavodoxin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Flavodoxin_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467724.1@GCF_003963555.1</td>\n",
       "      <td>862970.SAIN_1577</td>\n",
       "      <td>1.313000e-186</td>\n",
       "      <td>582.0</td>\n",
       "      <td>COG1396@1|root,COG1396@2|Bacteria,1VIH9@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>K</td>\n",
       "      <td>Helix-turn-helix XRE-family like proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HTH_19,HTH_3,TPR_12,TPR_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467790.1@GCF_003963555.1</td>\n",
       "      <td>862970.SAIN_1636</td>\n",
       "      <td>2.776000e-131</td>\n",
       "      <td>418.0</td>\n",
       "      <td>COG1564@1|root,COG1564@2|Bacteria,1VA0W@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>H</td>\n",
       "      <td>Thiamin pyrophosphokinase, vitamin B1 binding ...</td>\n",
       "      <td>thiN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K00949</td>\n",
       "      <td>ko00730,ko01100,map00730,map01100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R00619</td>\n",
       "      <td>RC00002,RC00017</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TPK_B1_binding,TPK_catalytic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467857.1@GCF_003963555.1</td>\n",
       "      <td>176090.SSIN_0693</td>\n",
       "      <td>1.181000e-201</td>\n",
       "      <td>627.0</td>\n",
       "      <td>COG3677@1|root,COG3677@2|Bacteria,1V4D1@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>L</td>\n",
       "      <td>ISXO2-like transposase domain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDE_Tnp_IS1595,Zn_Tnp_IS1595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_003963555.1</th>\n",
       "      <td>WP_126467993.1@GCF_003963555.1</td>\n",
       "      <td>862969.SCI_1925</td>\n",
       "      <td>1.812000e-256</td>\n",
       "      <td>791.0</td>\n",
       "      <td>COG0612@1|root,COG0612@2|Bacteria,1TPN6@1239|F...</td>\n",
       "      <td>2|Bacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>Peptidase M16 inactive</td>\n",
       "      <td>ymfF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peptidase_M16,Peptidase_M16_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        protein_id            seed_ortholog  \\\n",
       "assembly_accession                                                            \n",
       "GCF_003963555.1     WP_126467658.1@GCF_003963555.1  1000570.HMPREF9966_1759   \n",
       "GCF_003963555.1     WP_126467724.1@GCF_003963555.1         862970.SAIN_1577   \n",
       "GCF_003963555.1     WP_126467790.1@GCF_003963555.1         862970.SAIN_1636   \n",
       "GCF_003963555.1     WP_126467857.1@GCF_003963555.1         176090.SSIN_0693   \n",
       "GCF_003963555.1     WP_126467993.1@GCF_003963555.1          862969.SCI_1925   \n",
       "\n",
       "                           evalue  score  \\\n",
       "assembly_accession                         \n",
       "GCF_003963555.1      9.704000e-82  271.0   \n",
       "GCF_003963555.1     1.313000e-186  582.0   \n",
       "GCF_003963555.1     2.776000e-131  418.0   \n",
       "GCF_003963555.1     1.181000e-201  627.0   \n",
       "GCF_003963555.1     1.812000e-256  791.0   \n",
       "\n",
       "                                                           eggNOG_OGs  \\\n",
       "assembly_accession                                                      \n",
       "GCF_003963555.1     COG0716@1|root,COG0716@2|Bacteria,1V45R@1239|F...   \n",
       "GCF_003963555.1     COG1396@1|root,COG1396@2|Bacteria,1VIH9@1239|F...   \n",
       "GCF_003963555.1     COG1564@1|root,COG1564@2|Bacteria,1VA0W@1239|F...   \n",
       "GCF_003963555.1     COG3677@1|root,COG3677@2|Bacteria,1V4D1@1239|F...   \n",
       "GCF_003963555.1     COG0612@1|root,COG0612@2|Bacteria,1TPN6@1239|F...   \n",
       "\n",
       "                   max_annot_lvl COG_category  \\\n",
       "assembly_accession                              \n",
       "GCF_003963555.1       2|Bacteria            C   \n",
       "GCF_003963555.1       2|Bacteria            K   \n",
       "GCF_003963555.1       2|Bacteria            H   \n",
       "GCF_003963555.1       2|Bacteria            L   \n",
       "GCF_003963555.1       2|Bacteria            S   \n",
       "\n",
       "                                                          Description  \\\n",
       "assembly_accession                                                      \n",
       "GCF_003963555.1                                            Flavodoxin   \n",
       "GCF_003963555.1             Helix-turn-helix XRE-family like proteins   \n",
       "GCF_003963555.1     Thiamin pyrophosphokinase, vitamin B1 binding ...   \n",
       "GCF_003963555.1                         ISXO2-like transposase domain   \n",
       "GCF_003963555.1                                Peptidase M16 inactive   \n",
       "\n",
       "                   Preferred_name  GOs  ...    KEGG_ko  \\\n",
       "assembly_accession                      ...              \n",
       "GCF_003963555.1               NaN  NaN  ...        NaN   \n",
       "GCF_003963555.1               NaN  NaN  ...        NaN   \n",
       "GCF_003963555.1              thiN  NaN  ...  ko:K00949   \n",
       "GCF_003963555.1               NaN  NaN  ...        NaN   \n",
       "GCF_003963555.1              ymfF  NaN  ...        NaN   \n",
       "\n",
       "                                         KEGG_Pathway KEGG_Module  \\\n",
       "assembly_accession                                                  \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "GCF_003963555.1     ko00730,ko01100,map00730,map01100         NaN   \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "GCF_003963555.1                                   NaN         NaN   \n",
       "\n",
       "                   KEGG_Reaction      KEGG_rclass                    BRITE  \\\n",
       "assembly_accession                                                           \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "GCF_003963555.1           R00619  RC00002,RC00017  ko00000,ko00001,ko01000   \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "GCF_003963555.1              NaN              NaN                      NaN   \n",
       "\n",
       "                   KEGG_TC CAZy BiGG_Reaction                          PFAMs  \n",
       "assembly_accession                                                            \n",
       "GCF_003963555.1        NaN  NaN           NaN                   Flavodoxin_4  \n",
       "GCF_003963555.1        NaN  NaN           NaN      HTH_19,HTH_3,TPR_12,TPR_8  \n",
       "GCF_003963555.1        NaN  NaN           NaN   TPK_B1_binding,TPK_catalytic  \n",
       "GCF_003963555.1        NaN  NaN           NaN   DDE_Tnp_IS1595,Zn_Tnp_IS1595  \n",
       "GCF_003963555.1        NaN  NaN           NaN  Peptidase_M16,Peptidase_M16_C  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggNOG_annotations_path = strep_folder / 'Streptococcus_eggNOG_annotations.csv'\n",
    "eggNOG_df = pd.read_csv(eggNOG_annotations_path, index_col='assembly_accession')\n",
    "eggNOG_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_ogs(eggNOG_df, accessions, og_whitelist=None):\n",
    "    index = eggNOG_df.index\n",
    "\n",
    "    og_union = set()\n",
    "    for acc in accessions:\n",
    "        if acc not in index:\n",
    "            print(f'Not in index: {acc}')\n",
    "            continue\n",
    "\n",
    "        genme_ogs = eggNOG_df.loc[[acc]].set_index('protein_id')\n",
    "\n",
    "        ogs = set()\n",
    "        for protein_id in genme_ogs.index:\n",
    "            protein_ogs = genme_ogs.loc[protein_id, 'eggNOG_OGs']\n",
    "\n",
    "            candidate_ogs = {\n",
    "                og_with_tax.split('@')[0]\n",
    "                for og_with_tax in protein_ogs.split(',')\n",
    "            }\n",
    "\n",
    "            if og_whitelist is not None:\n",
    "                candidate_ogs = candidate_ogs & og_whitelist\n",
    "                \n",
    "            ogs |= candidate_ogs\n",
    "\n",
    "        og_union = og_union.union(ogs)\n",
    "\n",
    "    return og_union"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ogs_present_in_all(eggNOG_df, accessions, og_whitelist=None):\n",
    "    index = eggNOG_df.index\n",
    "\n",
    "    og_intersection = None\n",
    "    for acc in accessions:\n",
    "        if acc not in index:\n",
    "            print(f'Not in index: {acc}')\n",
    "            continue\n",
    "\n",
    "        genme_ogs = eggNOG_df.loc[[acc]].set_index('protein_id')\n",
    "\n",
    "        ogs = set()\n",
    "        for protein_id in genme_ogs.index:\n",
    "            protein_ogs = genme_ogs.loc[protein_id, 'eggNOG_OGs']\n",
    "\n",
    "            candidate_ogs = {\n",
    "                og_with_tax.split('@')[0]\n",
    "                for og_with_tax in protein_ogs.split(',')\n",
    "            }\n",
    "\n",
    "            if og_whitelist is not None:\n",
    "                candidate_ogs = candidate_ogs & og_whitelist\n",
    "                \n",
    "            ogs |= candidate_ogs\n",
    "\n",
    "        if og_intersection is None:\n",
    "            og_intersection = ogs\n",
    "        else:\n",
    "            og_intersection &= ogs\n",
    "\n",
    "    return og_intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2190"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_ogs = {\n",
    "    og\n",
    "    for og_str in eggNOG_df['eggNOG_OGs']\n",
    "    for og_raw in og_str.split(',')\n",
    "    if (og := og_raw.split('@')[0]).startswith('COG')\n",
    "}\n",
    "\n",
    "len(cog_ogs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "og_metadata = pd.read_csv(\n",
    "    strep_folder / 'eggNOG' / 'e5.og_annotations.tsv',\n",
    "    sep='\\t', \n",
    "    header=None,\n",
    "    names=['og', 'og_level', 'description'],\n",
    ").drop_duplicates('og').set_index('og')\n",
    "og_metadata.head()\n",
    "\n",
    "def display_ogs(ogs):\n",
    "    for og in ogs:\n",
    "        desc = og_metadata.loc[og, 'description']\n",
    "        print(f'{og}: {desc}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = [\n",
    "    'Streptococcus pyogens',\n",
    "    'Streptococcus dysgalactiae',\n",
    "    'Streptococcus equi',\n",
    "    'Streptococcus canis',\n",
    "]\n",
    "metadata_subset = metadata_df[metadata_df['gtdb_species'].isin(species)].copy()\n",
    "\n",
    "accessions = metadata_subset.index\n",
    "\n",
    "eggNOG_subset = eggNOG_df.loc[accessions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "accessions_S_protein = sorted(set(eggNOG_subset[eggNOG_subset['eggNOG_OGs'].str.contains('4HHT0')].index))\n",
    "accessions_no_S_protein = sorted(set(accessions) - set(accessions_S_protein))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(856, 1237)"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_S_protein = get_ogs_present_in_all(eggNOG_subset, accessions_no_S_protein, cog_ogs)\n",
    "no_S_protein_all = get_unique_ogs(eggNOG_subset, accessions_no_S_protein, cog_ogs)\n",
    "\n",
    "len(no_S_protein), len(no_S_protein_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(887, 1396)"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yes_S_protein = get_ogs_present_in_all(eggNOG_subset, accessions_S_protein, cog_ogs)\n",
    "yes_S_protein_all = get_unique_ogs(eggNOG_subset, accessions_S_protein, cog_ogs)\n",
    "\n",
    "len(yes_S_protein), len(yes_S_protein_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gain in S. equi\n",
    "\n",
    "COGs present in all S. equi but in no related Strep."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gain: 13 OGs\n",
      "COG1345: cell adhesion\n",
      "COG1376: peptidoglycan L,D-transpeptidase activity\n",
      "COG1434: Gram-negative-bacterium-type cell wall biogenesis\n",
      "COG2848: Uncharacterised ACR (DUF711)\n",
      "COG3458: cephalosporin-C deacetylase activity\n",
      "COG3830: Belongs to the UPF0237 family\n",
      "COG3958: Transketolase\n",
      "COG3959: Transketolase, thiamine diphosphate binding domain\n",
      "COG4495: Domain of unknown function (DUF4176)\n",
      "COG4935: serine-type endopeptidase activity\n",
      "COG5353: nan\n",
      "COG5444: nuclease activity\n",
      "COG5585: self proteolysis\n"
     ]
    }
   ],
   "source": [
    "missing_in_yes = no_S_protein - yes_S_protein_all\n",
    "print('Gain:', len(missing_in_yes), 'OGs')\n",
    "display_ogs(sorted(missing_in_yes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loss in S. equi\n",
    "\n",
    "COGs present in all related Strep with S protein but in no rS. equi."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loss: 34 OGs\n",
      "COG0010: Belongs to the arginase family\n",
      "COG0574: phosphotransferase activity, paired acceptors\n",
      "COG0578: glycerol-3-phosphate dehydrogenase (quinone) activity\n",
      "COG0625: glutathione transferase activity\n",
      "COG0727: PFAM Uncharacterised protein family (UPF0153)\n",
      "COG0730: response to heat\n",
      "COG0813: purine-nucleoside phosphorylase activity\n",
      "COG1054: Rhodanese Homology Domain\n",
      "COG1228: amidohydrolase\n",
      "COG1252: NADH oxidation\n",
      "COG1271: oxidase subunit\n",
      "COG1575: vitamin K biosynthetic process\n",
      "COG1596: polysaccharide transmembrane transporter activity\n",
      "COG1600: epoxyqueuosine reductase activity\n",
      "COG1765: OsmC-like protein\n",
      "COG1806: Bifunctional serine threonine kinase and phosphorylase involved in the regulation of the\n",
      "COG1902: FMN binding\n",
      "COG2110: O-acetyl-ADP-ribose deacetylase activity\n",
      "COG2141: oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen\n",
      "COG2223: nitrate assimilation\n",
      "COG2608: superoxide dismutase copper chaperone activity\n",
      "COG2764: nan\n",
      "COG2986: histidine ammonia-lyase\n",
      "COG2987: Catalyzes the conversion of urocanate to 4-imidazolone- 5-propionate\n",
      "COG3275: phosphorelay sensor kinase activity\n",
      "COG3404: glutamate formimidoyltransferase activity\n",
      "COG3448: nan\n",
      "COG3629: intracellular signal transduction\n",
      "COG3643: Formiminotransferase domain\n",
      "COG3758: protein conserved in bacteria\n",
      "COG3775: system Galactitol-specific IIC component\n",
      "COG3899: AAA ATPase domain\n",
      "COG4767: VanZ like family\n",
      "COG5341: Protein conserved in bacteria\n"
     ]
    }
   ],
   "source": [
    "missing_in_no = yes_S_protein - no_S_protein_all\n",
    "print('Loss:', len(missing_in_no), 'OGs')\n",
    "display_ogs(sorted(missing_in_no))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tree annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gtdb_species\n",
       "Streptococcus equi    38\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_subset.loc[accessions_no_S_protein]['gtdb_species'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cwb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
